{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  },
  "orig_nbformat": 2,
  "kernelspec": {
   "name": "python394jvsc74a57bd070de0b75b2a3f35b1bb892b463cd56f2ed390cd02c9f7d4168b3d96fea273b5e",
   "display_name": "Python 3.9.4 64-bit"
  },
  "metadata": {
   "interpreter": {
    "hash": "70de0b75b2a3f35b1bb892b463cd56f2ed390cd02c9f7d4168b3d96fea273b5e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "### pip is the package installer for python3. The command \"py -m pip install\"\n",
    "### installs the packages where in \"-m pip\" executes pip using the latest python interpreter installed in windows.\n",
    "\n",
    "import os\n",
    "import subprocess\n",
    "import re\n",
    "\n",
    "### packages for reading excel files\n",
    "from openpyxl import Workbook\n",
    "from openpyxl import load_workbook\n",
    "\n",
    "### packages for writing word documents\n",
    "from docx import Document\n",
    "from docx.shared import Cm\n",
    "import docx\n",
    "from docx.enum.dml import MSO_THEME_COLOR_INDEX\n",
    "#py -m pip install --user python-docx\n",
    "\n",
    "### packages for webscraping\n",
    "from selenium import webdriver\n",
    "from selenium.webdriver.common.keys import Keys\n",
    "import time\n",
    "from selenium.webdriver.common.by import By\n",
    "from selenium.webdriver.support.ui import WebDriverWait\n",
    "from selenium.webdriver.support import expected_conditions as EC\n",
    "from selenium.webdriver.common.action_chains import ActionChains\n",
    "\n",
    "#Setting up selenium driver\n",
    "#Install the corresponding version of chromedriver on https://sites.google.com/a/chromium.org/chromedriver/downloads and place the file in a map. Store the path to the file in the PATH variable below.\n",
    "'''\n",
    "option = webdriver.ChromeOptions()\n",
    "option.add_argument('headless')\n",
    "PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "driver = webdriver.Chrome(PATH, options=option)\n",
    "'''\n",
    "# package to read the pdf files\n",
    "from tabula import read_pdf\n",
    "from tabulate import tabulate\n",
    "import tabula\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "BRCA1/BRCA2?\n",
      "BRCA1\n",
      "Insert nucleotide\n"
     ]
    }
   ],
   "source": [
    "###########Input\n",
    "print(\"BRCA1/BRCA2?\")\n",
    "brca_input = input()\n",
    "brca_input = brca_input.upper()\n",
    "print(brca_input)\n",
    "print(\"Insert nucleotide\")\n",
    "search = input()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "####Setting up the path to the location of the excel files and pdf file\n",
    "\n",
    "folder_location = \"C:/Users/jensv/Desktop/BRCA_-_prostate_cancer_and_ovarian_cancer_-_PARPi\"\n",
    "functional_categorization = \"/2020 BRCA1 functional categorization of BRCA1 VUS - CCR.xlsx\"\n",
    "variants_excel = \"/allenigmavariants_BICsubmission_2013-07-01.xlsx\"\n",
    "bicbnl = \"/bicbnl 27  fh 2-10-13_nov2013.xlsx\"\n",
    "enigma = \"/BRCA_Multifac_published data_ENIGMAwebsite 2015-03-27.xlsx\"\n",
    "pdf_lindor = \"/Lindor 2012.pdf\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#1) Reading excel sheet: 2020 functional categorization of BRCA1 VUS - CCR\n",
    "\n",
    "### Create word document\n",
    "document = Document()\n",
    "\n",
    "### Add heading to word document\n",
    "document.add_heading(\"2020 functional categorization of BRCA1 VUS - CCR\", 1)\n",
    "\n",
    "### Create table in word\n",
    "table = document.add_table(rows=1, cols=4)\n",
    "\n",
    "### Loading a style template and creating the header cells\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'DNA variant'\n",
    "hdr_cells[1].text = 'Cisplatin Assay'\n",
    "hdr_cells[2].text = 'Olaparib Assay'\n",
    "hdr_cells[3].text = 'DR-GFP assay'\n",
    "\n",
    "\n",
    "#If the input is BRCA1 the script will load the excel file. It will iterate over each row of the first column (max_col = 1) of the active sheet.\n",
    "#If the search input is found in the first column the script will add the values of the 3rd, 4th and 5th column\n",
    "#Example cell.offset(column=2).value means that the script will take the value of the 3rd column (1st column + 2 to the right = 3)\n",
    "if brca_input == \"BRCA1\":\n",
    "\n",
    "    wb = load_workbook(folder_location + functional_categorization)\n",
    "\n",
    "    sheet = wb.active\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 1):\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                    row_cells = table.add_row().cells\n",
    "                    row_cells[0].text = cell.value\n",
    "                    row_cells[1].text = cell.offset(column=2).value\n",
    "                    row_cells[2].text = cell.offset(column=3).value\n",
    "                    row_cells[3].text = cell.offset(column=4).value\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#2) Reading excel sheet : allenigmavariants_BICsubmission_2013-07-01\n",
    "\n",
    "# The first column in this file is BRCA1/BRCA2, the second column is hgvsnucleotide and the fourth column is bic nucleotide\n",
    "# The script will iterate over each row, if the brca_input matches with the gene in the 1st column AND the search is found in the 2nd OR 4th column it will give the class in column 7 as output\n",
    "\n",
    "\n",
    "wb = load_workbook(folder_location + variants_excel)\n",
    "\n",
    "sheet = wb.active\n",
    "\n",
    "document.add_heading(\"allenigmavariants_BICsubmission_2013-07-01\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=4)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'HGSV nucleotide'\n",
    "hdr_cells[2].text = 'BIC nucleotide'\n",
    "hdr_cells[3].text = 'Class'\n",
    "\n",
    "\n",
    "for row in sheet.iter_rows():\n",
    "    for cell in row:\n",
    "        # if BRCA gene is in 1st column AND nucleotide search is in 2ndcolumn (hgvs nucleotide)\n",
    "        if brca_input in str(cell.value) and search in str(cell.offset(column=1).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value #gene\n",
    "            row_cells[1].text = cell.offset(column=1).value #hgvs nucleotide\n",
    "            row_cells[3].text = cell.offset(column=6).value #class\n",
    "        #else if BRCA gene is in 1st column AND nucleotide search is in 4th column (bic nucleotide)  \n",
    "        elif brca_input in str(cell.value) and search in str(cell.offset(column=3).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value #gene\n",
    "            row_cells[2].text = cell.offset(column=3).value #bic nucleotide\n",
    "            row_cells[3].text = cell.offset(column=6).value #class\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#3)Reading excel sheet: Published Multifactoral data enigma\n",
    "\n",
    "# This is likewise to searching in the allenigmavariants_BICsubmission_2013-07-01 excel file\n",
    "\n",
    "\n",
    "wb = load_workbook(folder_location + enigma)\n",
    "\n",
    "sheet = wb.active\n",
    "\n",
    "document.add_heading(\"BRCA_Multifac_published data_ENIGMAwebsite 2015-03-27\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=4)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'HGSV nucleotide'\n",
    "hdr_cells[2].text = 'BIC nucleotide'\n",
    "hdr_cells[3].text = 'Class'\n",
    "\n",
    "\n",
    "for row in sheet.iter_rows():\n",
    "    for cell in row:\n",
    "        if brca_input in str(cell.value) and search in str(cell.offset(column=1).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value\n",
    "            row_cells[1].text = cell.offset(column=1).value\n",
    "            row_cells[3].text = cell.offset(column=6).value\n",
    "            Result = False\n",
    "        elif brca_input in str(cell.value) and search in str(cell.offset(column=3).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value\n",
    "            row_cells[2].text = cell.offset(column=3).value\n",
    "            row_cells[3].text = cell.offset(column=6).value\n",
    "            Result = False\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "C:\\Users\\jensv\\AppData\\Local\\Programs\\Python\\Python39\\lib\\site-packages\\openpyxl\\worksheet\\header_footer.py:48: UserWarning: Cannot parse header or footer so it will be ignored\n  warn(\"\"\"Cannot parse header or footer so it will be ignored\"\"\")\n"
     ]
    }
   ],
   "source": [
    "#4) Reading excel sheet: bicbnl 27  fh 2-10-13_nov2013\n",
    "# Python gives warning it can't parse the header but the script still works\n",
    "# Similar to the other excel scripts except it differentiates between two tabs first.\n",
    "\n",
    "wb = load_workbook(folder_location + bicbnl)\n",
    "\n",
    "document.add_heading(\"bicbnl 27  fh 2-10-13_nov2013\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=5)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'c.nom'\n",
    "hdr_cells[2].text = 'Type'\n",
    "hdr_cells[3].text = 'opm LOB indeling'\n",
    "hdr_cells[4].text = 'opmerking'\n",
    "\n",
    "if brca_input == \"BRCA1\":\n",
    "    # The script will open the right sheet first.\n",
    "    sheet = wb['totaal BRCA1']\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 5, min_col = 5):\n",
    "        # It will search in the 5th column for the nucleotide\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                row_cells = table.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                row_cells[1].text = cell.value\n",
    "                row_cells[2].text = str(cell.offset(column=2).value) # column G (type)\n",
    "                row_cells[3].text = str(cell.offset(column=14).value) # column S (opmerking LOB indeling)\n",
    "                row_cells[4].text = str(cell.offset(column=15).value) # column T (opmerking)\n",
    "\n",
    "elif brca_input == \"BRCA2\":\n",
    "    sheet = wb['totaal BRCA2']\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 5, min_col = 5):\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                row_cells = table.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                row_cells[1].text = cell.value\n",
    "                row_cells[2].text = str(cell.offset(column=2).value)\n",
    "                row_cells[3].text = str(cell.offset(column=14).value)\n",
    "                row_cells[4].text = str(cell.offset(column=15).value)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating a function that allows for the creation of working hyperlinks\n",
    "\n",
    "def add_hyperlink(paragraph, text, url):\n",
    "    # This gets access to the document.xml.rels file and gets a new relation id value\n",
    "    part = paragraph.part\n",
    "    r_id = part.relate_to(url, docx.opc.constants.RELATIONSHIP_TYPE.HYPERLINK, is_external=True)\n",
    "\n",
    "    # Create the w:hyperlink tag and add needed values\n",
    "    hyperlink = docx.oxml.shared.OxmlElement('w:hyperlink')\n",
    "    hyperlink.set(docx.oxml.shared.qn('r:id'), r_id, )\n",
    "\n",
    "    # Create a w:r element and a new w:rPr element\n",
    "    new_run = docx.oxml.shared.OxmlElement('w:r')\n",
    "    rPr = docx.oxml.shared.OxmlElement('w:rPr')\n",
    "\n",
    "    # Join all the xml elements together add add the required text to the w:r element\n",
    "    new_run.append(rPr)\n",
    "    new_run.text = text\n",
    "    hyperlink.append(new_run)\n",
    "\n",
    "    # Create a new Run object and add the hyperlink into it\n",
    "    r = paragraph.add_run ()\n",
    "    r._r.append (hyperlink)\n",
    "\n",
    "    # A workaround for the lack of a hyperlink style (doesn't go purple after using the link)\n",
    "    # Delete this if using a template that has the hyperlink style in it\n",
    "    r.font.color.theme_color = MSO_THEME_COLOR_INDEX.HYPERLINK\n",
    "    r.font.underline = True\n",
    "\n",
    "    return hyperlink\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "# Searching in the LOVD database using selenium webscraping.\n",
    "# The path to the chromedriver needs to be defined in PATH\n",
    "# driver will open and interact with the website and has an optional option\n",
    "# In this case the option is 'headless' which means the driver won't visibly open the website and do the search\n",
    "# removing the options=option part will disable this so you can follow how the search is done\n",
    "\n",
    "# The elements can be found by rightclicking on the page when visiting the website and clicking inspect\n",
    "# The script finds the elements mostly by xpath\n",
    "\n",
    "option = webdriver.ChromeOptions()\n",
    "option.add_argument('headless')\n",
    "PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "#Try is used so the script doesn't fail completely if something goes wrong with one website (it doesn't load, the site got updated, it can't find the search, ...)\n",
    "try:\n",
    "\n",
    "    if brca_input == \"BRCA1\":\n",
    "\n",
    "        # The script will search for the searchbar on the site and enter the nucleotide search in it.\n",
    "\n",
    "        driver.get(\"https://databases.lovd.nl/shared/variants/BRCA1/unique\")\n",
    "        searchbar = driver.find_element_by_name(\"search_VariantOnTranscript/DNA\")\n",
    "        searchbar.send_keys(search)\n",
    "        searchbar.send_keys(Keys.RETURN)\n",
    "    \n",
    "        #time.sleep() is used to pause the script to give the website the time to load\n",
    "\n",
    "        time.sleep(1)\n",
    "\n",
    "        document.add_heading(\"LOVD\")\n",
    "        table_word = document.add_table(rows=1, cols=3)\n",
    "        table_word.style = 'Colorful List Accent 1'\n",
    "        hdr_cells = table_word.rows[0].cells\n",
    "        hdr_cells[0].text = 'Gene'\n",
    "        hdr_cells[1].text = 'DNA Change (cDNA)'\n",
    "        hdr_cells[2].text = 'Clinical Classification'\n",
    "\n",
    "        hit = False\n",
    "    # The script will search for the results table using xpath and iterate over it. hit = False is used so it only adds the relevant rows containing the search query.\n",
    "\n",
    "        count = 0\n",
    "        table = driver.find_element_by_xpath(\"//table[@id='viewlistTable_CustomVL_VOTunique_VOG_BRCA1']\")\n",
    "        for row in table.find_elements_by_xpath(\"//tr[@class='data']\"):\n",
    "            #The script will iterate over the results table in the page row by row\n",
    "            count = 0\n",
    "            for table_data in row.find_elements_by_xpath(\".//td\"):\n",
    "                count = count + 1\n",
    "                # For a row, the script will iterate over each cell\n",
    "                # It will keep a count for each time it goes over a cell\n",
    "                # when it reaches the 4th cell AND the search is in this cell it will append the information to the word document\n",
    "                # hit = True for this row so the script knows to append the information of the 9th cell to the word document aswell\n",
    "                if count == 4 and search in table_data.text:\n",
    "                    row_cells = table_word.add_row().cells\n",
    "                    row_cells[0].text = brca_input\n",
    "                    row_cells[1].text = table_data.text\n",
    "                    hit = True\n",
    "                if count == 9 and hit == True:\n",
    "                    row_cells[2].text = table_data.text\n",
    "                    hit = False\n",
    "\n",
    "        # It will append the results to the word document alongside with a hyperlink of the current url location\n",
    "        p = document.add_paragraph(\"Results: \")    \n",
    "        site_location = str(driver.current_url)\n",
    "        add_hyperlink(p, 'Link', driver.current_url)\n",
    "\n",
    "        driver.quit()\n",
    "\n",
    "    elif brca_input == \"BRCA2\":\n",
    "        #If BRCA2 is given as input it will start the search on the BRCA2 tab of the website\n",
    "        driver.get(\"https://databases.lovd.nl/shared/variants/BRCA2/unique\")\n",
    "        searchbar = driver.find_element_by_name(\"search_VariantOnTranscript/DNA\")\n",
    "        search = \"5095\"\n",
    "        searchbar.send_keys(search)\n",
    "        searchbar.send_keys(Keys.RETURN)\n",
    "    \n",
    "        time.sleep(1)\n",
    "\n",
    "        document.add_heading(\"LOVD\")\n",
    "        table_word = document.add_table(rows=1, cols=3)\n",
    "        table_word.style = 'Colorful List Accent 1'\n",
    "        hdr_cells = table_word.rows[0].cells\n",
    "        hdr_cells[0].text = 'Gene'\n",
    "        hdr_cells[1].text = 'DNA Change (cDNA)'\n",
    "        hdr_cells[2].text = 'Clinical Classification'\n",
    "    \n",
    "        hit = False\n",
    "    \n",
    "        count = 0\n",
    "        table = driver.find_element_by_xpath(\"//table[@id='viewlistTable_CustomVL_VOTunique_VOG_BRCA2']\")\n",
    "        for row in table.find_elements_by_xpath(\"//tr[@class='data']\"):\n",
    "            count = 0\n",
    "            for table_data in row.find_elements_by_xpath(\".//td\"):\n",
    "                count = count + 1\n",
    "                if count == 4 and search in table_data.text:\n",
    "                    row_cells = table_word.add_row().cells\n",
    "                    row_cells[0].text = brca_input\n",
    "                    row_cells[1].text = table_data.text\n",
    "                    hit = True\n",
    "                if count == 9 and hit == True:\n",
    "                    row_cells[2].text = table_data.text\n",
    "                    hit = False\n",
    "\n",
    "        p = document.add_paragraph(\"Results: \")    \n",
    "        site_location = str(driver.current_url)\n",
    "        add_hyperlink(p, 'Link', driver.current_url)\n",
    "        print(driver.current_url)\n",
    "\n",
    "        driver.quit()\n",
    "\n",
    "except Exception:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Searching BRCA Exchange database\"\n",
    "\n",
    "search = \"5074\"\n",
    "try:\n",
    "\n",
    "\n",
    "    option = webdriver.ChromeOptions()\n",
    "    option.add_argument('headless')\n",
    "    PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "    driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "    document.add_heading(\"BRCA Exchange\")\n",
    "    table_word = document.add_table(rows=1, cols=3)\n",
    "    table_word.style = 'Colorful List Accent 1'\n",
    "    hdr_cells = table_word.rows[0].cells\n",
    "    hdr_cells[0].text = 'Gene'\n",
    "    hdr_cells[1].text = 'GVS Nucleotide'\n",
    "    hdr_cells[2].text = 'Clinical Classification'\n",
    "\n",
    "\n",
    "    driver.get(\"https://brcaexchange.org/\")\n",
    "\n",
    "    # Looking for the searchbar and enter the search\n",
    "    searchbar = driver.find_element_by_xpath(\"//input[@type='text']\")\n",
    "    searchbar.send_keys(brca_input + \" \" +search)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "    #wait for the page to load\n",
    "    time.sleep(10)\n",
    "    '''\n",
    "    try:\n",
    "        element = WebDriverWait(driver, 10).until(\n",
    "            EC.presence_of_element_located((By.xpath, \"//*[@id='data-table-container']/div/table/tbody/tr[1]\")))\n",
    "\n",
    "    except TimeoutException:\n",
    "        print(\"Timed out waiting for page to load\")\n",
    "    \n",
    "\n",
    "    table = driver.find_element_by_xpath(\"//*[@id='data-table-container']/div/table/tbody\")\n",
    "'''\n",
    "\n",
    "    hit = False\n",
    "    #iterate over each cell\n",
    "    for row in table.find_elements_by_xpath(\"//tr\"):\n",
    "        count = 0\n",
    "        for table_data in row.find_elements_by_xpath('.//td'):\n",
    "            count = count + 1\n",
    "            if count == 2 and search in table_data.text:\n",
    "                row_cells[0].text = brca_input\n",
    "                row_cells = table_word.add_row().cells\n",
    "                row_cells[1].text = table_data.text\n",
    "                print(table_data.text)\n",
    "                hit = True\n",
    "            if count == 6 and hit == True:\n",
    "                row_cells[2].text = table_data.text\n",
    "                hit = False\n",
    "\n",
    "    #create and save a hyperlink to word\n",
    "    p = document.add_paragraph(\"Results: \")    \n",
    "    site_location = str(driver.current_url)\n",
    "    add_hyperlink(p, 'Link', driver.current_url)\n",
    "\n",
    "    #driver.quit()\n",
    "\n",
    "except Exception:\n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "https://sge.gs.washington.edu/BRCA1/#tab-9159-2\n"
     ]
    }
   ],
   "source": [
    "#Database of Functional Classifications of BRCA1 Variants based on Saturation Genome Editing\n",
    "\n",
    "try:\n",
    "    \n",
    "    if brca_input == \"BRCA1\":\n",
    "\n",
    "        option = webdriver.ChromeOptions()\n",
    "        option.add_argument('headless')\n",
    "        PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "        driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "        document.add_heading(\"Database of Functional Classifications of BRCA1 Variants based on Saturation Genome Editing\")\n",
    "        table_word = document.add_table(rows=1, cols=3)\n",
    "        table_word.style = 'Colorful List Accent 1'\n",
    "        hdr_cells = table_word.rows[0].cells\n",
    "        hdr_cells[0].text = 'Gene'\n",
    "        hdr_cells[1].text = 'Transcript Variant'\n",
    "        hdr_cells[2].text = 'Functional CLass'\n",
    "\n",
    "\n",
    "        driver.get(\"https://sge.gs.washington.edu/BRCA1/#tab-9159-2\")\n",
    "\n",
    "        time.sleep(1)\n",
    "        #The script will look up and click on the right tab\n",
    "        tab = driver.find_element_by_xpath(\"//*[@id='main_page']/div/div/nav/div/ul/li[2]/a\")\n",
    "        tab.click()\n",
    "        '''\n",
    "        tab = driver.find_element_by_xpath(\"//ul[@class='nav navbar-nav']\")\n",
    "        tab.find_element_by_xpath(\".//a[@href='#tab-9159-2']\").click()\n",
    "        '''\n",
    "        #time.sleep(5)\n",
    "        try:\n",
    "            element = WebDriverWait(driver, 10).until(EC.presence_of_element_located((By.ID, \"DataTables_Table_0_filter\")))\n",
    "\n",
    "        except TimeoutException:\n",
    "            print(\"Timed out waiting for page to load\")\n",
    "\n",
    "        # look up the searchbar and enter the query\n",
    "        searchbar = driver.find_element_by_xpath(\"//*[@id='DataTables_Table_0_filter']/label/input\")\n",
    "        searchbar.send_keys(search)\n",
    "        searchbar.send_keys(Keys.RETURN)\n",
    "    \n",
    "        table = driver.find_element_by_xpath(\"//*[@id='DataTables_Table_0']/tbody\")\n",
    " \n",
    "        count = 0\n",
    "        time.sleep(5)\n",
    "        hit = False\n",
    "\n",
    "        for row in table.find_elements_by_xpath(\"//tr\"):\n",
    "            count = 0\n",
    "            for table_data in row.find_elements_by_xpath((\".//td\")):\n",
    "                count = count + 1\n",
    "                if count == 6 and search in table_data.text:\n",
    "                    row_cells = table_word.add_row().cells\n",
    "                    row_cells[0].text = brca_input\n",
    "                    hit = True\n",
    "                    row_cells[1].text = table_data.text \n",
    "                if count == 10 and hit == True:\n",
    "                    row_cells[2].text = table_data.text\n",
    "                    hit = False\n",
    "\n",
    "    p = document.add_paragraph(\"Results: \")    \n",
    "    site_location = str(driver.current_url)\n",
    "    add_hyperlink(p, 'Link', driver.current_url)\n",
    "    print(driver.current_url)\n",
    "\n",
    "    driver.quit()\n",
    "except Exception:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "########Franklin database\n",
    "\n",
    "try:\n",
    "\n",
    "    option = webdriver.ChromeOptions()\n",
    "    option.add_argument('headless')\n",
    "    PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "    driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "    document.add_heading(\"Franklin\")\n",
    "    table_word = document.add_table(rows=1, cols=2)\n",
    "    table_word.style = 'Colorful List Accent 1'\n",
    "    hdr_cells = table_word.rows[0].cells\n",
    "    hdr_cells[0].text = 'Gene'\n",
    "    hdr_cells[1].text = 'Classification'\n",
    "    row_cells = table_word.add_row().cells\n",
    "\n",
    "    driver.get(\"https://franklin.genoox.com/clinical-db/home\")\n",
    "\n",
    "    time.sleep(10)\n",
    "\n",
    "    #a pop up appeared on this page which can be removed by pressing escape\n",
    "    ActionChains(driver).send_keys(Keys.ESCAPE).perform()\n",
    "\n",
    "    #The script will search for the searchbar and the button(somatic)\n",
    "    searchbar = driver.find_element_by_xpath(\"/html/body/app-root/div/gnx-home-page/div/gnx-search/div      [2]/input\")\n",
    "\n",
    "    button = driver.find_element_by_xpath(\"/html/body/app-root/div/gnx-home-page/div/gnx-search/div[2]/     gnx-variant-type-toggle/button[2]\")\n",
    "\n",
    "    button.click()\n",
    "\n",
    "    searchbar.send_keys(search)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "\n",
    "    #time.sleep(10)\n",
    "    try:\n",
    "        element = WebDriverWait(driver, 10).until(EC.presence_of_element_located((By.ID, \"mat-tab-label-2-3\")))\n",
    "\n",
    "    except TimeoutException:\n",
    "        print(\"Timed out waiting for page to load\")\n",
    "\n",
    "    driver.find_element_by_xpath(\"//*[@id='mat-tab-label-2-3']\").click()\n",
    "\n",
    "    time.sleep(3)\n",
    "\n",
    "    classification = driver.find_element_by_xpath(\"//*[@class='indicator-text ng-star-inserted']\")\n",
    "\n",
    "    row_cells[0].text = brca_input\n",
    "    row_cells[1].text = classification.text\n",
    "    franklin_class = classification.text\n",
    "\n",
    "\n",
    "    p = document.add_paragraph(\"Results: \")    \n",
    "    site_location = str(driver.current_url)\n",
    "    add_hyperlink(p, 'Link', driver.current_url)\n",
    "\n",
    "    driver.quit()\n",
    "\n",
    "except Exception:\n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Clinvar\n",
    "try:\n",
    "\n",
    "    option = webdriver.ChromeOptions()\n",
    "    option.add_argument('headless')\n",
    "    PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "    driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "    driver.get(\"https://www.ncbi.nlm.nih.gov/clinvar/\")\n",
    "\n",
    "    document.add_heading(\"Clinvar\")\n",
    "    table_word = document.add_table(rows=1, cols=2)\n",
    "    table_word.style = 'Colorful List Accent 1'\n",
    "    hdr_cells = table_word.rows[0].cells\n",
    "    hdr_cells[0].text = 'Classification'\n",
    "    hdr_cells[1].text = 'Review Status'\n",
    "    row_cells = table_word.add_row().cells\n",
    "\n",
    "    time.sleep(1)\n",
    "\n",
    "    searchbar = driver.find_element_by_xpath(\"//*[@id='term']\")\n",
    "    searchbar.send_keys(brca_input + \" \" + search)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "\n",
    "    table = driver.find_element_by_xpath(\"//*[@id='tabdocsumtable']/tbody\")\n",
    "\n",
    "    driver.find_element_by_xpath(\"//*[@id='tabdocsumtable']/tbody/tr/td[2]/div/a\").click()\n",
    "\n",
    "\n",
    "    #time.sleep(10)\n",
    "    try:\n",
    "        element = WebDriverWait(driver, 10).until(EC.presence_of_element_located((By.class_name, \"fa-star\")))\n",
    "\n",
    "    except TimeoutException:\n",
    "        print(\"Timed out waiting for page to load\")\n",
    "\n",
    "    star_count = 0\n",
    "    classification = driver.find_element_by_xpath(\"//*[@id='main_content']/main/div[3]/div/div/dl/dd[1]\")\n",
    "    row_cells[0].text = classification.text\n",
    "    star_list = driver.find_elements_by_xpath(\"//*[@id='main_content']/main/div[3]/div/div/dl/dd[2]/a/span\")\n",
    "\n",
    "    # On the element of the colored stars have a certain name\n",
    "    # The script will search for every element containing this name and print a * in the the word document for each star it finds on the page\n",
    "    stars = driver.find_elements_by_class_name(\"fa-star\")\n",
    "    for star in stars:\n",
    "        star_count = star_count + 1\n",
    "\n",
    "    row_cells[1].text = star_count * \"*\"\n",
    "\n",
    "    p = document.add_paragraph(\"Results: \")    \n",
    "    site_location = str(driver.current_url)\n",
    "    add_hyperlink(p, 'Link', driver.current_url)\n",
    "    print(driver.current_url)\n",
    "\n",
    "    driver.quit()\n",
    "\n",
    "except Exception:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Varsome login required if too many searches are done\n",
    "\n",
    "try:\n",
    "\n",
    "    option = webdriver.ChromeOptions()\n",
    "    option.add_argument('headless')\n",
    "    PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "    driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "    driver.get(\"https://varsome.com/\")\n",
    "\n",
    "    document.add_heading(\"Varsome\")\n",
    "    table_word = document.add_table(rows=1, cols=2)\n",
    "    table_word.style = 'Colorful List Accent 1'\n",
    "    hdr_cells = table_word.rows[0].cells\n",
    "    hdr_cells[0].text = \"Gene\"\n",
    "    hdr_cells[1].text = 'Classification'\n",
    "    row_cells = table_word.add_row().cells\n",
    "\n",
    "    time.sleep(1)\n",
    "\n",
    "    searchbar = driver.find_element_by_xpath(\"//*[@id='search']\")\n",
    "    searchbar.send_keys(brca_input + \" \" + search)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "\n",
    "    time.sleep(1)\n",
    "\n",
    "    #Popup asking to accept cookies \n",
    "    cookie = driver.find_element_by_xpath(\"//*[@id='onetrust-accept-btn-handler']\")\n",
    "    cookie.click()\n",
    "\n",
    "    button = driver.find_element_by_xpath(\"//*[@id='proceedBtn']/h4\")\n",
    "    button.click()\n",
    "\n",
    "    time.sleep(10)\n",
    "    try:\n",
    "        element = WebDriverWait(driver, 10).until(EC.presence_of_element_located((By.ID, \"acmg_annotation\")))\n",
    "\n",
    "    except TimeoutException:\n",
    "        print(\"Timed out waiting for page to load\")\n",
    "    \n",
    "\n",
    "    classification = driver.find_element_by_xpath(\"//*[@id='acmg_annotation']/div/div[1]/div/div[2]/div/    div/div[1]/div/div/span/div[2]/div/div\")\n",
    "\n",
    "    row_cells = table_word.add_row().cells\n",
    "    row_cells[0].text = brca_input + \" \" + search\n",
    "    row_cells[1].text = classification.text\n",
    "\n",
    "    p = document.add_paragraph(\"Results: \")    \n",
    "    site_location = str(driver.current_url)\n",
    "    add_hyperlink(p, 'Link', driver.current_url)\n",
    "    print(driver.current_url)\n",
    "\n",
    "    driver.quit()\n",
    "\n",
    "except Exception:\n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "https://www.oncokb.org/gene/BRCA1/5074\n"
     ]
    }
   ],
   "source": [
    "#Oncokb.org\n",
    "\n",
    "try:\n",
    "\n",
    "    option = webdriver.ChromeOptions()\n",
    "    option.add_argument('headless')\n",
    "    PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "    driver = webdriver.Chrome(PATH, options = option)\n",
    "\n",
    "    driver.get(\"https://www.oncokb.org/\")\n",
    "\n",
    "    document.add_heading(\"OncoKB\")\n",
    "    table_word = document.add_table(rows=1, cols=2)\n",
    "    table_word.style = 'Colorful List Accent 1'\n",
    "    hdr_cells = table_word.rows[0].cells\n",
    "    hdr_cells[0].text = 'Search'\n",
    "    hdr_cells[1].text = 'Results'\n",
    "\n",
    "    #time.sleep(5)\n",
    "\n",
    "    searchbar = driver.find_element_by_xpath(\"/html/body/div[1]/div[2]/div[3]/div/div/div[3]/div/div/       div/div[1]/div[2]/div/input\")\n",
    "    searchbar.send_keys(brca_input + \" \" + search)\n",
    "    time.sleep(2)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "    time.sleep(5)\n",
    "    result = driver.find_element_by_xpath(\"//*[@id='root']/div[2]/div[3]/div/div/div/div[2]/div/div[2]\")\n",
    "\n",
    "    row_cells = table_word.add_row().cells\n",
    "    row_cells[0].text = brca_input + search\n",
    "    row_cells[1].text = result.text\n",
    "\n",
    "    p = document.add_paragraph(\"Results: \")    \n",
    "    site_location = str(driver.current_url)\n",
    "    add_hyperlink(p, 'Link', driver.current_url)\n",
    "    print(driver.current_url)\n",
    "\n",
    "    driver.quit()\n",
    "    \n",
    "except Exception:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "# The script will read the pdf file containing the tables\n",
    "# It will create a folder and paste the tables in excel files, this step is skipped if this folder already exists\n",
    "# The script will iterate over each excel file in the folder\n",
    "\n",
    "tables_pdf = read_pdf(folder_location + pdf_lindor, pages=\"17-28\")\n",
    "tables_output_location = \"/tables excel lindor 2012\"\n",
    "\n",
    "tables_folder = folder_location + tables_output_location\n",
    "\n",
    "if not os.path.exists(tables_folder):\n",
    "    os.makedirs(tables_folder)\n",
    "\n",
    "for i, table_pdf in enumerate(tables_pdf, start=1):\n",
    "    table_pdf.to_excel(os.path.join(tables_folder, f\"table_{i}.xlsx\"), index=False)\n",
    "\n",
    "# The excel files are divided in different groups based on the similarity of their structure (Equal amount of columns, columns that have the same meaning)\n",
    "group_one = ('table_1.xlsx')\n",
    "group_two = (\"2.xlsx\", \"3.xlsx\", \"4.xlsx\", \"5.xlsx\")\n",
    "group_three = (\"6.xlsx\", \"7.xlsx\", \"8.xlsx\", \"9.xlsx\", \"10.xlsx\")\n",
    "group_four = (\"11.xlsx\")\n",
    "group_five = (\"12.xlsx\")\n",
    "\n",
    "document.add_heading(\"Lindor 2012 pdf\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=4)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'HGSV nucleotide'\n",
    "hdr_cells[2].text = 'BIC nucleotide'\n",
    "hdr_cells[3].text = 'Class'\n",
    "\n",
    "tables_dir = os.listdir(tables_folder)\n",
    "\n",
    "#In the first half it will search for BRCA1 genes\n",
    "\n",
    "for table_file in tables_dir:\n",
    "    if brca_input == \"BRCA1\":\n",
    "        if table_file.endswith(group_one):\n",
    "\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "\n",
    "            for row in sheet.iter_rows():\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[3].text = str(cell.offset(column=5).value)\n",
    "                        #protein?\n",
    "\n",
    "\n",
    "        if table_file.endswith(group_two):\n",
    "\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "\n",
    "            for row in sheet.iter_rows(max_col = 3, min_col = 3):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[2].text = cell.value\n",
    "                        row_cells[3].text = str(cell.offset(column=5).value)\n",
    "                        #bic\n",
    "                        \n",
    "            for row in sheet.iter_rows(max_col = 4, min_col = 4):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[1].text = cell.value\n",
    "                        row_cells[3].text = str(cell.offset(column=4).value)\n",
    "                        #hgvs\n",
    "                        \n",
    "            \n",
    "        if table_file.endswith(group_four):\n",
    "\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "            for row in sheet.iter_rows():\n",
    "                for cell in row:\n",
    "                    if brca_input and search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = cell.value\n",
    "                        row_cells[3].text = str(cell.offset(column=2).value)\n",
    "                        #brca1/brca2\n",
    "#Second half: it will search for the BRCA2 gene in the applicable file\n",
    "    if brca_input == \"BRCA2\":\n",
    "        if table_file.endswith(group_three):\n",
    "\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "            for row in sheet.iter_rows(max_col = 3, min_col = 3):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[2].text = cell.value\n",
    "                        row_cells[3].text = str(cell.offset(column=8).value)\n",
    "                        #bic\n",
    "\n",
    "            for row in sheet.iter_rows(max_col = 4, min_col = 4):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[1].text = cell.value\n",
    "                        row_cells[3].text = str(cell.offset(column=5).value)\n",
    "                        #hgvs\n",
    "\n",
    "        if table_file.endswith(group_four):\n",
    "\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "            for row in sheet.iter_rows():\n",
    "                for cell in row:\n",
    "                    if brca_input and search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = cell.value\n",
    "                        row_cells[3].text = str(cell.offset(column=2).value)\n",
    "                        \n",
    "        if table_file.endswith(group_five):\n",
    "\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "            for row in sheet.iter_rows(max_col = 4, min_col = 4):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[1].text = cell.value\n",
    "                        row_cells[3].text = str(cell.offset(column=5).value)\n",
    "                        #hgvs\n",
    "\n",
    "                     \n",
    "            for row in sheet.iter_rows(max_col = 5, min_col = 5):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[2].text = cell.value\n",
    "                        row_cells[3].text = str(cell.offset(column=4).value)\n",
    "                        #bic\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "####Transformed the text files into excel files for easier use by dragging them into excel.\n",
    "\n",
    "\n",
    "cmg_brca1 = \"/BRCA1 CMG 05 2021.xlsx\"\n",
    "cmg_brca2 = \"/BRCA2 CMG 05 2021.xlsx\"\n",
    "full_path = folder_location + cmg_brca1\n",
    "\n",
    "document.add_heading(\"cmg\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=3)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'Nuc name'\n",
    "hdr_cells[2].text = 'Mut effect'\n",
    "\n",
    "\n",
    "if brca_input == \"BRCA1\":\n",
    "    wb = load_workbook(folder_location + cmg_brca1)\n",
    "    sheet = wb.active\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 8, min_col = 8):\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                row_cells = table.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                row_cells[1].text = cell.value\n",
    "                row_cells[2].text = str(cell.offset(column=2).value)\n",
    "\n",
    "elif brca_input == \"BRCA2\":\n",
    "    sheet = wb.active\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 8, min_col = 8):\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                row_cells = table.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                row_cells[1].text = cell.value\n",
    "                row_cells[2].text = str(cell.offset(column=2).value)\n",
    "\n",
    "document.save('Output.docx')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "'\\npdf_enigma = \"/ENIGMA_Rules_2017-06-29.pdf\"\\nimport tabula\\nimport tabulate\\nfolder_location = \"C:/Users/jensv/Desktop/BRCA_-_prostate_cancer_and_ovarian_cancer_-_PARPi\"\\ntables_enigma = read_pdf(folder_location + pdf_enigma, pages=\\'14-16\\')\\ntables_output_location = \"/tables ENIGMA\"\\n\\nprint(tables_enigma)\\ntables_folder = folder_location + tables_output_location\\n\\nif not os.path.exists(tables_folder):\\n    os.makedirs(tables_folder)\\n\\nfor i, table_pdf in enumerate(tables_enigma, start=1):\\n    table_pdf.to_excel(os.path.join(tables_folder, f\"table_{i}.xlsx\"), index=False)\\n'"
      ]
     },
     "metadata": {},
     "execution_count": 18
    }
   ],
   "source": [
    "'''\r\n",
    "pdf_enigma = \"/ENIGMA_Rules_2017-06-29.pdf\"\r\n",
    "import tabula\r\n",
    "import tabulate\r\n",
    "folder_location = \"C:/Users/jensv/Desktop/BRCA_-_prostate_cancer_and_ovarian_cancer_-_PARPi\"\r\n",
    "tables_enigma = read_pdf(folder_location + pdf_enigma, pages='14-16')\r\n",
    "tables_output_location = \"/tables ENIGMA\"\r\n",
    "\r\n",
    "print(tables_enigma)\r\n",
    "tables_folder = folder_location + tables_output_location\r\n",
    "\r\n",
    "if not os.path.exists(tables_folder):\r\n",
    "    os.makedirs(tables_folder)\r\n",
    "\r\n",
    "for i, table_pdf in enumerate(tables_enigma, start=1):\r\n",
    "    table_pdf.to_excel(os.path.join(tables_folder, f\"table_{i}.xlsx\"), index=False)\r\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "https://pubmed.ncbi.nlm.nih.gov/?term=BRCA1+5074\n"
     ]
    }
   ],
   "source": [
    "# Pubmed\n",
    "p = document.add_paragraph(\"pubmed: \") \n",
    "pubmed_site = \"https://pubmed.ncbi.nlm.nih.gov/?term=\"  \n",
    "pubmed_full = pubmed_site + brca_input + \"+\" + search\n",
    "add_hyperlink(p, 'Link', pubmed_full)\n",
    "print(pubmed_full)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "95 is found between 1 and 101 and has classification Class-5 if at least one clinically relevant residue is removed. Class-3 otherwise.\n95 is found between 1 and 101 and has classification Class-5 if at least one clinically relevant residue is removed. Class-3 otherwise.\n95 is found between 1 and 101 and has classification Class-5 if at least one clinically relevant residue is removed. Class-3 otherwise.\n95 is found between 1 and 101 and has classification Class-5 if at least one clinically relevant residue is removed. Class-3 otherwise.\n95 is found between 1 and 101 and has classification Class-5 if at least one clinically relevant residue is removed. Class-3 otherwise.\n95 is found between 1 and 101 and has classification Class-5 if at least one clinically relevant residue is removed. Class-3 otherwise.\n95 is found between 1 and 101 and has classification Class-5 if at least one clinically relevant residue is removed. Class-3 otherwise.\n95 is found between 81 and 99 and has classification Class-3\nT1685A (c.5053A>G (p.Thr1685Ala)) T1685I (c.5054C>T (p.Thr1685Ile)) V1688del (c.5062_5064del (p.Val1688del)) R1699W (c.5095C>T (p.Arg1699Trp)) G1706E (c.5117G>A (p.Gly1706Glu)) A1708E (c.5123C>A (p.Ala1708Glu)) S1715R (c.5143A>C (p.Ser1715Arg)) G1738R (c.5212G>A (p.Gly1738Arg)) L1764P (c.5291T>C (p.Leu1764Pro)) I1766S (c.5297T>G (p.Ile1766Ser)) M1775K (c.5324T>A (p.Met1775Lys)) M1775R (c.5324T>G (p.Met1775Arg)) C1787S (c.5359T>A (p.Cys1787Ser)) G1788V (c.5363G>T (p.Gly1788Val)) V1838E (c.5513T>A (p.Val1838Glu))\nClass-5 if at least one clinically relevant residue is removed. Class-3 otherwise.\n"
     ]
    }
   ],
   "source": [
    "#Enigma rules manual made table\n",
    "enigma_table_brca1 = \"/ENIGMA RULES BRCA1.xlsx\"\n",
    "enigma_table_brca2 = \"/ENIGMA RULES BRCA2.xlsx\"\n",
    "\n",
    "brca_input = \"BRCA1\"\n",
    "search = \"95\"\n",
    "int_search = int(search)\n",
    "\n",
    "\n",
    "\n",
    "wb = load_workbook(folder_location + enigma_table_brca1)\n",
    "\n",
    "sheet = wb.active\n",
    "\n",
    "document.add_heading(\"Table 3: Catalogue of BRCA1 conserved domains/motifs and currently known clinically important amino acid residues, and relevance for classification of BRCA1 in-frame and terminal exon sequence variants.\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=5)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Domain/motif'\n",
    "hdr_cells[1].text = 'AA start'\n",
    "hdr_cells[2].text = 'AA end'\n",
    "hdr_cells[3].text = 'AA alterations'\n",
    "hdr_cells[3].text = 'Classification'\n",
    "\n",
    "# searching for between AA start and AA end\n",
    "for row in sheet.iter_rows(min_row=2, min_col=2, max_col=2):\n",
    "    for cell in row:\n",
    "        if int_search >= int(cell.value) and int_search <= int(cell.offset(column=1).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = brca_input\n",
    "            AA_start = str(cell.value)\n",
    "            AA_end = str(cell.offset(column=1).value)\n",
    "            classification = str(cell.offset(column=3).value)\n",
    "\n",
    "          #  print(\"{} is found between {} and {} and has classification {}\".format(search, AA_start, AA_end, classification))\n",
    "            \n",
    "###AA alteration with demonstrated clinical importance\n",
    "for row in sheet.iter_rows(min_col=4, max_col=4):\n",
    "    for cell in row:\n",
    "        if search in str(cell.value):\n",
    "            row_cells = table.add_row().cells\n",
    "\n",
    "            AA_alteration = cell.value ####grep?\n",
    "            classification = str(cell.offset(column=1).value)\n",
    "            print(AA_alteration)\n",
    "            print(classification)\n",
    "\n",
    "wb = load_workbook(folder_location + enigma_table_brca2)\n",
    "\n",
    "sheet = wb.active\n",
    "\n",
    "document.add_heading(\"Table 3: Catalogue of BRCA1 conserved domains/motifs and currently known clinically important amino acid residues, and relevance for classification of BRCA1 in-frame and terminal exon sequence variants.\")\n",
    "#row_cells[2].text = str(cell.offset(column=2).value)\n",
    "\n",
    "###BRCA2\n",
    "\n",
    "table = document.add_table(rows=1, cols=5)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Domain/motif'\n",
    "hdr_cells[1].text = 'AA start'\n",
    "hdr_cells[2].text = 'AA end'\n",
    "hdr_cells[3].text = 'AA alterations'\n",
    "hdr_cells[3].text = 'Classification'\n",
    "\n",
    "# searching for between AA start and AA end\n",
    "for row in sheet.iter_rows(min_row=2, min_col=2, max_col=2):\n",
    "    for cell in row:\n",
    "        if int_search >= int(cell.value) and int_search <= int(cell.offset(column=1).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = brca_input\n",
    "            AA_start = str(cell.value)\n",
    "            AA_end = str(cell.offset(column=1).value)\n",
    "            classification = str(cell.offset(column=3).value)\n",
    "\n",
    "            print(\"{} is found between {} and {} and has classification {}\".format(search, AA_start, AA_end, classification))\n",
    "            \n",
    "###AA alteration with demonstrated clinical importance\n",
    "for row in sheet.iter_rows(min_col=4, max_col=4):\n",
    "    for cell in row:\n",
    "        if search in str(cell.value):\n",
    "            row_cells = table.add_row().cells\n",
    "            AA_alteration = cell.value ####grep?\n",
    "            classification = str(cell.offset(column=1).value)\n",
    "            print(AA_alteration)\n",
    "            print(classification)\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ]
}