{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  },
  "orig_nbformat": 2,
  "kernelspec": {
   "name": "python394jvsc74a57bd070de0b75b2a3f35b1bb892b463cd56f2ed390cd02c9f7d4168b3d96fea273b5e",
   "display_name": "Python 3.9.4 64-bit"
  },
  "metadata": {
   "interpreter": {
    "hash": "70de0b75b2a3f35b1bb892b463cd56f2ed390cd02c9f7d4168b3d96fea273b5e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "###pip is the package installer for python3. The command \"py -m pip install\"\n",
    "### installs the packages where in \"-m pip\" executes pip using the latest python\n",
    "### interpreter installed in windows.\n",
    "\n",
    "import os\n",
    "import subprocess\n",
    "import re\n",
    "\n",
    "packagelist = \"openpyxl\", \"beautifulsoup4\"\n",
    "\n",
    "for package in packagelist:\n",
    "    try:\n",
    "        import package\n",
    "    except:\n",
    "        subprocess.call([\"py\",\"-m\",\"pip\", \"install\", package], stdout=subprocess.DEVNULL)\n",
    "       # os.system(\"py -m pip install \" + package)\n",
    "\n",
    "from openpyxl import Workbook\n",
    "from openpyxl import load_workbook\n",
    "from selenium import webdriver\n",
    "from selenium.webdriver.common.keys import Keys\n",
    "import time\n",
    "from selenium.webdriver.common.by import By\n",
    "from selenium.webdriver.support.ui import WebDriverWait\n",
    "from selenium.webdriver.support import expected_conditions as EC\n",
    "from selenium.webdriver.common.action_chains import ActionChains\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "###########Input\n",
    "#print(\"BRCA1/BRCA2?\")\n",
    "#brca_input = input()\n",
    "brca_input = \"BRCA1\"\n",
    "brca_input = brca_input.upper()\n",
    "#print(\"Search?\")\n",
    "#search = input()\n",
    "search = \"4308\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "####Setting up the location of the excel files\n",
    "\n",
    "folder_location = \"C:/Users/jensv/Desktop/BRCA_-_prostate_cancer_and_ovarian_cancer_-_PARPi\"\n",
    "functional_categorization = \"/2020 BRCA1 functional categorization of BRCA1 VUS - CCR.xlsx\"\n",
    "variants_excel = \"/allenigmavariants_BICsubmission_2013-07-01.xlsx\"\n",
    "bicbnl = \"/bicbnl 27  fh 2-10-13_nov2013.xlsx\"\n",
    "enigma = \"/BRCA_Multifac_published data_ENIGMAwebsite 2015-03-27.xlsx\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "##################Write data to excel\n",
    "#Writing data to word\n",
    "from docx import Document\n",
    "from docx.shared import Cm\n",
    "#py -m pip install --user python-docx\n",
    "document = Document()\n",
    "\n",
    "document.add_heading(\"2020 functional categorization of BRCA1 VUS - CCR\", 1)\n",
    "\n",
    "table = document.add_table(rows=1, cols=4)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'DNA variant'\n",
    "hdr_cells[1].text = 'Cisplatin Assay'\n",
    "hdr_cells[2].text = 'Olaparib Assay'\n",
    "hdr_cells[3].text = 'DR-GFP assay'\n",
    "\n",
    "document.save('test.docx')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "output_type": "error",
     "ename": "NameError",
     "evalue": "name 'folder_location' is not defined",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-52-3bfcf5d47e01>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     24\u001b[0m \u001b[1;32mif\u001b[0m \u001b[0mbrca_input\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;34m\"BRCA1\"\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     25\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 26\u001b[1;33m     \u001b[0mwb\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mload_workbook\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfolder_location\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mfunctional_categorization\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     27\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     28\u001b[0m     \u001b[0msheet\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mwb\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mactive\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'folder_location' is not defined"
     ]
    }
   ],
   "source": [
    "#1) Reading excel sheet: 2020 functional categorization of BRCA1 VUS - CCR\n",
    "#The script will load the excel workbook and sheet and create empty lists of the different assays and found dna variants\n",
    "#it will iterate over each cell and append the results to the lists if the input is found in the first column: DNA variant\n",
    "\n",
    "##################Write data to excel\n",
    "#Writing data to word\n",
    "from docx import Document\n",
    "from docx.shared import Cm\n",
    "#py -m pip install --user python-docx\n",
    "document = Document()\n",
    "\n",
    "document.add_heading(\"2020 functional categorization of BRCA1 VUS - CCR\", 1)\n",
    "\n",
    "table = document.add_table(rows=1, cols=4)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'DNA variant'\n",
    "hdr_cells[1].text = 'Cisplatin Assay'\n",
    "hdr_cells[2].text = 'Olaparib Assay'\n",
    "hdr_cells[3].text = 'DR-GFP assay'\n",
    "\n",
    "\n",
    "if brca_input == \"BRCA1\":\n",
    "\n",
    "    wb = load_workbook(folder_location + functional_categorization)\n",
    "\n",
    "    sheet = wb.active\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 1):\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                row_cells = table.add_row().cells\n",
    "                Result = True\n",
    "                if Result == True:\n",
    "                    row_cells[0].text = cell.value\n",
    "                    row_cells[1].text = cell.offset(column=2).value\n",
    "                    row_cells[2].text = cell.offset(column=3).value\n",
    "                    row_cells[3].text = cell.offset(column=4).value\n",
    "                    Result = False\n",
    "\n",
    "document.save('test.docx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "#2) Reading excel sheet : allenigmavariants_BICsubmission_2013-07-01\n",
    "#The script will iterate over cell but will only append to the lists if the search is found in the cell and the right BRCA input is present in the first column\n",
    "#Another elif statement is added which seperates the search found between HGVS and BIC\n",
    "\n",
    "wb = load_workbook(folder_location + variants_excel)\n",
    "\n",
    "sheet = wb.active\n",
    "\n",
    "search = \"4308\"\n",
    "brca_input = \"BRCA1\"\n",
    "\n",
    "document.add_heading(\"allenigmavariants_BICsubmission_2013-07-01\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=4)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'HGSV nucleotide'\n",
    "hdr_cells[2].text = 'BIC nucleotide'\n",
    "hdr_cells[3].text = 'Class'\n",
    "\n",
    "\n",
    "for row in sheet.iter_rows():\n",
    "    for cell in row:\n",
    "        if brca_input in str(cell.value) and search in str(cell.offset(column=1).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value\n",
    "            row_cells[1].text = cell.offset(column=1).value\n",
    "            row_cells[3].text = cell.offset(column=6).value\n",
    "            Result = False\n",
    "            \n",
    "        elif brca_input in str(cell.value) and search in str(cell.offset(column=3).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value\n",
    "            row_cells[2].text = cell.offset(column=3).value\n",
    "            row_cells[3].text = cell.offset(column=6).value\n",
    "            Result = False\n",
    "\n",
    "\n",
    "document.save('test.docx')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "#3)Reading excel sheet: Published Multifactoral data enigma\n",
    "\n",
    "\n",
    "wb = load_workbook(folder_location + enigma)\n",
    "\n",
    "sheet = wb.active\n",
    "\n",
    "\n",
    "search = \"2002\"\n",
    "brca_input = \"BRCA1\"\n",
    "\n",
    "document.add_heading(\"BRCA_Multifac_published data_ENIGMAwebsite 2015-03-27\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=4)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'HGSV nucleotide'\n",
    "hdr_cells[2].text = 'BIC nucleotide'\n",
    "hdr_cells[3].text = 'Class'\n",
    "\n",
    "\n",
    "for row in sheet.iter_rows():\n",
    "    for cell in row:\n",
    "        if brca_input in str(cell.value) and search in str(cell.offset(column=1).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value\n",
    "            row_cells[1].text = cell.offset(column=1).value\n",
    "            row_cells[3].text = cell.offset(column=6).value\n",
    "            Result = False\n",
    "        elif brca_input in str(cell.value) and search in str(cell.offset(column=3).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value\n",
    "            row_cells[2].text = cell.offset(column=3).value\n",
    "            row_cells[3].text = cell.offset(column=6).value\n",
    "            Result = False\n",
    "\n",
    "document.save('test.docx')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "#4) Reading excel sheet: bicbnl 27  fh 2-10-13_nov2013\n",
    "\n",
    "wb = load_workbook(folder_location + bicbnl)\n",
    "\n",
    "search = \"2685\"\n",
    "brca_input = \"BRCA1\"\n",
    "\n",
    "document.add_heading(\"bicbnl 27  fh 2-10-13_nov2013\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=5)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'c.nom'\n",
    "hdr_cells[2].text = 'Type'\n",
    "hdr_cells[3].text = 'opm LOB indeling'\n",
    "hdr_cells[4].text = 'opmerking'\n",
    "\n",
    "if brca_input == \"BRCA1\":\n",
    "    sheet = wb['totaal BRCA1']\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 5, min_col = 5):\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                row_cells = table.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                row_cells[1].text = cell.value\n",
    "                row_cells[2].text = str(cell.offset(column=2).value)\n",
    "                row_cells[3].text = str(cell.offset(column=14).value)\n",
    "                row_cells[4].text = str(cell.offset(column=15).value)\n",
    "                Result = False\n",
    "\n",
    "elif brca_input == \"BRCA2\":\n",
    "    sheet = wb['totaal BRCA2']\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 5, min_col = 5):\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                row_cells = table.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                row_cells[1].text = cell.value\n",
    "                row_cells[2].text = str(cell.offset(column=2).value)\n",
    "                row_cells[3].text = str(cell.offset(column=14).value)\n",
    "                row_cells[4].text = str(cell.offset(column=15).value)\n",
    "                Result = False\n",
    "\n",
    "document.save('test.docx')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Setting up selenium driver\n",
    "#Install the corresponding version of chromedriver on https://sites.google.com/a/chromium.org/chromedriver/downloads and place the file in a map. Store the path to the file in the PATH variable below.\n",
    "option = webdriver.ChromeOptions()\n",
    "option.add_argument('headless')\n",
    "PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "driver = webdriver.Chrome(PATH, options=option)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "#Searching in the LOVD database\n",
    "option = webdriver.ChromeOptions()\n",
    "option.add_argument('headless')\n",
    "PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "brca_input = \"BRCA1\"\n",
    "if brca_input == \"BRCA1\":\n",
    "    driver.get(\"https://databases.lovd.nl/shared/variants/BRCA1/unique\")\n",
    "    searchbar = driver.find_element_by_name(\"search_VariantOnTranscript/DNA\")\n",
    "    search = \"5095\"\n",
    "    searchbar.send_keys(search)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "    \n",
    "    time.sleep(1)\n",
    "\n",
    "    document.add_heading(\"LOVD\")\n",
    "    table_word = document.add_table(rows=1, cols=3)\n",
    "    table_word.style = 'Colorful List Accent 1'\n",
    "    hdr_cells = table_word.rows[0].cells\n",
    "    hdr_cells[0].text = 'Gene'\n",
    "    hdr_cells[1].text = 'DNA Change (cDNA)'\n",
    "    hdr_cells[2].text = 'Clinical Classification'\n",
    "    \n",
    "\n",
    "    count = 0\n",
    "    table = driver.find_element_by_xpath(\"//table[@id='viewlistTable_CustomVL_VOTunique_VOG_BRCA1']\")\n",
    "    for row in table.find_elements_by_xpath(\"//tr[@class='data']\"):\n",
    "        row_cells = table_word.add_row().cells\n",
    "        count = 0\n",
    "        for table_data in row.find_elements_by_xpath(\".//td\"):\n",
    "            row_cells[0].text = brca_input\n",
    "            count = count + 1\n",
    "            if count == 4:\n",
    "                row_cells[1].text = table_data.text\n",
    "            if count == 9:\n",
    "                row_cells[2].text = table_data.text\n",
    "\n",
    "    driver.quit()\n",
    "\n",
    "elif brca_input == \"BRCA2\":\n",
    "    driver.get(\"https://databases.lovd.nl/shared/variants/BRCA2/unique\")\n",
    "    searchbar = driver.find_element_by_name(\"search_VariantOnTranscript/DNA\")\n",
    "    search = \"5095\"\n",
    "    searchbar.send_keys(search)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "    \n",
    "    time.sleep(1)\n",
    "\n",
    "    document.add_heading(\"LOVD\")\n",
    "    table_word = document.add_table(rows=1, cols=3)\n",
    "    table_word.style = 'Colorful List Accent 1'\n",
    "    hdr_cells = table_word.rows[0].cells\n",
    "    hdr_cells[0].text = 'Gene'\n",
    "    hdr_cells[1].text = 'DNA Change (cDNA)'\n",
    "    hdr_cells[2].text = 'Clinical Classification'\n",
    "    \n",
    "    \n",
    "    count = 0\n",
    "    table = driver.find_element_by_xpath(\"//table[@id='viewlistTable_CustomVL_VOTunique_VOG_BRCA2']\")\n",
    "    for row in table.find_elements_by_xpath(\"//tr[@class='data']\"):\n",
    "        row_cells = table_word.add_row().cells\n",
    "        count = 0\n",
    "        for table_data in row.find_elements_by_xpath(\".//td\"):\n",
    "            row_cells[0].text = brca_input\n",
    "            count = count + 1\n",
    "            if count == 4:\n",
    "                row_cells[1].text = table_data.text\n",
    "            if count == 9:\n",
    "                row_cells[2].text = table_data.text\n",
    "\n",
    "    driver.quit()\n",
    "document.save('test.docx') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Searching BRCA Exchange database\n",
    "search = \"5073\"\n",
    "brca_input = \"BRCA2\"\n",
    "\n",
    "option = webdriver.ChromeOptions()\n",
    "option.add_argument('headless')\n",
    "PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "document.add_heading(\"BRCA Exchange\")\n",
    "table_word = document.add_table(rows=1, cols=3)\n",
    "table_word.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table_word.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'GVS Nucleotide'\n",
    "hdr_cells[2].text = 'Clinical Classification'\n",
    "\n",
    "\n",
    "driver.get(\"https://brcaexchange.org/\")\n",
    "\n",
    "searchbar = driver.find_element_by_xpath(\"//input[@type='text']\")\n",
    "searchbar.send_keys(brca_input + \" \" +search)\n",
    "searchbar.send_keys(Keys.RETURN)\n",
    "\n",
    "time.sleep(5)\n",
    "table = driver.find_element_by_xpath(\"//*[@id='data-table-container']/div/table/tbody\")\n",
    "\n",
    "for row in table.find_elements_by_xpath(\"//tr\"):\n",
    "    count = 0\n",
    "    row_cells = table_word.add_row().cells\n",
    "    for table_data in row.find_elements_by_xpath('.//td'):\n",
    "        count = count + 1\n",
    "        if count == 1:\n",
    "            row_cells[0].text = table_data.text\n",
    "        if count == 2:\n",
    "            row_cells[1].text = table_data.text\n",
    "        if count == 6:\n",
    "            row_cells[2].text = table_data.text\n",
    "driver.quit()\n",
    "document.save('test.docx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "exp. chr. pos.hg19 ref alt transcript.variant protein.variant conseq. function.score functional.class RNA.score protein.variant.HGVS\n",
      "exon3 17 41267771 A C c.106T>G p.S36A Missense 0.067965073 FUNC 0.16032109 p.Ser36Ala\n",
      "exon16 17 41219690 C T c.5009G>A p.R1670K Missense -1.552697925 LOF -0.650732372 p.Arg1670Lys\n",
      "exon16 17 41219626 T A c.5073A>T p.T1691= Splice region -1.485894342 LOF -2.71412048 p.Thr1691=\n",
      "exon16 17 41219626 T C c.5073A>G p.T1691= Splice region -0.38257879 FUNC -2.712178636 p.Thr1691=\n",
      "exon16 17 41219626 T G c.5073A>C p.T1691= Splice region -1.426685321 LOF -4.282318624 p.Thr1691=\n",
      "exon21 17 41201191 G T c.5353C>A p.Q1785K Missense -0.283400305 FUNC 0.062507334 p.Gln1785Lys\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "search = \"5073\"\n",
    "brca_input = \"BRCA1\"\n",
    "if brca_input == \"BRCA1\":\n",
    "\n",
    "    option = webdriver.ChromeOptions()\n",
    "    option.add_argument('headless')\n",
    "    PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "    driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "    document.add_heading(\"BRCA Exchange\")\n",
    "    table_word = document.add_table(rows=1, cols=3)\n",
    "    table_word.style = 'Colorful List Accent 1'\n",
    "    hdr_cells = table_word.rows[0].cells\n",
    "    hdr_cells[0].text = 'Gene'\n",
    "    hdr_cells[1].text = 'Transcript Variant'\n",
    "    hdr_cells[2].text = 'Functional CLass'\n",
    "\n",
    "\n",
    "    driver.get(\"https://sge.gs.washington.edu/BRCA1/#tab-9159-2\")\n",
    "\n",
    "    time.sleep(1)\n",
    "    tab = driver.find_element_by_xpath(\"//*[@id='main_page']/div/div/nav/div/ul/li[2]/a\")\n",
    "    tab.click()\n",
    "    '''\n",
    "    tab = driver.find_element_by_xpath(\"//ul[@class='nav navbar-nav']\")\n",
    "    tab.find_element_by_xpath(\".//a[@href='#tab-9159-2']\").click()\n",
    "    '''\n",
    "    time.sleep(5)\n",
    "\n",
    "    searchbar = driver.find_element_by_xpath(\"//*[@id='DataTables_Table_0_filter']/label/input\")\n",
    "    searchbar.send_keys(search)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "    \n",
    "    table = driver.find_element_by_xpath(\"//*[@id='DataTables_Table_0']/tbody\")\n",
    " \n",
    "    count = 0\n",
    "    time.sleep(5)\n",
    "    hit = False\n",
    "\n",
    "    for row in table.find_elements_by_xpath(\"//tr\"):\n",
    "        row_cells = table_word.add_row().cells\n",
    "        count = 0\n",
    "        for table_data in row.find_elements_by_xpath((\".//td\")):\n",
    "            count = count + 1\n",
    "            if count == 6 and search in table_data.text:\n",
    "                row_cells[0].text = brca_input\n",
    "                hit = True\n",
    "                row_cells[1].text = table_data.text \n",
    "            if count == 10 and hit == True:\n",
    "                row_cells[2].text = table_data.text\n",
    "                hit = False\n",
    "     \n",
    "driver.quit()\n",
    "document.save('test2.docx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Likely Pathogenic\n"
     ]
    }
   ],
   "source": [
    "########Franklin database\n",
    "\n",
    "search = \"BRCA2:c.50del\"\n",
    "\n",
    "option = webdriver.ChromeOptions()\n",
    "option.add_argument('headless')\n",
    "PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "document.add_heading(\"Franklin\")\n",
    "table_word = document.add_table(rows=1, cols=2)\n",
    "table_word.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table_word.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'Classification'\n",
    "row_cells = table_word.add_row().cells\n",
    "\n",
    "driver.get(\"https://franklin.genoox.com/clinical-db/home\")\n",
    "\n",
    "time.sleep(10)\n",
    "\n",
    "ActionChains(driver).send_keys(Keys.ESCAPE).perform()\n",
    "searchbar = driver.find_element_by_xpath(\"/html/body/app-root/div/gnx-home-page/div/gnx-search/div[2]/input\")\n",
    "button = driver.find_element_by_xpath(\"/html/body/app-root/div/gnx-home-page/div/gnx-search/div[2]/gnx-variant-type-toggle/button[2]\")\n",
    "\n",
    "button.click()\n",
    "\n",
    "searchbar.send_keys(search)\n",
    "searchbar.send_keys(Keys.RETURN)\n",
    "\n",
    "time.sleep(10)\n",
    "\n",
    "driver.find_element_by_xpath(\"//*[@id='mat-tab-label-2-3']\").click()\n",
    "\n",
    "time.sleep(3)\n",
    "\n",
    "classification = driver.find_element_by_xpath(\"//*[@class='indicator-text ng-star-inserted']\")\n",
    "\n",
    "row_cells[0].text = brca_input\n",
    "row_cells[1].text = classification.text\n",
    "franklin_class = classification.text\n",
    "\n",
    "document.save('test2.docx')\n",
    "driver.quit()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Clinvar\n",
    "brca_input = \"BRCA2\"\n",
    "search = \"c.5073del\"\n",
    "option = webdriver.ChromeOptions()\n",
    "option.add_argument('headless')\n",
    "PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "driver = webdriver.Chrome(PATH, options = option)\n",
    "\n",
    "driver.get(\"https://www.ncbi.nlm.nih.gov/clinvar/\")\n",
    "\n",
    "document.add_heading(\"Clinvar\")\n",
    "table_word = document.add_table(rows=1, cols=2)\n",
    "table_word.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table_word.rows[0].cells\n",
    "hdr_cells[0].text = 'Classification'\n",
    "hdr_cells[1].text = 'Review Status'\n",
    "row_cells = table_word.add_row().cells\n",
    "\n",
    "time.sleep(1)\n",
    "\n",
    "searchbar = driver.find_element_by_xpath(\"//*[@id='term']\")\n",
    "searchbar.send_keys(brca_input + \" \" + search)\n",
    "searchbar.send_keys(Keys.RETURN)\n",
    "\n",
    "table = driver.find_element_by_xpath(\"//*[@id='tabdocsumtable']/tbody\")\n",
    "\n",
    "driver.find_element_by_xpath(\"//*[@id='tabdocsumtable']/tbody/tr/td[2]/div/a\").click()\n",
    "\n",
    "Classification = driver.find_element_by_xpath(\"//*[@id='main_content']/main/div[3]/div/div/dl/dd[1]\")\n",
    "\n",
    "row_cells[0].text = classification.text\n",
    "\n",
    "time.sleep(3)\n",
    "star_count = 0\n",
    "star_list = driver.find_elements_by_xpath(\"//*[@id='main_content']/main/div[3]/div/div/dl/dd[2]/a/span\")\n",
    "stars = driver.find_elements_by_class_name(\"fa-star\")\n",
    "for star in stars:\n",
    "    star_count = star_count + 1\n",
    "\n",
    "row_cells[1].text = star_count * \"*\"\n",
    "'''\n",
    "for row in table.find_elements_by_xpath(\".//tr\"):\n",
    "        row_cells = table_word.add_row().cells\n",
    "        count = 0\n",
    "        for table_data in row.find_elements_by_xpath((\".//td\")):\n",
    "            count = count + 1\n",
    "            if count == 2:\n",
    "                row_cells[0].text = table_data.text\n",
    "            if count == 6:\n",
    "                row_cells[1].text = table_data.text\n",
    "'''\n",
    "driver.quit()\n",
    "document.save(\"test2.docx\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "output_type": "error",
     "ename": "NoSuchElementException",
     "evalue": "Message: no such element: Unable to locate element: {\"method\":\"xpath\",\"selector\":\"//*[@id='acmg_annotation']/div/div[1]/div/div[2]/div/div/div[1]/div/div/span/div[2]/div/div\"}\n  (Session info: chrome=90.0.4430.93)\n",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNoSuchElementException\u001b[0m                    Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-122-57ae62a102bf>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     34\u001b[0m \u001b[0mtime\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msleep\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;36m10\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     35\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 36\u001b[1;33m \u001b[0mclassification\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mdriver\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfind_element_by_xpath\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"//*[@id='acmg_annotation']/div/div[1]/div/div[2]/div/div/div[1]/div/div/span/div[2]/div/div\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     37\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     38\u001b[0m \u001b[0mrow_cells\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mtable_word\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0madd_row\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcells\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Roaming\\Python\\Python39\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\u001b[0m in \u001b[0;36mfind_element_by_xpath\u001b[1;34m(self, xpath)\u001b[0m\n\u001b[0;32m    392\u001b[0m             \u001b[0melement\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mdriver\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfind_element_by_xpath\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'//div/td[1]'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    393\u001b[0m         \"\"\"\n\u001b[1;32m--> 394\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfind_element\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mby\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mBy\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mXPATH\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mxpath\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    395\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    396\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0mfind_elements_by_xpath\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mxpath\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Roaming\\Python\\Python39\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\u001b[0m in \u001b[0;36mfind_element\u001b[1;34m(self, by, value)\u001b[0m\n\u001b[0;32m    974\u001b[0m                 \u001b[0mby\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mBy\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mCSS_SELECTOR\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    975\u001b[0m                 \u001b[0mvalue\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;34m'[name=\"%s\"]'\u001b[0m \u001b[1;33m%\u001b[0m \u001b[0mvalue\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 976\u001b[1;33m         return self.execute(Command.FIND_ELEMENT, {\n\u001b[0m\u001b[0;32m    977\u001b[0m             \u001b[1;34m'using'\u001b[0m\u001b[1;33m:\u001b[0m \u001b[0mby\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    978\u001b[0m             'value': value})['value']\n",
      "\u001b[1;32m~\\AppData\\Roaming\\Python\\Python39\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\u001b[0m in \u001b[0;36mexecute\u001b[1;34m(self, driver_command, params)\u001b[0m\n\u001b[0;32m    319\u001b[0m         \u001b[0mresponse\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcommand_executor\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mexecute\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdriver_command\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mparams\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    320\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mresponse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 321\u001b[1;33m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0merror_handler\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcheck_response\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mresponse\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    322\u001b[0m             response['value'] = self._unwrap_value(\n\u001b[0;32m    323\u001b[0m                 response.get('value', None))\n",
      "\u001b[1;32m~\\AppData\\Roaming\\Python\\Python39\\site-packages\\selenium\\webdriver\\remote\\errorhandler.py\u001b[0m in \u001b[0;36mcheck_response\u001b[1;34m(self, response)\u001b[0m\n\u001b[0;32m    240\u001b[0m                 \u001b[0malert_text\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mvalue\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'alert'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'text'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    241\u001b[0m             \u001b[1;32mraise\u001b[0m \u001b[0mexception_class\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmessage\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mscreen\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mstacktrace\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0malert_text\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 242\u001b[1;33m         \u001b[1;32mraise\u001b[0m \u001b[0mexception_class\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmessage\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mscreen\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mstacktrace\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    243\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    244\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_value_or_default\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mobj\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mdefault\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNoSuchElementException\u001b[0m: Message: no such element: Unable to locate element: {\"method\":\"xpath\",\"selector\":\"//*[@id='acmg_annotation']/div/div[1]/div/div[2]/div/div/div[1]/div/div/span/div[2]/div/div\"}\n  (Session info: chrome=90.0.4430.93)\n"
     ]
    }
   ],
   "source": [
    "#Varsome login required\n",
    "'''\n",
    "brca_input = \"BRCA2\"\n",
    "search = \"c.5073del\"\n",
    "option = webdriver.ChromeOptions()\n",
    "option.add_argument('headless')\n",
    "PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "driver = webdriver.Chrome(PATH)\n",
    "\n",
    "driver.get(\"https://varsome.com/\")\n",
    "\n",
    "document.add_heading(\"Varsome\")\n",
    "table_word = document.add_table(rows=1, cols=2)\n",
    "table_word.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table_word.rows[0].cells\n",
    "hdr_cells[0].text = \"Gene\"\n",
    "hdr_cells[1].text = 'Classification'\n",
    "row_cells = table_word.add_row().cells\n",
    "\n",
    "time.sleep(1)\n",
    "\n",
    "searchbar = driver.find_element_by_xpath(\"//*[@id='search']\")\n",
    "searchbar.send_keys(brca_input + \" \" + search)\n",
    "searchbar.send_keys(Keys.RETURN)\n",
    "\n",
    "time.sleep(1)\n",
    "\n",
    "cookie = driver.find_element_by_xpath(\"//*[@id='onetrust-accept-btn-handler']\")\n",
    "cookie.click()\n",
    "\n",
    "button = driver.find_element_by_xpath(\"//*[@id='proceedBtn']/h4\")\n",
    "button.click()\n",
    "\n",
    "time.sleep(10)\n",
    "\n",
    "classification = driver.find_element_by_xpath(\"//*[@id='acmg_annotation']/div/div[1]/div/div[2]/div/div/div[1]/div/div/span/div[2]/div/div\")\n",
    "\n",
    "row_cells = table_word.add_row().cells\n",
    "row_cells[0].text = brca_input + search\n",
    "row_cells[1].text = classification.text\n",
    "\n",
    "document.save('test2.docx')\n",
    "driver.quit()\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ]
}