{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  },
  "orig_nbformat": 2,
  "kernelspec": {
   "name": "python394jvsc74a57bd070de0b75b2a3f35b1bb892b463cd56f2ed390cd02c9f7d4168b3d96fea273b5e",
   "display_name": "Python 3.9.4 64-bit"
  },
  "metadata": {
   "interpreter": {
    "hash": "70de0b75b2a3f35b1bb892b463cd56f2ed390cd02c9f7d4168b3d96fea273b5e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "###pip is the package installer for python3. The command \"py -m pip install\"\n",
    "### installs the packages where in \"-m pip\" executes pip using the latest python\n",
    "### interpreter installed in windows.\n",
    "\n",
    "import os\n",
    "import subprocess\n",
    "import re\n",
    "'''\n",
    "packagelist = \"openpyxl\", \"beautifulsoup4\"\n",
    "\n",
    "for package in packagelist:\n",
    "    try:\n",
    "        import package\n",
    "    except:\n",
    "        subprocess.call([\"py\",\"-m\",\"pip\", \"install\", package], stdout=subprocess.DEVNULL)\n",
    "       # os.system(\"py -m pip install \" + package)\n",
    "'''\n",
    "\n",
    "from openpyxl import Workbook\n",
    "from openpyxl import load_workbook\n",
    "from selenium import webdriver\n",
    "from selenium.webdriver.common.keys import Keys\n",
    "import time\n",
    "from selenium.webdriver.common.by import By\n",
    "from selenium.webdriver.support.ui import WebDriverWait\n",
    "from selenium.webdriver.support import expected_conditions as EC\n",
    "from selenium.webdriver.common.action_chains import ActionChains\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "###########Input\n",
    "#print(\"BRCA1/BRCA2?\")\n",
    "#brca_input = input()\n",
    "brca_input = \"BRCA1\"\n",
    "brca_input = brca_input.upper()\n",
    "#print(\"Search?\")\n",
    "#search = input()\n",
    "search = \"4308\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "####Setting up the location of the excel files\n",
    "\n",
    "folder_location = \"C:/Users/jensv/Desktop/BRCA_-_prostate_cancer_and_ovarian_cancer_-_PARPi\"\n",
    "functional_categorization = \"/2020 BRCA1 functional categorization of BRCA1 VUS - CCR.xlsx\"\n",
    "variants_excel = \"/allenigmavariants_BICsubmission_2013-07-01.xlsx\"\n",
    "bicbnl = \"/bicbnl 27  fh 2-10-13_nov2013.xlsx\"\n",
    "enigma = \"/BRCA_Multifac_published data_ENIGMAwebsite 2015-03-27.xlsx\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "##################Write data to excel\n",
    "#Writing data to word\n",
    "from docx import Document\n",
    "from docx.shared import Cm\n",
    "#py -m pip install --user python-docx\n",
    "document = Document()\n",
    "\n",
    "document.add_heading(\"2020 functional categorization of BRCA1 VUS - CCR\", 1)\n",
    "\n",
    "table = document.add_table(rows=1, cols=4)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'DNA variant'\n",
    "hdr_cells[1].text = 'Cisplatin Assay'\n",
    "hdr_cells[2].text = 'Olaparib Assay'\n",
    "hdr_cells[3].text = 'DR-GFP assay'\n",
    "\n",
    "document.save('test.docx')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "#1) Reading excel sheet: 2020 functional categorization of BRCA1 VUS - CCR\n",
    "#The script will load the excel workbook and sheet and create empty lists of the different assays and found dna variants\n",
    "#it will iterate over each cell and append the results to the lists if the input is found in the first column: DNA variant\n",
    "\n",
    "##################Write data to excel\n",
    "#Writing data to word\n",
    "from docx import Document\n",
    "from docx.shared import Cm\n",
    "#py -m pip install --user python-docx\n",
    "document = Document()\n",
    "\n",
    "document.add_heading(\"2020 functional categorization of BRCA1 VUS - CCR\", 1)\n",
    "\n",
    "table = document.add_table(rows=1, cols=4)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'DNA variant'\n",
    "hdr_cells[1].text = 'Cisplatin Assay'\n",
    "hdr_cells[2].text = 'Olaparib Assay'\n",
    "hdr_cells[3].text = 'DR-GFP assay'\n",
    "\n",
    "\n",
    "if brca_input == \"BRCA1\":\n",
    "\n",
    "    wb = load_workbook(folder_location + functional_categorization)\n",
    "\n",
    "    sheet = wb.active\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 1):\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                row_cells = table.add_row().cells\n",
    "                Result = True\n",
    "                if Result == True:\n",
    "                    row_cells[0].text = cell.value\n",
    "                    row_cells[1].text = cell.offset(column=2).value\n",
    "                    row_cells[2].text = cell.offset(column=3).value\n",
    "                    row_cells[3].text = cell.offset(column=4).value\n",
    "                    Result = False\n",
    "\n",
    "document.save('test.docx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "#2) Reading excel sheet : allenigmavariants_BICsubmission_2013-07-01\n",
    "#The script will iterate over cell but will only append to the lists if the search is found in the cell and the right BRCA input is present in the first column\n",
    "#Another elif statement is added which seperates the search found between HGVS and BIC\n",
    "\n",
    "wb = load_workbook(folder_location + variants_excel)\n",
    "\n",
    "sheet = wb.active\n",
    "\n",
    "search = \"4308\"\n",
    "brca_input = \"BRCA1\"\n",
    "\n",
    "document.add_heading(\"allenigmavariants_BICsubmission_2013-07-01\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=4)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'HGSV nucleotide'\n",
    "hdr_cells[2].text = 'BIC nucleotide'\n",
    "hdr_cells[3].text = 'Class'\n",
    "\n",
    "\n",
    "for row in sheet.iter_rows():\n",
    "    for cell in row:\n",
    "        if brca_input in str(cell.value) and search in str(cell.offset(column=1).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value\n",
    "            row_cells[1].text = cell.offset(column=1).value\n",
    "            row_cells[3].text = cell.offset(column=6).value\n",
    "            Result = False\n",
    "            \n",
    "        elif brca_input in str(cell.value) and search in str(cell.offset(column=3).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value\n",
    "            row_cells[2].text = cell.offset(column=3).value\n",
    "            row_cells[3].text = cell.offset(column=6).value\n",
    "            Result = False\n",
    "\n",
    "\n",
    "document.save('test.docx')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "#3)Reading excel sheet: Published Multifactoral data enigma\n",
    "\n",
    "\n",
    "wb = load_workbook(folder_location + enigma)\n",
    "\n",
    "sheet = wb.active\n",
    "\n",
    "\n",
    "search = \"2002\"\n",
    "brca_input = \"BRCA1\"\n",
    "\n",
    "document.add_heading(\"BRCA_Multifac_published data_ENIGMAwebsite 2015-03-27\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=4)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'HGSV nucleotide'\n",
    "hdr_cells[2].text = 'BIC nucleotide'\n",
    "hdr_cells[3].text = 'Class'\n",
    "\n",
    "\n",
    "for row in sheet.iter_rows():\n",
    "    for cell in row:\n",
    "        if brca_input in str(cell.value) and search in str(cell.offset(column=1).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value\n",
    "            row_cells[1].text = cell.offset(column=1).value\n",
    "            row_cells[3].text = cell.offset(column=6).value\n",
    "            Result = False\n",
    "        elif brca_input in str(cell.value) and search in str(cell.offset(column=3).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value\n",
    "            row_cells[2].text = cell.offset(column=3).value\n",
    "            row_cells[3].text = cell.offset(column=6).value\n",
    "            Result = False\n",
    "\n",
    "document.save('test.docx')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "C:\\Users\\jensv\\AppData\\Local\\Programs\\Python\\Python39\\lib\\site-packages\\openpyxl\\worksheet\\header_footer.py:48: UserWarning: Cannot parse header or footer so it will be ignored\n  warn(\"\"\"Cannot parse header or footer so it will be ignored\"\"\")\n"
     ]
    }
   ],
   "source": [
    "#4) Reading excel sheet: bicbnl 27  fh 2-10-13_nov2013\n",
    "\n",
    "wb = load_workbook(folder_location + bicbnl)\n",
    "\n",
    "search = \"2685\"\n",
    "brca_input = \"BRCA1\"\n",
    "\n",
    "document.add_heading(\"bicbnl 27  fh 2-10-13_nov2013\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=5)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'c.nom'\n",
    "hdr_cells[2].text = 'Type'\n",
    "hdr_cells[3].text = 'opm LOB indeling'\n",
    "hdr_cells[4].text = 'opmerking'\n",
    "\n",
    "if brca_input == \"BRCA1\":\n",
    "    sheet = wb['totaal BRCA1']\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 5, min_col = 5):\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                row_cells = table.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                row_cells[1].text = cell.value\n",
    "                row_cells[2].text = str(cell.offset(column=2).value)\n",
    "                row_cells[3].text = str(cell.offset(column=14).value)\n",
    "                row_cells[4].text = str(cell.offset(column=15).value)\n",
    "                Result = False\n",
    "\n",
    "elif brca_input == \"BRCA2\":\n",
    "    sheet = wb['totaal BRCA2']\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 5, min_col = 5):\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                row_cells = table.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                row_cells[1].text = cell.value\n",
    "                row_cells[2].text = str(cell.offset(column=2).value)\n",
    "                row_cells[3].text = str(cell.offset(column=14).value)\n",
    "                row_cells[4].text = str(cell.offset(column=15).value)\n",
    "                Result = False\n",
    "\n",
    "document.save('test.docx')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Setting up selenium driver\n",
    "#Install the corresponding version of chromedriver on https://sites.google.com/a/chromium.org/chromedriver/downloads and place the file in a map. Store the path to the file in the PATH variable below.\n",
    "option = webdriver.ChromeOptions()\n",
    "option.add_argument('headless')\n",
    "PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "driver = webdriver.Chrome(PATH, options=option)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "import docx\n",
    "from docx.enum.dml import MSO_THEME_COLOR_INDEX\n",
    "\n",
    "def add_hyperlink(paragraph, text, url):\n",
    "    # This gets access to the document.xml.rels file and gets a new relation id value\n",
    "    part = paragraph.part\n",
    "    r_id = part.relate_to(url, docx.opc.constants.RELATIONSHIP_TYPE.HYPERLINK, is_external=True)\n",
    "\n",
    "    # Create the w:hyperlink tag and add needed values\n",
    "    hyperlink = docx.oxml.shared.OxmlElement('w:hyperlink')\n",
    "    hyperlink.set(docx.oxml.shared.qn('r:id'), r_id, )\n",
    "\n",
    "    # Create a w:r element and a new w:rPr element\n",
    "    new_run = docx.oxml.shared.OxmlElement('w:r')\n",
    "    rPr = docx.oxml.shared.OxmlElement('w:rPr')\n",
    "\n",
    "    # Join all the xml elements together add add the required text to the w:r element\n",
    "    new_run.append(rPr)\n",
    "    new_run.text = text\n",
    "    hyperlink.append(new_run)\n",
    "\n",
    "    # Create a new Run object and add the hyperlink into it\n",
    "    r = paragraph.add_run ()\n",
    "    r._r.append (hyperlink)\n",
    "\n",
    "    # A workaround for the lack of a hyperlink style (doesn't go purple after using the link)\n",
    "    # Delete this if using a template that has the hyperlink style in it\n",
    "    r.font.color.theme_color = MSO_THEME_COLOR_INDEX.HYPERLINK\n",
    "    r.font.underline = True\n",
    "\n",
    "    return hyperlink\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "#Searching in the LOVD database\n",
    "option = webdriver.ChromeOptions()\n",
    "option.add_argument('headless')\n",
    "PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "brca_input = \"BRCA1\"\n",
    "if brca_input == \"BRCA1\":\n",
    "    driver.get(\"https://databases.lovd.nl/shared/variants/BRCA1/unique\")\n",
    "    searchbar = driver.find_element_by_name(\"search_VariantOnTranscript/DNA\")\n",
    "    search = \"5095\"\n",
    "    searchbar.send_keys(search)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "    \n",
    "    time.sleep(1)\n",
    "\n",
    "    document.add_heading(\"LOVD\")\n",
    "    table_word = document.add_table(rows=1, cols=3)\n",
    "    table_word.style = 'Colorful List Accent 1'\n",
    "    hdr_cells = table_word.rows[0].cells\n",
    "    hdr_cells[0].text = 'Gene'\n",
    "    hdr_cells[1].text = 'DNA Change (cDNA)'\n",
    "    hdr_cells[2].text = 'Clinical Classification'\n",
    "    \n",
    "\n",
    "    count = 0\n",
    "    table = driver.find_element_by_xpath(\"//table[@id='viewlistTable_CustomVL_VOTunique_VOG_BRCA1']\")\n",
    "    for row in table.find_elements_by_xpath(\"//tr[@class='data']\"):\n",
    "        row_cells = table_word.add_row().cells\n",
    "        count = 0\n",
    "        for table_data in row.find_elements_by_xpath(\".//td\"):\n",
    "            row_cells[0].text = brca_input\n",
    "            count = count + 1\n",
    "            if count == 4:\n",
    "                row_cells[1].text = table_data.text\n",
    "            if count == 9:\n",
    "                row_cells[2].text = table_data.text\n",
    "\n",
    "    p = document.add_paragraph(\"Results: \")    \n",
    "    site_location = str(driver.current_url)\n",
    "    add_hyperlink(p, 'Link', driver.current_url)\n",
    "\n",
    "    driver.quit()\n",
    "\n",
    "elif brca_input == \"BRCA2\":\n",
    "    driver.get(\"https://databases.lovd.nl/shared/variants/BRCA2/unique\")\n",
    "    searchbar = driver.find_element_by_name(\"search_VariantOnTranscript/DNA\")\n",
    "    search = \"5095\"\n",
    "    searchbar.send_keys(search)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "    \n",
    "    time.sleep(1)\n",
    "\n",
    "    document.add_heading(\"LOVD\")\n",
    "    table_word = document.add_table(rows=1, cols=3)\n",
    "    table_word.style = 'Colorful List Accent 1'\n",
    "    hdr_cells = table_word.rows[0].cells\n",
    "    hdr_cells[0].text = 'Gene'\n",
    "    hdr_cells[1].text = 'DNA Change (cDNA)'\n",
    "    hdr_cells[2].text = 'Clinical Classification'\n",
    "    \n",
    "    \n",
    "    count = 0\n",
    "    table = driver.find_element_by_xpath(\"//table[@id='viewlistTable_CustomVL_VOTunique_VOG_BRCA2']\")\n",
    "    for row in table.find_elements_by_xpath(\"//tr[@class='data']\"):\n",
    "        row_cells = table_word.add_row().cells\n",
    "        count = 0\n",
    "        for table_data in row.find_elements_by_xpath(\".//td\"):\n",
    "            row_cells[0].text = brca_input\n",
    "            count = count + 1\n",
    "            if count == 4:\n",
    "                row_cells[1].text = table_data.text\n",
    "            if count == 9:\n",
    "                row_cells[2].text = table_data.text\n",
    "\n",
    "    p = document.add_paragraph(\"Results: \")    \n",
    "    site_location = str(driver.current_url)\n",
    "    add_hyperlink(p, 'Link', driver.current_url)\n",
    "    print(driver.current_url)\n",
    "\n",
    "    driver.quit()\n",
    "document.save('test.docx') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Searching BRCA Exchange database\n",
    "search = \"5073\"\n",
    "brca_input = \"BRCA2\"\n",
    "\n",
    "option = webdriver.ChromeOptions()\n",
    "option.add_argument('headless')\n",
    "PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "document.add_heading(\"BRCA Exchange\")\n",
    "table_word = document.add_table(rows=1, cols=3)\n",
    "table_word.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table_word.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'GVS Nucleotide'\n",
    "hdr_cells[2].text = 'Clinical Classification'\n",
    "\n",
    "\n",
    "driver.get(\"https://brcaexchange.org/\")\n",
    "\n",
    "searchbar = driver.find_element_by_xpath(\"//input[@type='text']\")\n",
    "searchbar.send_keys(brca_input + \" \" +search)\n",
    "searchbar.send_keys(Keys.RETURN)\n",
    "\n",
    "time.sleep(5)\n",
    "table = driver.find_element_by_xpath(\"//*[@id='data-table-container']/div/table/tbody\")\n",
    "\n",
    "for row in table.find_elements_by_xpath(\"//tr\"):\n",
    "    count = 0\n",
    "    row_cells = table_word.add_row().cells\n",
    "    for table_data in row.find_elements_by_xpath('.//td'):\n",
    "        count = count + 1\n",
    "        if count == 1:\n",
    "            row_cells[0].text = table_data.text\n",
    "        if count == 2:\n",
    "            row_cells[1].text = table_data.text\n",
    "        if count == 6:\n",
    "            row_cells[2].text = table_data.text\n",
    "\n",
    "p = document.add_paragraph(\"Results: \")    \n",
    "site_location = str(driver.current_url)\n",
    "add_hyperlink(p, 'Link', driver.current_url)\n",
    "\n",
    "driver.quit()\n",
    "\n",
    "document.save('test.docx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "https://sge.gs.washington.edu/BRCA1/#tab-9159-2\n"
     ]
    }
   ],
   "source": [
    "#Washington\n",
    "\n",
    "search = \"eeeeeeegzgezrgzergzergz\"\n",
    "brca_input = \"BRCA1\"\n",
    "try:\n",
    "    \n",
    "    if brca_input == \"BRCA1\":\n",
    "\n",
    "        option = webdriver.ChromeOptions()\n",
    "        option.add_argument('headless')\n",
    "        PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "        driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "        document.add_heading(\"Washington\")\n",
    "        table_word = document.add_table(rows=1, cols=3)\n",
    "        table_word.style = 'Colorful List Accent 1'\n",
    "        hdr_cells = table_word.rows[0].cells\n",
    "        hdr_cells[0].text = 'Gene'\n",
    "        hdr_cells[1].text = 'Transcript Variant'\n",
    "        hdr_cells[2].text = 'Functional CLass'\n",
    "\n",
    "\n",
    "        driver.get(\"https://sge.gs.washington.edu/BRCA1/#tab-9159-2\")\n",
    "\n",
    "        time.sleep(1)\n",
    "        tab = driver.find_element_by_xpath(\"//*[@id='main_page']/div/div/nav/div/ul/li[2]/a\")\n",
    "        tab.click()\n",
    "        '''\n",
    "        tab = driver.find_element_by_xpath(\"//ul[@class='nav navbar-nav']\")\n",
    "        tab.find_element_by_xpath(\".//a[@href='#tab-9159-2']\").click()\n",
    "        '''\n",
    "        time.sleep(5)\n",
    "\n",
    "        searchbar = driver.find_element_by_xpath(\"//*[@id='DataTables_Table_0_filter']/label/input\")\n",
    "        searchbar.send_keys(search)\n",
    "        searchbar.send_keys(Keys.RETURN)\n",
    "    \n",
    "        table = driver.find_element_by_xpath(\"//*[@id='DataTables_Table_0']/tbody\")\n",
    " \n",
    "        count = 0\n",
    "        time.sleep(5)\n",
    "        hit = False\n",
    "\n",
    "        for row in table.find_elements_by_xpath(\"//tr\"):\n",
    "            row_cells = table_word.add_row().cells\n",
    "            count = 0\n",
    "            for table_data in row.find_elements_by_xpath((\".//td\")):\n",
    "                count = count + 1\n",
    "                if count == 6 and search in table_data.text:\n",
    "                    row_cells[0].text = brca_input\n",
    "                    hit = True\n",
    "                    row_cells[1].text = table_data.text \n",
    "                if count == 10 and hit == True:\n",
    "                    row_cells[2].text = table_data.text\n",
    "                    hit = False\n",
    "\n",
    "    p = document.add_paragraph(\"Results: \")    \n",
    "    site_location = str(driver.current_url)\n",
    "    add_hyperlink(p, 'Link', driver.current_url)\n",
    "    print(driver.current_url)\n",
    "\n",
    "    driver.quit()\n",
    "except Exception:\n",
    "    pass\n",
    "document.save('test2.docx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "########Franklin database\n",
    "\n",
    "search = \"fzefzef\"\n",
    "\n",
    "try:\n",
    "\n",
    "    option = webdriver.ChromeOptions()\n",
    "    option.add_argument('headless')\n",
    "    PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "    driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "    document.add_heading(\"Franklin\")\n",
    "    table_word = document.add_table(rows=1, cols=2)\n",
    "    table_word.style = 'Colorful List Accent 1'\n",
    "    hdr_cells = table_word.rows[0].cells\n",
    "    hdr_cells[0].text = 'Gene'\n",
    "    hdr_cells[1].text = 'Classification'\n",
    "    row_cells = table_word.add_row().cells\n",
    "\n",
    "    driver.get(\"https://franklin.genoox.com/clinical-db/home\")\n",
    "\n",
    "    time.sleep(10)\n",
    "\n",
    "    ActionChains(driver).send_keys(Keys.ESCAPE).perform()\n",
    "    searchbar = driver.find_element_by_xpath(\"/html/body/app-root/div/gnx-home-page/div/gnx-search/div      [2]/input\")\n",
    "    button = driver.find_element_by_xpath(\"/html/body/app-root/div/gnx-home-page/div/gnx-search/div[2]/     gnx-variant-type-toggle/button[2]\")\n",
    "\n",
    "    button.click()\n",
    "\n",
    "    searchbar.send_keys(search)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "\n",
    "    time.sleep(10)\n",
    "\n",
    "    driver.find_element_by_xpath(\"//*[@id='mat-tab-label-2-3']\").click()\n",
    "\n",
    "    time.sleep(3)\n",
    "\n",
    "    classification = driver.find_element_by_xpath(\"//*[@class='indicator-text ng-star-inserted']\")\n",
    "\n",
    "    row_cells[0].text = brca_input\n",
    "    row_cells[1].text = classification.text\n",
    "    franklin_class = classification.text\n",
    "\n",
    "\n",
    "    p = document.add_paragraph(\"Results: \")    \n",
    "    site_location = str(driver.current_url)\n",
    "    add_hyperlink(p, 'Link', driver.current_url)\n",
    "\n",
    "    driver.quit()\n",
    "\n",
    "except Exception:\n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "https://www.ncbi.nlm.nih.gov/clinvar/variation/51762/\n"
     ]
    }
   ],
   "source": [
    "#Clinvar\n",
    "brca_input = \"BRCA2\"\n",
    "search = \"c.5073del\"\n",
    "try:\n",
    "\n",
    "    option = webdriver.ChromeOptions()\n",
    "    option.add_argument('headless')\n",
    "    PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "    driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "    driver.get(\"https://www.ncbi.nlm.nih.gov/clinvar/\")\n",
    "\n",
    "    document.add_heading(\"Clinvar\")\n",
    "    table_word = document.add_table(rows=1, cols=2)\n",
    "    table_word.style = 'Colorful List Accent 1'\n",
    "    hdr_cells = table_word.rows[0].cells\n",
    "    hdr_cells[0].text = 'Classification'\n",
    "    hdr_cells[1].text = 'Review Status'\n",
    "    row_cells = table_word.add_row().cells\n",
    "\n",
    "    time.sleep(1)\n",
    "\n",
    "    searchbar = driver.find_element_by_xpath(\"//*[@id='term']\")\n",
    "    searchbar.send_keys(brca_input + \" \" + search)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "\n",
    "    table = driver.find_element_by_xpath(\"//*[@id='tabdocsumtable']/tbody\")\n",
    "\n",
    "    driver.find_element_by_xpath(\"//*[@id='tabdocsumtable']/tbody/tr/td[2]/div/a\").click()\n",
    "\n",
    "\n",
    "    time.sleep(10)\n",
    "    star_count = 0\n",
    "    classification = driver.find_element_by_xpath(\"//*[@id='main_content']/main/div[3]/div/div/dl/dd[1]\")\n",
    "    row_cells[0].text = classification.text\n",
    "    star_list = driver.find_elements_by_xpath(\"//*[@id='main_content']/main/div[3]/div/div/dl/dd[2]/a/span\")\n",
    "    stars = driver.find_elements_by_class_name(\"fa-star\")\n",
    "    for star in stars:\n",
    "        star_count = star_count + 1\n",
    "\n",
    "    row_cells[1].text = star_count * \"*\"\n",
    "\n",
    "    p = document.add_paragraph(\"Results: \")    \n",
    "    site_location = str(driver.current_url)\n",
    "    add_hyperlink(p, 'Link', driver.current_url)\n",
    "    print(driver.current_url)\n",
    "\n",
    "    driver.quit()\n",
    "    document.save(\"test2.docx\")\n",
    "\n",
    "except Exception:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Varsome login required\n",
    "\n",
    "brca_input = \"BRCA2\"\n",
    "search = \"c.5073del\"\n",
    "try:\n",
    "\n",
    "    option = webdriver.ChromeOptions()\n",
    "    option.add_argument('headless')\n",
    "    PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "    driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "    driver.get(\"https://varsome.com/\")\n",
    "\n",
    "    document.add_heading(\"Varsome\")\n",
    "    table_word = document.add_table(rows=1, cols=2)\n",
    "    table_word.style = 'Colorful List Accent 1'\n",
    "    hdr_cells = table_word.rows[0].cells\n",
    "    hdr_cells[0].text = \"Gene\"\n",
    "    hdr_cells[1].text = 'Classification'\n",
    "    row_cells = table_word.add_row().cells\n",
    "\n",
    "    time.sleep(1)\n",
    "\n",
    "    searchbar = driver.find_element_by_xpath(\"//*[@id='search']\")\n",
    "    searchbar.send_keys(brca_input + \" \" + search)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "\n",
    "    time.sleep(1)\n",
    "\n",
    "    cookie = driver.find_element_by_xpath(\"//*[@id='onetrust-accept-btn-handler']\")\n",
    "    cookie.click()\n",
    "\n",
    "    button = driver.find_element_by_xpath(\"//*[@id='proceedBtn']/h4\")\n",
    "    button.click()\n",
    "\n",
    "    time.sleep(10)\n",
    "\n",
    "    classification = driver.find_element_by_xpath(\"//*[@id='acmg_annotation']/div/div[1]/div/div[2]/div/    div/div[1]/div/div/span/div[2]/div/div\")\n",
    "\n",
    "    row_cells = table_word.add_row().cells\n",
    "    row_cells[0].text = brca_input + search\n",
    "    row_cells[1].text = classification.text\n",
    "\n",
    "    document.save('test2.docx')\n",
    "\n",
    "    p = document.add_paragraph(\"Results: \")    \n",
    "    site_location = str(driver.current_url)\n",
    "    add_hyperlink(p, 'Link', driver.current_url)\n",
    "    print(driver.current_url)\n",
    "\n",
    "    driver.quit()\n",
    "\n",
    "except Exception:\n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "https://www.oncokb.org/gene/BRCA2/c.5073del\n"
     ]
    }
   ],
   "source": [
    "#Oncokb.org\n",
    "\n",
    "brca_input = \"BRCA2\"\n",
    "search = \"c.5073del\"\n",
    "\n",
    "try:\n",
    "\n",
    "    option = webdriver.ChromeOptions()\n",
    "    option.add_argument('headless')\n",
    "    PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "    driver = webdriver.Chrome(PATH, options = option)\n",
    "\n",
    "    driver.get(\"https://www.oncokb.org/\")\n",
    "\n",
    "    document.add_heading(\"OncoKB\")\n",
    "    table_word = document.add_table(rows=1, cols=2)\n",
    "    table_word.style = 'Colorful List Accent 1'\n",
    "    hdr_cells = table_word.rows[0].cells\n",
    "    hdr_cells[0].text = 'Search'\n",
    "    hdr_cells[1].text = 'Results'\n",
    "\n",
    "    time.sleep(5)\n",
    "\n",
    "    searchbar = driver.find_element_by_xpath(\"/html/body/div[1]/div[2]/div[3]/div/div/div[3]/div/div/       div/div[1]/div[2]/div/input\")\n",
    "    searchbar.send_keys(brca_input + \" \" + search)\n",
    "    time.sleep(2)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "    time.sleep(5)\n",
    "    result = driver.find_element_by_xpath(\"//*[@id='root']/div[2]/div[3]/div/div/div/div[2]/div/div[2]\")\n",
    "\n",
    "    row_cells = table_word.add_row().cells\n",
    "    row_cells[0].text = brca_input + search\n",
    "    row_cells[1].text = result.text\n",
    "\n",
    "    p = document.add_paragraph(\"Results: \")    \n",
    "    site_location = str(driver.current_url)\n",
    "    add_hyperlink(p, 'Link', driver.current_url)\n",
    "    print(driver.current_url)\n",
    "\n",
    "    driver.quit()\n",
    "    document.save('test2.docx')\n",
    "    \n",
    "except Exception:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "c.7436-14T>G\nhgvs\n2\n"
     ]
    }
   ],
   "source": [
    "from tabula import read_pdf\n",
    "from tabulate import tabulate\n",
    "import tabula\n",
    "folder_location = \"C:/Users/jensv/Desktop/BRCA_-_prostate_cancer_and_ovarian_cancer_-_PARPi\"\n",
    "pdf_test = \"/Lindor 2012.pdf\"\n",
    "\n",
    "\n",
    "tables_pdf = read_pdf(folder_location + pdf_test, pages=\"17-28\")\n",
    "tables_output_location = \"/tables excel lindor 2012\"\n",
    "\n",
    "tables_folder = folder_location + tables_output_location\n",
    "\n",
    "if not os.path.exists(tables_folder):\n",
    "    os.makedirs(tables_folder)\n",
    "\n",
    "for i, table in enumerate(tables_pdf, start=1):\n",
    "    table.to_excel(os.path.join(tables_folder, f\"table_{i}.xlsx\"), index=False)\n",
    "\n",
    "group_one = ('table_1.xlsx')\n",
    "group_two = (\"2.xlsx\", \"3.xlsx\", \"4.xlsx\", \"5.xlsx\")\n",
    "group_three = (\"6.xlsx\", \"7.xlsx\", \"8.xlsx\", \"9.xlsx\", \"10.xlsx\")\n",
    "group_four = (\"11.xlsx\")\n",
    "group_five = (\"12.xlsx\")\n",
    "\n",
    "\n",
    "search = \"7436\"\n",
    "brca_input = \"BRCA2\"\n",
    "\n",
    "\n",
    "tables_dir = os.listdir(tables_folder)\n",
    "for table_file in tables_dir:\n",
    "    if brca_input == \"BRCA1\":\n",
    "        if table_file.endswith(group_one):\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "            for row in sheet.iter_rows():\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        print(str(cell.offset(column=5).value))\n",
    "                        #protein?\n",
    "\n",
    "\n",
    "        elif table_file.endswith(group_two):\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "            for row in sheet.iter_rows(max_col = 3, min_col = 3):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        print(\"bic\")\n",
    "                        print(str(cell.offset(column=5).value))\n",
    "            for row in sheet.iter_rows(max_col = 4, min_col = 4):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        print(\"hgvs\")\n",
    "                        print(str(cell.offset(column=4).value))\n",
    "            \n",
    "        elif table_file.endswith(group_four):\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "            for row in sheet.iter_rows():\n",
    "                for cell in row:\n",
    "                    if brca_input and search in str(cell.value):\n",
    "                        print(str(cell.offset(column=2).value))\n",
    "\n",
    "    if brca_input == \"BRCA2\":\n",
    "        if table_file.endswith(group_three):\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "            for row in sheet.iter_rows(max_col = 3, min_col = 3):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        print(\"bic\")\n",
    "                        print(str(cell.offset(column=8).value))\n",
    "            for row in sheet.iter_rows(max_col = 4, min_col = 4):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        print(\"hgvs\")\n",
    "                        print(str(cell.offset(column=5).value))\n",
    "                        \n",
    "        elif table_file.endswith(group_four):\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "            for row in sheet.iter_rows():\n",
    "                for cell in row:\n",
    "                    if brca_input and search in str(cell.value):\n",
    "                        print(str(cell.offset(column=2).value))\n",
    "                        \n",
    "        elif table_file.endswith(group_five):\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "            for row in sheet.iter_rows(max_col = 4, min_col = 4):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        print(str(cell.value))\n",
    "                        print(\"hgvs\")\n",
    "                        print(str(cell.offset(column=5).value)) \n",
    "\n",
    "                     \n",
    "            for row in sheet.iter_rows(max_col = 5, min_col = 5):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        print(\"bic\")\n",
    "                        print(str(cell.offset(column=4).value))\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "####text files\n",
    "\n",
    "search = \"192\"\n",
    "brca_input = \"BRCA1\"\n",
    "\n",
    "cmg_brca1 = \"/BRCA1 CMG 05 2021.xlsx\"\n",
    "cmg_brca2 = \"/BRCA2 CMG 05 2021.xlsx\"\n",
    "full_path = folder_location + cmg_brca1\n",
    "\n",
    "document.add_heading(\"cmg\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=3)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'Nuc name'\n",
    "hdr_cells[2].text = 'Mut effect'\n",
    "\n",
    "\n",
    "if brca_input == \"BRCA1\":\n",
    "    wb = load_workbook(folder_location + cmg_brca1)\n",
    "    sheet = wb.active\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 8, min_col = 8):\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                row_cells = table.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                row_cells[1].text = cell.value\n",
    "                row_cells[2].text = str(cell.offset(column=2).value)\n",
    "\n",
    "elif brca_input == \"BRCA2\":\n",
    "    sheet = wb.active\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 8, min_col = 8):\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                row_cells = table.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                row_cells[1].text = cell.value\n",
    "                row_cells[2].text = str(cell.offset(column=2).value)\n",
    "\n",
    "document.save('test.docx')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ]
}