{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  },
  "orig_nbformat": 2,
  "kernelspec": {
   "name": "python394jvsc74a57bd070de0b75b2a3f35b1bb892b463cd56f2ed390cd02c9f7d4168b3d96fea273b5e",
   "display_name": "Python 3.9.4 64-bit"
  },
  "metadata": {
   "interpreter": {
    "hash": "70de0b75b2a3f35b1bb892b463cd56f2ed390cd02c9f7d4168b3d96fea273b5e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 429,
   "metadata": {},
   "outputs": [],
   "source": [
    "### pip is the package installer for python3. The command \"py -m pip install\"\n",
    "### installs the packages where in \"-m pip\" executes pip using the latest python interpreter installed in windows.\n",
    "\n",
    "import os\n",
    "import subprocess\n",
    "import re\n",
    "\n",
    "### packages for reading excel files\n",
    "from openpyxl import Workbook\n",
    "from openpyxl import load_workbook\n",
    "\n",
    "### packages for writing word documents\n",
    "from docx import Document\n",
    "from docx.shared import Cm\n",
    "import docx\n",
    "from docx.enum.dml import MSO_THEME_COLOR_INDEX\n",
    "#py -m pip install --user python-docx\n",
    "\n",
    "### packages for webscraping\n",
    "from selenium import webdriver\n",
    "from selenium.webdriver.common.keys import Keys\n",
    "import time\n",
    "from selenium.webdriver.common.by import By\n",
    "from selenium.webdriver.support.ui import WebDriverWait\n",
    "from selenium.webdriver.support import expected_conditions as EC\n",
    "from selenium.webdriver.common.action_chains import ActionChains\n",
    "\n",
    "#Setting up selenium driver\n",
    "#Install the corresponding version of chromedriver on https://sites.google.com/a/chromium.org/chromedriver/downloads and place the file in a map. Store the path to the file in the PATH variable below.\n",
    "'''\n",
    "option = webdriver.ChromeOptions()\n",
    "option.add_argument('headless')\n",
    "PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "driver = webdriver.Chrome(PATH, options=option)\n",
    "'''\n",
    "# package to read the pdf files\n",
    "from tabula import read_pdf\n",
    "from tabulate import tabulate\n",
    "import tabula\n",
    "\n",
    "import csv\n",
    "\n",
    "import re\n",
    "import requests as rq\n",
    "import json\n",
    "import pprint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 430,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creating a function that allows for the creation of working hyperlinks\n",
    "\n",
    "def add_hyperlink(paragraph, text, url):\n",
    "    # This gets access to the document.xml.rels file and gets a new relation id value\n",
    "    part = paragraph.part\n",
    "    r_id = part.relate_to(url, docx.opc.constants.RELATIONSHIP_TYPE.HYPERLINK, is_external=True)\n",
    "\n",
    "    # Create the w:hyperlink tag and add needed values\n",
    "    hyperlink = docx.oxml.shared.OxmlElement('w:hyperlink')\n",
    "    hyperlink.set(docx.oxml.shared.qn('r:id'), r_id, )\n",
    "\n",
    "    # Create a w:r element and a new w:rPr element\n",
    "    new_run = docx.oxml.shared.OxmlElement('w:r')\n",
    "    rPr = docx.oxml.shared.OxmlElement('w:rPr')\n",
    "\n",
    "    # Join all the xml elements together add add the required text to the w:r element\n",
    "    new_run.append(rPr)\n",
    "    new_run.text = text\n",
    "    hyperlink.append(new_run)\n",
    "\n",
    "    # Create a new Run object and add the hyperlink into it\n",
    "    r = paragraph.add_run ()\n",
    "    r._r.append (hyperlink)\n",
    "\n",
    "    # A workaround for the lack of a hyperlink style (doesn't go purple after using the link)\n",
    "    # Delete this if using a template that has the hyperlink style in it\n",
    "    r.font.color.theme_color = MSO_THEME_COLOR_INDEX.HYPERLINK\n",
    "    r.font.underline = True\n",
    "\n",
    "    return hyperlink\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 431,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "BRCA1/BRCA2?\n",
      "BRCA2\n",
      "Insert nucleotide\n"
     ]
    }
   ],
   "source": [
    "###########Input\n",
    "print(\"BRCA1/BRCA2?\")\n",
    "brca_input = input()\n",
    "brca_input = brca_input.upper()\n",
    "print(brca_input)\n",
    "print(\"Insert nucleotide\")\n",
    "search = input()\n",
    "\n",
    "if \".c\" in search:\n",
    "    search_c = search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 432,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "####Setting up the path to the location of the excel files and pdf file\n",
    "\n",
    "folder_location = \"C:/Users/jensv/Desktop/BRCA_-_prostate_cancer_and_ovarian_cancer_-_PARPi\"\n",
    "functional_categorization = \"/2020 BRCA1 functional categorization of BRCA1 VUS - CCR.xlsx\"\n",
    "variants_excel = \"/allenigmavariants_BICsubmission_2013-07-01.xlsx\"\n",
    "bicbnl = \"/bicbnl 27  fh 2-10-13_nov2013.xlsx\"\n",
    "enigma = \"/BRCA_Multifac_published data_ENIGMAwebsite 2015-03-27.xlsx\"\n",
    "pdf_lindor = \"/Lindor 2012.pdf\"\n",
    "cmg_brca1 = \"/BRCA1 CMG 05 2021.txt\"\n",
    "cmg_brca2 = \"/BRCA2 CMG 05 2021.txt\"\n",
    "brca1_exchange = \"/brca1_exchange\"\n",
    "brca2_exchange = \"/brca2_exchange\"\n",
    "new_path = \"C:/Users/jensv/Stage_UZA\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 433,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Word document\n",
    "\n",
    "### Create word document\n",
    "document = Document()\n",
    "sections = document.sections\n",
    "for section in sections:\n",
    "    section.top_margin = Cm(1)\n",
    "    section.bottom_margin = Cm(1)\n",
    "    section.left_margin = Cm(1)\n",
    "    section.right_margin = Cm(1)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 434,
   "metadata": {},
   "outputs": [],
   "source": [
    "### UZA CMG\n",
    "### priority: 1\n",
    "### The script will open the text file depending on if the gene input is BRCA1/2\n",
    "\n",
    "document.add_heading(\"UZA CMG\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=3)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'Nuc name'\n",
    "hdr_cells[2].text = 'Mut effect'\n",
    "\n",
    "if brca_input == \"BRCA1\":\n",
    "    full_path = folder_location + cmg_brca1\n",
    "    with open(full_path) as csv_file:\n",
    "        # The script will open the text file using csv.reader but using the tab (\"\\t\") delimiter so it will iterate over each line of text separated by tabs.\n",
    "        csv_reader = csv.reader(csv_file, delimiter=\"\\t\")\n",
    "        count = 0\n",
    "        hit = False\n",
    "        for row in csv_reader:\n",
    "            for cell in row:\n",
    "                if search in cell:\n",
    "                    row_cells = table.add_row().cells\n",
    "                    row_cells[0].text = brca_input\n",
    "                    row_cells[1].text = cell\n",
    "                    hit = True\n",
    "                    count = count + 1\n",
    "                elif hit == True:\n",
    "                    count = count + 1\n",
    "                    hit = False\n",
    "                elif count == 2:\n",
    "                    count = 0\n",
    "                    row_cells[2].text = cell\n",
    "\n",
    "elif brca_input == \"BRCA2\":\n",
    "    full_path = folder_location + cmg_brca2\n",
    "    with open(full_path) as csv_file:\n",
    "        csv_reader = csv.reader(csv_file, delimiter=\"\\t\")\n",
    "        count = 0\n",
    "        hit = False\n",
    "        for row in csv_reader:\n",
    "            for cell in row:\n",
    "                if search in cell:\n",
    "                    row_cells = table.add_row().cells\n",
    "                    row_cells[0].text = brca_input\n",
    "                    row_cells[1].text = cell\n",
    "                    hit = True\n",
    "                    count = count + 1\n",
    "                elif hit == True:\n",
    "                    count = count + 1\n",
    "                    hit = False\n",
    "                elif count == 2:\n",
    "                    count = 0\n",
    "                    row_cells[2].text = cell\n",
    "                    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 463,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Downloading the brca exchange databases locally\n",
    "download_database = \"YES\"\n",
    "if download_database == \"YES\":\n",
    "    query_brca1 = \"\"\"\n",
    "https://brcaexchange.org/backend/data/\n",
    "?format=json\n",
    "&search_term=BRCA1\n",
    "&include=Variant_in_ENIGMA\n",
    "&include=Variant_in_ClinVar\n",
    "&include=Variant_in_1000_Genomes\n",
    "&include=Variant_in_ExAC\n",
    "&include=Variant_in_LOVD\n",
    "&include=Variant_in_BIC\n",
    "&include=Variant_in_ESP\n",
    "&include=Variant_in_exLOVD\n",
    "&include=Variant_in_Findlay_BRCA1_Ring_Function_Scores\n",
    "&include=Variant_in_GnomAD\n",
    "&include=Variant_in_GnomADv3\n",
    "\"\"\"\n",
    "    response = rq.get(query_brca1.replace('\\n', '')).json()['data']\n",
    "    with open('brca1_exchange', 'w') as brca1_data:\n",
    "        json.dump(response, brca1_data)\n",
    "\n",
    "    query_brca2 = \"\"\"\n",
    "https://brcaexchange.org/backend/data/\n",
    "?format=json\n",
    "&search_term=BRCA1\n",
    "&include=Variant_in_ENIGMA\n",
    "&include=Variant_in_ClinVar\n",
    "&include=Variant_in_1000_Genomes\n",
    "&include=Variant_in_ExAC\n",
    "&include=Variant_in_LOVD\n",
    "&include=Variant_in_BIC\n",
    "&include=Variant_in_ESP\n",
    "&include=Variant_in_exLOVD\n",
    "&include=Variant_in_Findlay_BRCA1_Ring_Function_Scores\n",
    "&include=Variant_in_GnomAD\n",
    "&include=Variant_in_GnomADv3\n",
    "\"\"\"\n",
    "    response = rq.get(query_brca2.replace('\\n', '')).json()['data']\n",
    "    with open('brca2_exchange', 'w') as brca2_data:\n",
    "        json.dump(response, brca2_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 462,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "<Element {http://schemas.openxmlformats.org/wordprocessingml/2006/main}hyperlink at 0x248c23728c0>"
      ]
     },
     "metadata": {},
     "execution_count": 462
    }
   ],
   "source": [
    "#2\n",
    "#BRCA exchange\n",
    "\n",
    "# https://github.com/BRCAChallenge/brca-exchange/blob/master/website/content/api_docs/api_overview.md\n",
    "\n",
    "### Generating a list of numbers and converting them from integers into strings. The script will use this list as a comparison, if the search output gives a nucleotide sequence of +5 or -5 it will not be given as output because the list of disallowed numbers ranges from 4 to 9999 in this case, this can be changed below.\n",
    "pattern = list(range(4, 9999))\n",
    "pattern_string = [str(number) for number in pattern]\n",
    "\n",
    "document.add_heading(\"BRCA Exchange\")\n",
    "table_word = document.add_table(rows=1, cols=4)\n",
    "table_word.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table_word.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'HGVS Nucleotide'\n",
    "hdr_cells[2].text = 'HGVS Protein'\n",
    "hdr_cells[3].text = 'Clinical Classification'\n",
    "\n",
    "search = \"c.8954-5A>G\"\n",
    "### The script will split and transform the search input\n",
    "\n",
    "if brca_input == \"BRCA1\":\n",
    "    full_path_brca = new_path + brca1_exchange\n",
    "\n",
    "    f = open (full_path_brca, \"r\")\n",
    "\n",
    "    brca_data = json.loads(f.read())\n",
    "\n",
    "    if (\"+\" or \"-\") in search:\n",
    "        # if + or - is specified in the search the script will NOT split the search string and compare it to the list.\n",
    "        for data in brca_data:\n",
    "            if search in data[\"HGVS_cDNA\"]:\n",
    "                row_cells = table_word.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                cDNA = data[\"HGVS_cDNA\"]\n",
    "                # The accession number of each result is attached to the output in this case, it will first be splitted of.\n",
    "                cDNA_split = cDNA.split(\":\")\n",
    "                row_cells[1].text = cDNA_split[1]\n",
    "                row_cells[2].text = data[\"HGVS_Protein\"]\n",
    "                row_cells[3].text = data[\"Pathogenicity_expert\"]\n",
    "                \n",
    "    for data in brca_data:\n",
    "        if search in data[\"HGVS_cDNA\"]:\n",
    "            #It will print the outcome if there is no + in the output\n",
    "            if (\"+\" or \"-\") not in data[\"HGVS_cDNA\"]:\n",
    "                row_cells = table_word.add_row().cells\n",
    "                # Removing the accession numbers before the nucleotide/protein\n",
    "                row_cells[0].text = brca_input\n",
    "                cDNA = data[\"HGVS_cDNA\"]\n",
    "                cDNA_split = cDNA.split(\":\")\n",
    "                row_cells[1].text = cDNA_split[1]\n",
    "\n",
    "                pDNA = data[\"HGVS_Protein\"]\n",
    "                pDNA_split = pDNA.split(\":\")\n",
    "                row_cells[2].text = pDNA_split[1]                \n",
    "                row_cells[3].text = data[\"Pathogenicity_expert\"]\n",
    "\n",
    "            #If there is a plus sign in the output it will start to deconstruct the string and search the for the nucleotide numbers in the range. If there is not a match it will print the output.\n",
    "            elif \"+\" in data[\"HGVS_cDNA\"]:\n",
    "\n",
    "                hit = data[\"HGVS_cDNA\"]\n",
    "\n",
    "                # In each step the string will be further deconstructed resulting in only the nucleotide sequence\n",
    "                plus = hit.split(\"+\")\n",
    "                plus_two = plus[1]\n",
    "                plus_three = re.split('([a-zA-Z\\s]+)', plus_two)\n",
    "                plus_four = plus_three[0]\n",
    "                plus_five = plus_four.split('_')\n",
    "                plus_six = plus_five[0]\n",
    "\n",
    "\n",
    "                if not any(i == plus_six for i in pattern_string):\n",
    "                    row_cells = table_word.add_row().cells\n",
    "                    row_cells[0].text = brca_input\n",
    "                    cDNA = data[\"HGVS_cDNA\"]\n",
    "                    cDNA_split = cDNA.split(\":\")\n",
    "                    row_cells[1].text = cDNA_split[1]\n",
    "                   \n",
    "\n",
    "                    pDNA = data[\"HGVS_Protein\"]\n",
    "                    pDNA_split = pDNA.split(\":\")\n",
    "                    row_cells[2].text = pDNA_split[1]                \n",
    "                    row_cells[3].text = data[\"Pathogenicity_expert\"]\n",
    "\n",
    "\n",
    "            elif \"-\" in data[\"HGVS_cDNA\"]:\n",
    "\n",
    "                minus = hit.split(\"-\")\n",
    "                minus_two = minus[1]\n",
    "                minus_three = re.split('([a-zA-Z\\s]+)', plus_minus)\n",
    "                minus_four = minus_three[0]\n",
    "                minus_five = minus_four.split('_')\n",
    "                minus_six = minus_five[0]\n",
    "\n",
    "                if not any( i == minus_six for i in pattern_string):\n",
    "                    row_cells = table_word.add_row().cells\n",
    "                    row_cells[0].text = brca_input\n",
    "                    cDNA = data[\"HGVS_cDNA\"]\n",
    "                    cDNA_split = cDNA.split(\":\")\n",
    "                    row_cells[1].text = cDNA_split[1]\n",
    "\n",
    "                    pDNA = data[\"HGVS_Protein\"]\n",
    "                    pDNA_split = pDNA.split(\":\")\n",
    "                    row_cells[2].text = pDNA_split[1]                \n",
    "                    row_cells[3].text = data[\"Pathogenicity_expert\"]\n",
    "\n",
    "\n",
    "    #### if there is a match in the protein key it will also extract the data as output\n",
    "\n",
    "        elif search in data[\"HGVS_Protein\"]:\n",
    "                row_cells = table_word.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                cDNA = data[\"HGVS_cDNA\"]\n",
    "                cDNA_split = cDNA.split(\":\")\n",
    "                row_cells[1].text = cDNA_split[1]\n",
    "\n",
    "                pDNA = data[\"HGVS_Protein\"]\n",
    "                pDNA_split = pDNA.split(\":\")\n",
    "                row_cells[2].text = pDNA_split[1]                \n",
    "                row_cells[3].text = data[\"Pathogenicity_expert\"]\n",
    "\n",
    "if brca_input == \"BRCA2\":\n",
    "    full_path_brca = new_path + brca2_exchange\n",
    "\n",
    "    f = open (full_path_brca, \"r\")\n",
    "\n",
    "    brca_data = json.loads(f.read())\n",
    "\n",
    "    if (\"-\") in search:\n",
    "        for data in brca_data:\n",
    "            if search in data[\"HGVS_cDNA\"]:\n",
    "                row_cells = table_word.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                cDNA = data[\"HGVS_cDNA\"]\n",
    "                cDNA_split = cDNA.split(\":\")\n",
    "                row_cells[1].text = cDNA_split[1]\n",
    "                row_cells[2].text = data[\"HGVS_Protein\"]\n",
    "                row_cells[3].text = data[\"Pathogenicity_expert\"]\n",
    "                print(search)\n",
    "\n",
    "    for data in brca_data:\n",
    "        if search in data[\"HGVS_cDNA\"]:\n",
    "            #It will print the outcome if there is no + in the output\n",
    "            if (\"+\" or \"-\") not in data[\"HGVS_cDNA\"]:\n",
    "                row_cells = table_word.add_row().cells\n",
    "                # Removing the accession numbers before the nucleotide/protein\n",
    "                row_cells[0].text = brca_input\n",
    "                cDNA = data[\"HGVS_cDNA\"]\n",
    "                cDNA_split = cDNA.split(\":\")\n",
    "                row_cells[1].text = cDNA_split[1]\n",
    "\n",
    "                pDNA = data[\"HGVS_Protein\"]\n",
    "                pDNA_split = pDNA.split(\":\")\n",
    "                row_cells[2].text = pDNA_split[1]                \n",
    "                row_cells[3].text = data[\"Pathogenicity_expert\"]\n",
    "\n",
    "            #If there is a plus sign in the output it will start to deconstruct the string and search the for the nucleotide numbers in the range. If there is not a match it will print the output.\n",
    "            elif \"+\" in data[\"HGVS_cDNA\"]:\n",
    "\n",
    "                hit = data[\"HGVS_cDNA\"]\n",
    "\n",
    "                # In each step the string will be further deconstructed resulting in only the nucleotide sequence\n",
    "                plus = hit.split(\"+\")\n",
    "                plus_two = plus[1]\n",
    "                plus_three = re.split('([a-zA-Z\\s]+)', plus_two)\n",
    "                plus_four = plus_three[0]\n",
    "                plus_five = plus_four.split('_')\n",
    "                plus_six = plus_five[0]\n",
    "\n",
    "\n",
    "                if not any(i == plus_six for i in pattern_string):\n",
    "                    row_cells = table_word.add_row().cells\n",
    "                    row_cells[0].text = brca_input\n",
    "                    cDNA = data[\"HGVS_cDNA\"]\n",
    "                    cDNA_split = cDNA.split(\":\")\n",
    "                    row_cells[1].text = cDNA_split[1]\n",
    "\n",
    "                    pDNA = data[\"HGVS_Protein\"]\n",
    "                    pDNA_split = pDNA.split(\":\")\n",
    "                    row_cells[2].text = pDNA_split[1]                \n",
    "                    row_cells[3].text = data[\"Pathogenicity_expert\"]\n",
    "\n",
    "\n",
    "            elif \"-\" in data[\"HGVS_cDNA\"]:\n",
    "\n",
    "                minus = hit.split(\"-\")\n",
    "                minus_two = minus[1]\n",
    "                minus_three = re.split('([a-zA-Z\\s]+)', plus_minus)\n",
    "                minus_four = minus_three[0]\n",
    "                minus_five = minus_four.split('_')\n",
    "                minus_six = minus_five[0]\n",
    "\n",
    "                if not any( i == minus_six for i in pattern_string):\n",
    "                    row_cells = table_word.add_row().cells\n",
    "                    row_cells[0].text = brca_input\n",
    "                    cDNA = data[\"HGVS_cDNA\"]\n",
    "                    cDNA_split = cDNA.split(\":\")\n",
    "                    row_cells[1].text = cDNA_split[1]\n",
    "\n",
    "                    pDNA = data[\"HGVS_Protein\"]\n",
    "                    pDNA_split = pDNA.split(\":\")\n",
    "                    row_cells[2].text = pDNA_split[1]                \n",
    "                    row_cells[3].text = data[\"Pathogenicity_expert\"]\n",
    "\n",
    "\n",
    "    #### if there is a match in the protein key it will also extract the data as output\n",
    "\n",
    "        elif search in data[\"HGVS_Protein\"]:\n",
    "                row_cells = table_word.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                cDNA = data[\"HGVS_cDNA\"]\n",
    "                cDNA_split = cDNA.split(\":\")\n",
    "                row_cells[1].text = cDNA_split[1]\n",
    "\n",
    "                pDNA = data[\"HGVS_Protein\"]\n",
    "                pDNA_split = pDNA.split(\":\")\n",
    "                row_cells[2].text = pDNA_split[1]                \n",
    "                row_cells[3].text = data[\"Pathogenicity_expert\"]\n",
    "\n",
    "#create and save a hyperlink to word\n",
    "current_url = \"https://brcaexchange.org/variants?search={}%20{}\".format(brca_input, search)\n",
    "p = document.add_paragraph(\"Results: \")    \n",
    "site_location = str(current_url)\n",
    "add_hyperlink(p, 'Link', current_url)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 437,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "https://databases.lovd.nl/shared/variants/BRCA2/unique#object_id=VariantOnTranscriptUnique%2CVariantOnGenome&id=BRCA2&order=VariantOnTranscript%2FDNA%2CASC&search_transcriptid=00003479&search_VariantOnTranscript/DNA=c.8954-5A%3EG&page_size=100&page=1\n"
     ]
    }
   ],
   "source": [
    "# 3\n",
    "# Searching in the LOVD database using selenium webscraping.\n",
    "# The path to the chromedriver needs to be defined in PATH\n",
    "# driver will open and interact with the website and has an optional option\n",
    "# In this case the option is 'headless' which means the driver won't visibly open the website and do the search\n",
    "# removing the options=option part will disable this so you can follow how the search is done\n",
    "\n",
    "# The elements can be found by rightclicking on the page when visiting the website and clicking inspect\n",
    "# The script finds the elements mostly by xpath\n",
    "pattern = list(range(4, 9999))\n",
    "pattern_string = [str(number) for number in pattern]\n",
    "option = webdriver.ChromeOptions()\n",
    "\n",
    "option.add_argument('headless')\n",
    "PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "#Try is used so the script doesn't fail completely if something goes wrong with one website (it doesn't load, the site got updated, it can't find the search, ...)\n",
    "try:\n",
    "\n",
    "    if brca_input == \"BRCA1\":\n",
    "\n",
    "        # The script will search for the searchbar on the site and enter the nucleotide search in it.\n",
    "\n",
    "        driver.get(\"https://databases.lovd.nl/shared/variants/BRCA1/unique\")\n",
    "        searchbar = driver.find_element_by_name(\"search_VariantOnTranscript/DNA\")\n",
    "        searchbar.send_keys(search)\n",
    "        searchbar.send_keys(Keys.RETURN)\n",
    "    \n",
    "        #time.sleep() is used to pause the script to give the website the time to load\n",
    "\n",
    "        time.sleep(1)\n",
    "\n",
    "        document.add_heading(\"LOVD\")\n",
    "        table_word = document.add_table(rows=1, cols=4)\n",
    "        table_word.style = 'Colorful List Accent 1'\n",
    "        hdr_cells = table_word.rows[0].cells\n",
    "        hdr_cells[0].text = 'Gene'\n",
    "        hdr_cells[1].text = 'DNA Change (cDNA)'\n",
    "        hdr_cells[2].text = 'Clinical Classification'\n",
    "        hdr_cells[3].text = 'Protein'\n",
    "\n",
    "        hit = False\n",
    "    # The script will search for the results table using xpath and iterate over it. hit = False is used so it only adds the relevant rows containing the search query.\n",
    "\n",
    "        count = 0\n",
    "        table = driver.find_element_by_xpath(\"//table[@id='viewlistTable_CustomVL_VOTunique_VOG_BRCA1']\")\n",
    "        for row in table.find_elements_by_xpath(\"//tr[@class='data']\"):\n",
    "            #The script will iterate over the results table in the page row by row\n",
    "            count = 0\n",
    "            for table_data in row.find_elements_by_xpath(\".//td\"):\n",
    "                count = count + 1\n",
    "                # For a row, the script will iterate over each cell\n",
    "                # It will keep a count for each time it goes over a cell\n",
    "                # when it reaches the 4th cell AND the search is in this cell it will append the information to the word document\n",
    "                # hit = True for this row so the script knows to append the information of the 9th cell to the word document aswell\n",
    "                if count == 4 and search in table_data.text:\n",
    "                    if (\"+\" or \"-\") in search:\n",
    "                        row_cells = table_word.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[1].text = table_data.text\n",
    "                        hit = True\n",
    "            \n",
    "\n",
    "                    elif (\"+\" or \"-\") not in table_data.text:\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[1].text = table_data.text\n",
    "                        hit = True\n",
    "                    \n",
    "                    elif \"+\" in table_data.text:\n",
    "                    \n",
    "                        hit = table_data.text      \n",
    "                        plus = hit.split(\"+\")\n",
    "                        plus_two = plus[1]\n",
    "                        plus_three = re.split('([a-zA-Z\\s]+)', plus_two)\n",
    "                        plus_four = plus_three[0]\n",
    "                        plus_five = plus_four.split('_')\n",
    "                        plus_six = plus_five[0]\n",
    "\n",
    "                        if not any(i == plus_six for i in pattern_string):\n",
    "\n",
    "                            row_cells = table_word.add_row().cells\n",
    "                            row_cells[0].text = brca_input\n",
    "                            row_cells[1].text = table_data.text\n",
    "                            hit = True\n",
    "\n",
    "                    elif \"-\" in table_data.text:\n",
    "\n",
    "                        minus = hit.split(\"-\")\n",
    "                        minus_two = minus[1]\n",
    "                        minus_three = re.split('([a-zA-Z\\s]+)', plus_minus)\n",
    "                        minus_four = minus_three[0]\n",
    "                        minus_five = minus_four.split('_')\n",
    "                        minus_six = minus_five[0]\n",
    "\n",
    "                        if not any( i == minus_six for i in pattern_string):\n",
    "                            row_cells = table_word.add_row().cells\n",
    "                            row_cells[0].text = brca_input\n",
    "                            row_cells[1].text = table_data.text\n",
    "                            hit = True\n",
    "\n",
    "                elif count == 6 and search in table_data.text:\n",
    "                    row_cells[O].text = brca_input\n",
    "                    row_cells[3].text = table_data.text\n",
    "                    hit = True\n",
    "\n",
    "\n",
    "                if count == 9 and hit == True:\n",
    "                    row_cells[2].text = table_data.text\n",
    "                    hit = False               \n",
    "\n",
    "\n",
    "\n",
    "        # It will append the results to the word document alongside with a hyperlink of the current url location\n",
    "        p = document.add_paragraph(\"Results: \")    \n",
    "        site_location = str(driver.current_url)\n",
    "        add_hyperlink(p, 'Link', driver.current_url)\n",
    "\n",
    "        driver.quit()\n",
    "\n",
    "    elif brca_input == \"BRCA2\":\n",
    "        #If BRCA2 is given as input it will start the search on the BRCA2 tab of the website\n",
    "        driver.get(\"https://databases.lovd.nl/shared/variants/BRCA2/unique\")\n",
    "        searchbar = driver.find_element_by_name(\"search_VariantOnTranscript/DNA\")\n",
    "        searchbar.send_keys(search)\n",
    "        searchbar.send_keys(Keys.RETURN)\n",
    "    \n",
    "        time.sleep(1)\n",
    "\n",
    "        document.add_heading(\"LOVD\")\n",
    "        table_word = document.add_table(rows=1, cols=3)\n",
    "        table_word.style = 'Colorful List Accent 1'\n",
    "        hdr_cells = table_word.rows[0].cells\n",
    "        hdr_cells[0].text = 'Gene'\n",
    "        hdr_cells[1].text = 'DNA Change (cDNA)'\n",
    "        hdr_cells[2].text = 'Clinical Classification'\n",
    "    \n",
    "        hit = False\n",
    "    \n",
    "        count = 0\n",
    "        table = driver.find_element_by_xpath(\"//table[@id='viewlistTable_CustomVL_VOTunique_VOG_BRCA2']\")\n",
    "        for row in table.find_elements_by_xpath(\"//tr[@class='data']\"):\n",
    "            count = 0\n",
    "            for table_data in row.find_elements_by_xpath(\".//td\"):\n",
    "                count = count + 1\n",
    "                if count == 4 and search in table_data.text:\n",
    "                    # Searching on nucleotide level\n",
    "                    if (\"+\" or \"-\") in search:\n",
    "                        row_cells = table_word.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[1].text = table_data.text\n",
    "                        hit = True\n",
    "\n",
    "                    elif (\"+\" or \"-\") not in table_data.text:\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[1].text = table_data.text\n",
    "                        hit = True\n",
    "                    \n",
    "                    elif \"+\" in table_data.text:\n",
    "                    \n",
    "                        hit = table_data.text      \n",
    "                        plus = hit.split(\"+\")\n",
    "                        plus_two = plus[1]\n",
    "                        plus_three = re.split('([a-zA-Z\\s]+)', plus_two)\n",
    "                        plus_four = plus_three[0]\n",
    "                        plus_five = plus_four.split('_')\n",
    "                        plus_six = plus_five[0]\n",
    "\n",
    "                        if not any(i == plus_six for i in pattern_string):\n",
    "\n",
    "                            row_cells = table_word.add_row().cells\n",
    "                            row_cells[0].text = brca_input\n",
    "                            row_cells[1].text = table_data.text\n",
    "                            hit = True\n",
    "\n",
    "                    elif \"-\" in table_data.text:\n",
    "\n",
    "                        minus = hit.split(\"-\")\n",
    "                        minus_two = minus[1]\n",
    "                        minus_three = re.split('([a-zA-Z\\s]+)', plus_minus)\n",
    "                        minus_four = minus_three[0]\n",
    "                        minus_five = minus_four.split('_')\n",
    "                        minus_six = minus_five[0]\n",
    "\n",
    "                        if not any( i == minus_six for i in pattern_string):\n",
    "                            row_cells = table_word.add_row().cells\n",
    "                            row_cells[0].text = brca_input\n",
    "                            row_cells[1].text = table_data.text\n",
    "                            hit = True\n",
    "\n",
    "                elif count == 6 and search in table_data.text:\n",
    "                    # Searching on protein lev\n",
    "                    row_cells[O].text = brca_input\n",
    "                    row_cells[3].text = table_data.text\n",
    "                    hit = True\n",
    "\n",
    "                if count == 9 and hit == True:\n",
    "                    row_cells[2].text = table_data.text\n",
    "                    hit = False\n",
    "\n",
    "        p = document.add_paragraph(\"Results: \")    \n",
    "        site_location = str(driver.current_url)\n",
    "        add_hyperlink(p, 'Link', driver.current_url)\n",
    "        print(driver.current_url)\n",
    "\n",
    "        driver.quit()\n",
    "    document.save(\"test.docx\")\n",
    "\n",
    "except Exception:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 438,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "# Lindor pdf\n",
    "# 4\n",
    "# The script will read the pdf file containing the tables\n",
    "# It will create a folder and paste the tables in excel files, this step is skipped if this folder already exists\n",
    "# The script will iterate over each excel file in the folder\n",
    "\n",
    "\n",
    "tables_pdf = read_pdf(folder_location + pdf_lindor, pages=\"17-28\")\n",
    "tables_output_location = \"/tables excel lindor 2012\"\n",
    "\n",
    "tables_folder = folder_location + tables_output_location\n",
    "\n",
    "if not os.path.exists(tables_folder):\n",
    "    os.makedirs(tables_folder)\n",
    "\n",
    "for i, table_pdf in enumerate(tables_pdf, start=1):\n",
    "    table_pdf.to_excel(os.path.join(tables_folder, f\"table_{i}.xlsx\"), index=False)\n",
    "\n",
    "# The excel files are divided in different groups based on the similarity of their structure (Equal amount of columns, columns that have the same meaning)\n",
    "group_one = ('table_1.xlsx')\n",
    "group_two = (\"2.xlsx\", \"3.xlsx\", \"4.xlsx\", \"5.xlsx\")\n",
    "group_three = (\"6.xlsx\", \"7.xlsx\", \"8.xlsx\", \"9.xlsx\", \"10.xlsx\")\n",
    "group_four = (\"11.xlsx\")\n",
    "group_five = (\"12.xlsx\")\n",
    "\n",
    "document.add_heading(\"Lindor 2012 pdf\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=5)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'HGSV nucleotide'\n",
    "hdr_cells[2].text = 'BIC nucleotide'\n",
    "hdr_cells[3].text = \"HGVS protein\"\n",
    "hdr_cells[4].text = 'Class'\n",
    "\n",
    "tables_dir = os.listdir(tables_folder)\n",
    "\n",
    "#In the first half it will search for BRCA1 genes\n",
    "\n",
    "for table_file in tables_dir:\n",
    "    if brca_input == \"BRCA1\":\n",
    "        if table_file.endswith(group_one):\n",
    "\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "\n",
    "            for row in sheet.iter_rows():\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[3].text = brca_input\n",
    "                        row_cells[4].text = str(cell.offset(column=5).value)\n",
    "                        #protein?\n",
    "\n",
    "\n",
    "        if table_file.endswith(group_two):\n",
    "\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "\n",
    "            for row in sheet.iter_rows(max_col = 3, min_col = 3):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[2].text = cell.value\n",
    "                        row_cells[3].text = str(cell.offset(column=-1).value)\n",
    "                        row_cells[4].text = str(cell.offset(column=5).value)\n",
    "                        #bic\n",
    "                        \n",
    "            for row in sheet.iter_rows(max_col = 4, min_col = 4):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[1].text = cell.value\n",
    "                        row_cells[2].text = str(cell.offset(column=-2).value)\n",
    "                        row_cells[4].text = str(cell.offset(column=4).value)\n",
    "                        #hgvs\n",
    "\n",
    "            for row in sheet.iter_rows(max_col = 2, min_col = 2):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[1].text = str(cell.offset(column=2).value)\n",
    "                        row_cells[2].text = str(cell.offset(column=1).value)\n",
    "                        row_cells[3].text = cell.value\n",
    "                        row_cells[4].text = str(cell.offset(column=6).value)\n",
    "                        #hgvs protein\n",
    "                        \n",
    "            \n",
    "        if table_file.endswith(group_four):\n",
    "\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "            for row in sheet.iter_rows():\n",
    "                for cell in row:\n",
    "                    if brca_input and search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = cell.value\n",
    "                        row_cells[4].text = str(cell.offset(column=2).value)\n",
    "                        #brca1/brca2\n",
    "#Second half: it will search for the BRCA2 gene in the applicable file\n",
    "\n",
    "    if brca_input == \"BRCA2\":\n",
    "        if table_file.endswith(group_three):\n",
    "\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "            for row in sheet.iter_rows(max_col = 3, min_col = 3):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[2].text = cell.value\n",
    "                        row_cells[3].text = str(cell.offset(column=-1).value)\n",
    "                        row_cells[4].text = str(cell.offset(column=8).value)\n",
    "                        #bic\n",
    "\n",
    "            for row in sheet.iter_rows(max_col = 4, min_col = 4):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[1].text = cell.value\n",
    "                        row_cells[3].text = str(cell.offset(column=-2).value)\n",
    "                        row_cells[4].text = str(cell.offset(column=5).value)\n",
    "                        #hgvs\n",
    "\n",
    "            for row in sheet.iter_rows(max_col = 2, min_col = 2):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[1].text = str(cell.offset(column=2).value)\n",
    "                        row_cells[2].text = str(cell.offset(column=1).value)\n",
    "                        row_cells[3].text = cell.value\n",
    "                        row_cells[4].text = str(cell.offset(column=9).value)\n",
    "                        #hgvs protein\n",
    "\n",
    "\n",
    "        if table_file.endswith(group_four):\n",
    "\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "            for row in sheet.iter_rows():\n",
    "                for cell in row:\n",
    "                    if brca_input and search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = cell.value\n",
    "                        row_cells[3].text = str(cell.offset(column=2).value)\n",
    "                        \n",
    "        if table_file.endswith(group_five):\n",
    "\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "            for row in sheet.iter_rows(max_col = 4, min_col = 4):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[1].text = cell.value\n",
    "                        row_cells[3].text = str(cell.offset(column=5).value)\n",
    "                        #hgvs\n",
    "\n",
    "                     \n",
    "            for row in sheet.iter_rows(max_col = 5, min_col = 5):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[2].text = cell.value\n",
    "                        row_cells[3].text = str(cell.offset(column=4).value)\n",
    "                        #bic\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 439,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "C:\\Users\\jensv\\AppData\\Local\\Programs\\Python\\Python39\\lib\\site-packages\\openpyxl\\worksheet\\header_footer.py:48: UserWarning: Cannot parse header or footer so it will be ignored\n  warn(\"\"\"Cannot parse header or footer so it will be ignored\"\"\")\n"
     ]
    }
   ],
   "source": [
    "#5) Reading excel sheet: bicbnl 27  fh 2-10-13_nov2013\n",
    "# Python gives warning it can't parse the header but the script still works\n",
    "# Similar to the other excel scripts except it differentiates between two tabs first.\n",
    "\n",
    "wb = load_workbook(folder_location + bicbnl)\n",
    "\n",
    "document.add_heading(\"LOB\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=5)\n",
    "table.allow_outfit = True\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'c.nom'\n",
    "hdr_cells[2].text = 'Type'\n",
    "hdr_cells[3].text = 'opm LOB'\n",
    "hdr_cells[4].text = 'opmerking'\n",
    "    \n",
    "if brca_input == \"BRCA1\":\n",
    "    # The script will open the right sheet first.\n",
    "    sheet = wb['totaal BRCA1']\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 5, min_col = 5):\n",
    "        # It will search in the 5th column for the nucleotide\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                row_cells = table.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                row_cells[1].text = cell.value\n",
    "                row_cells[2].text = str(cell.offset(column=2).value) # column G (type)\n",
    "                row_cells[3].text = str(cell.offset(column=14).value) # column S (opmerking LOB indeling)\n",
    "                row_cells[4].text = str(cell.offset(column=15).value) # column T (opmerking) \n",
    "\n",
    "elif brca_input == \"BRCA2\":\n",
    "    sheet = wb['totaal BRCA2']\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 5, min_col = 5):\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                row_cells = table.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                row_cells[1].text = cell.value\n",
    "                row_cells[2].text = str(cell.offset(column=2).value)\n",
    "                row_cells[3].text = str(cell.offset(column=14).value)\n",
    "                row_cells[4].text = str(cell.offset(column=15).value)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 440,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "<Element {http://schemas.openxmlformats.org/wordprocessingml/2006/main}hyperlink at 0x248c178db40>"
      ]
     },
     "metadata": {},
     "execution_count": 440
    }
   ],
   "source": [
    "from eutils import Client\n",
    "from Bio import Entrez\n",
    "import requests as rq\n",
    "import pprint\n",
    "#4/5/6 clinvar\n",
    "eclient = Client(api_key=\"84c4d9fa268d5e41511ce024a0cf00537808\")\n",
    "\n",
    "gene_esearch = eclient.esearch(db='clinvar',term='{}[gene] AND {}'.format(brca_input, search))\n",
    "\n",
    "id_list = gene_esearch.ids\n",
    "\n",
    "document.add_heading(\"Clinvar\")\n",
    "table_word = document.add_table(rows=1, cols=4)\n",
    "table_word.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table_word.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'cDNA'\n",
    "hdr_cells[2].text = 'Classification'\n",
    "hdr_cells[3].text = 'Review Status'\n",
    "\n",
    "\n",
    "for i in id_list:\n",
    "    query = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=clinvar&id={}&retmode=json\".format(i)\n",
    "    response = rq.get(query).json()\n",
    "    id_string = str(i)\n",
    "    if search in response[\"result\"][id_string][\"title\"]:\n",
    "        row_cells = table_word.add_row().cells\n",
    "        row_cells[0].text = brca_input\n",
    "        sequence = response[\"result\"][id_string][\"title\"]\n",
    "        sequence_split = sequence.split(\":\")\n",
    "        row_cells[1].text = sequence_split[1]\n",
    "        row_cells[2].text = response[\"result\"][id_string][\"clinical_significance\"][\"description\"]\n",
    "        if response[\"result\"][id_string][\"clinical_significance\"][\"review_status\"] == \"practice guideline\":\n",
    "            row_cells[3].text = \"****\"\n",
    "        elif response[\"result\"][id_string][\"clinical_significance\"][\"review_status\"] == \"reviewed by expert panel\":\n",
    "            row_cells[3].text = \"***\"\n",
    "        elif response[\"result\"][id_string][\"clinical_significance\"][\"review_status\"] == \"criteria provided, multiple submitters, no conflicts\":\n",
    "            row_cells[3].text = \"**\"\n",
    "        elif response[\"result\"][id_string][\"clinical_significance\"][\"review_status\"] == \"criteria provided, conflicting interpretations\" or 'criteria provided, single submitter':\n",
    "            row_cells[3].text = \"*\"\n",
    "        elif response[\"result\"][id_string][\"clinical_significance\"][\"review_status\"] == \"no assertion for the individual variant\" or \"no assertion criteria provided\" or \"no assertion provided\":\n",
    "            row_cells[3].text = \"/\"\n",
    "\n",
    "current_url = \"https://www.ncbi.nlm.nih.gov/clinvar/?term={}%5Bgene%5D+{}\".format(brca_input, search)\n",
    "p = document.add_paragraph(\"Results: \")    \n",
    "site_location = current_url\n",
    "add_hyperlink(p, 'Link', current_url)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 441,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "'\\nimport re\\ntest = \\'c.301+7G>A\\'\\ntest_split = re.split(\"[+, ., -, >, d-zD-Z]\", test)\\nprint(test_split)\\n'"
      ]
     },
     "metadata": {},
     "execution_count": 441
    }
   ],
   "source": [
    "\"\"\"\n",
    "import re\n",
    "test = 'c.301+7G>A'\n",
    "test_split = re.split(\"[+, ., -, >, d-zD-Z]\", test)\n",
    "print(test_split)\n",
    "\"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 442,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Enigma rules manual made table\n",
    "# 4?\n",
    "enigma_table_brca1 = \"/ENIGMA RULES BRCA1.xlsx\"\n",
    "enigma_table_brca2 = \"/ENIGMA RULES BRCA2.xlsx\"\n",
    "\n",
    "\n",
    "search_split = re.split(\"[\\-!?:.a-zA-Z]+\", search)\n",
    "search_split = search_split[1]\n",
    "int_search = int(search_split)\n",
    "\n",
    "'''\n",
    "if (\".\") in search:\n",
    "    search_split = re.split(\"[.]\", search)\n",
    "    search_split = search_split[1]\n",
    "\n",
    "if (\"+\" or \"-\" ) in search:\n",
    "    search_split = re.split(\"[+, ., -, >, d-zD-Z]\", search_split)\n",
    "    int_search = int(search_split[0])\n",
    "elif (\">\" or \"del\") in search:\n",
    "    search_split = re.split(\"[+, ., -, >, d-zD-Z]\", search_split)\n",
    "    int_search = int(search_split[0])\n",
    "    print(search_split[1])\n",
    "if (\"del\") in search:\n",
    "    search_split = re.split(\"[d-zD-Z]\", search_split)\n",
    "else:\n",
    "    int_search = int(search)\n",
    "'''\n",
    "if brca_input == \"BRCA1\":\n",
    "    wb = load_workbook(folder_location + enigma_table_brca1)\n",
    "\n",
    "    sheet = wb.active\n",
    "\n",
    "    document.add_heading(\"Table 3: Catalogue of BRCA1 conserved domains/motifs and currently known clinically important amino acid residues, and relevance for classification of BRCA1 in-frame and terminal exon sequence variants.\")\n",
    "\n",
    "    table = document.add_table(rows=1, cols=5)\n",
    "\n",
    "    table.style = 'Colorful List Accent 1'\n",
    "    hdr_cells = table.rows[0].cells\n",
    "    hdr_cells[0].text = 'Domain/motif'\n",
    "    hdr_cells[1].text = 'AA start'\n",
    "    hdr_cells[2].text = 'AA end'\n",
    "    hdr_cells[3].text = 'AA alterations'\n",
    "    hdr_cells[3].text = 'Classification'\n",
    "\n",
    "    # searching for between AA start and AA end\n",
    "    for row in sheet.iter_rows(min_row=2, min_col=2, max_col=2):\n",
    "        for cell in row:\n",
    "            if int_search >= int(cell.value) and int_search <= int(cell.offset(column=1).value):\n",
    "                row_cells = table.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                AA_start = str(cell.value)\n",
    "                AA_end = str(cell.offset(column=1).value)\n",
    "                classification = str(cell.offset(column=3).value)\n",
    "\n",
    "            #  print(\"{} is found between {} and {} and has classification {}\".format(search, AA_start, AA_end, classification))\n",
    "            \n",
    "    ###AA alteration with demonstrated clinical importance\n",
    "    for row in sheet.iter_rows(min_col=4, max_col=4):\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                row_cells = table.add_row().cells\n",
    "                AA_alteration = cell.value ####grep?\n",
    "                classification = str(cell.offset(column=1).value)\n",
    "\n",
    "if brca_input == \"BRCA2\":  \n",
    "\n",
    "    wb = load_workbook(folder_location + enigma_table_brca2)\n",
    "\n",
    "    sheet = wb.active\n",
    "\n",
    "    document.add_heading(\"Table 3: Catalogue of BRCA1 conserved domains/motifs and currently known clinically important amino acid residues, and relevance for classification of BRCA1 in-frame and terminal exon sequence variants.\")\n",
    "    #row_cells[2].text = str(cell.offset(column=2).value)\n",
    "\n",
    "    ###BRCA2\n",
    "\n",
    "    table = document.add_table(rows=1, cols=5)\n",
    "\n",
    "    table.style = 'Colorful List Accent 1'\n",
    "    hdr_cells = table.rows[0].cells\n",
    "    hdr_cells[0].text = 'Domain/motif'\n",
    "    hdr_cells[1].text = 'AA start'\n",
    "    hdr_cells[2].text = 'AA end'\n",
    "    hdr_cells[3].text = 'AA alterations'\n",
    "    hdr_cells[3].text = 'Classification'\n",
    "\n",
    "    # searching for between AA start and AA end\n",
    "    for row in sheet.iter_rows(min_row=2, min_col=2, max_col=2):\n",
    "        for cell in row:\n",
    "            if int_search >= int(cell.value) and int_search <= int(cell.offset(column=1).value):\n",
    "                row_cells = table.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                AA_start = str(cell.value)\n",
    "                AA_end = str(cell.offset(column=1).value)\n",
    "                classification = str(cell.offset(column=3).value)\n",
    "\n",
    "                #print(\"{} is found between {} and {} and has classification {}\".format(search, AA_start, AA_end, classification))\n",
    "            \n",
    "    ###AA alteration with demonstrated clinical importance\n",
    "    for row in sheet.iter_rows(min_col=4, max_col=4):\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                row_cells = table.add_row().cells\n",
    "                AA_alteration = cell.value ####grep?\n",
    "                classification = str(cell.offset(column=1).value)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 443,
   "metadata": {},
   "outputs": [],
   "source": [
    "#2) Reading excel sheet : allenigmavariants_BICsubmission_2013-07-01\n",
    "#7A\n",
    "\n",
    "# The first column in this file is BRCA1/BRCA2, the second column is hgvsnucleotide and the fourth column is bic nucleotide\n",
    "# The script will iterate over each row, if the brca_input matches with the gene in the 1st column AND the search is found in the 2nd OR 4th column it will give the class in column 7 as output\n",
    "\n",
    "#cell.offset(column=2).value gives an error as output if the cell is empty but not when it is converted to string first.\n",
    "\n",
    "# Zoeken op protein change\n",
    "wb = load_workbook(folder_location + variants_excel)\n",
    "\n",
    "sheet = wb.active\n",
    "\n",
    "document.add_heading(\"ENIGMA class 1 lijst met SNVs (freq >1>%% in 1000g)\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=4)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'HGSV nucleotide'\n",
    "hdr_cells[2].text = 'HGSV protein'\n",
    "hdr_cells[3].text = 'Class'\n",
    "\n",
    "\n",
    "for row in sheet.iter_rows():\n",
    "    for cell in row:\n",
    "        # if BRCA gene is in 1st column AND nucleotide search is in 2ndcolumn (hgvs nucleotide)\n",
    "        if brca_input in str(cell.value) and search in str(cell.offset(column=1).value) or brca_input in str(cell.value) and search in str(cell.offset(column=2).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value #gene\n",
    "            row_cells[1].text = cell.offset(column=1).value #hgvs nucleotide\n",
    "            row_cells[2].text = str(cell.offset(column=2).value) #hgvs protein\n",
    "            row_cells[3].text = str(cell.offset(column=6).value) #class\n",
    "        '''\n",
    "        #else if BRCA gene is in 1st column AND nucleotide search is in 4th column (bic nucleotide)  \n",
    "        elif brca_input in str(cell.value) and search in str(cell.offset(column=3).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value #gene\n",
    "            row_cells[2].text = cell.offset(column=3).value #bic nucleotide\n",
    "            row_cells[3].text = cell.offset(column=6).value #class\n",
    "        '''\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 444,
   "metadata": {},
   "outputs": [],
   "source": [
    "#3)Reading excel sheet: Published Multifactoral data enigma\n",
    "#7B samen met 7A\n",
    "# This is likewise to searching in the allenigmavariants_BICsubmission_2013-07-01 excel file\n",
    "\n",
    "\n",
    "wb = load_workbook(folder_location + enigma)\n",
    "\n",
    "sheet = wb.active\n",
    "\n",
    "document.add_heading(\"ENIGMA MLM lijst ( variants classified based on multifactorial likelihood model)\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=4)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'HGSV nucleotide'\n",
    "hdr_cells[2].text = 'HGVS protein'\n",
    "hdr_cells[3].text = 'Class'\n",
    "\n",
    "\n",
    "for row in sheet.iter_rows():\n",
    "    for cell in row:\n",
    "        if brca_input in str(cell.value) and search in str(cell.offset(column=1).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value\n",
    "            row_cells[1].text = cell.offset(column=1).value\n",
    "            row_cells[2].text = str(cell.offset(column=2).value)\n",
    "            row_cells[3].text = cell.offset(column=6).value\n",
    "            Result = False\n",
    "        if brca_input in str(cell.value) and search in str(cell.offset(column=2).value):\n",
    "            row_cells[0].text = cell.value\n",
    "            row_cells[1].text = cell.offset(column=1).value\n",
    "            row_cells[2].text = str(cell.offset(column=2).value)\n",
    "            row_cells[3].text = cell.offset(column=6).value\n",
    "\n",
    "            '''\n",
    "        elif brca_input in str(cell.value) and search in str(cell.offset(column=3).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value\n",
    "            row_cells[2].text = cell.offset(column=3).value\n",
    "            row_cells[3].text = cell.offset(column=6).value\n",
    "            Result = False\n",
    "            '''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 445,
   "metadata": {},
   "outputs": [],
   "source": [
    "#1) Reading excel sheet: 2020 functional categorization of BRCA1 VUS - CCR\n",
    "# Zoeken op protein\n",
    "# 8\n",
    "# Bij group B SNVS smaen met enigma en function scores\n",
    "\n",
    "### Add heading to word document\n",
    "document.add_heading(\"BRCA1 SNV lijst met functionele data (PMID: 32546644)\", 1)\n",
    "\n",
    "### Create table in word\n",
    "table = document.add_table(rows=1, cols=4)\n",
    "\n",
    "### Loading a style template and creating the header cells\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'DNA variant'\n",
    "hdr_cells[1].text = 'Cisplatin Assay'\n",
    "hdr_cells[2].text = 'Olaparib Assay'\n",
    "hdr_cells[3].text = 'DR-GFP assay'\n",
    "\n",
    "\n",
    "#If the input is BRCA1 the script will load the excel file. It will iterate over each row of the first column (max_col = 1) of the active sheet.\n",
    "#If the search input is found in the first column the script will add the values of the 3rd, 4th and 5th column\n",
    "#Example cell.offset(column=2).value means that the script will take the value of the 3rd column (1st column + 2 to the right = 3)\n",
    "if brca_input == \"BRCA1\":\n",
    "\n",
    "    wb = load_workbook(folder_location + functional_categorization)\n",
    "\n",
    "    sheet = wb.active\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 2, min_col = 2):\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                    row_cells = table.add_row().cells\n",
    "                    row_cells[0].text = cell.value\n",
    "                    row_cells[1].text = cell.offset(column=1).value\n",
    "                    row_cells[2].text = cell.offset(column=2).value\n",
    "                    row_cells[3].text = cell.offset(column=3).value\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 446,
   "metadata": {},
   "outputs": [],
   "source": [
    "########Franklin database\n",
    "#9\n",
    "try:\n",
    "\n",
    "    option = webdriver.ChromeOptions()\n",
    "    option.add_argument('headless')\n",
    "    PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "    driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "    document.add_heading(\"Franklin\")\n",
    "    table_word = document.add_table(rows=1, cols=2)\n",
    "    table_word.style = 'Colorful List Accent 1'\n",
    "    hdr_cells = table_word.rows[0].cells\n",
    "    hdr_cells[0].text = 'Gene'\n",
    "    hdr_cells[1].text = 'Classification'\n",
    "    row_cells = table_word.add_row().cells\n",
    "\n",
    "    driver.get(\"https://franklin.genoox.com/clinical-db/home\")\n",
    "\n",
    "    time.sleep(10)\n",
    "\n",
    "    #a pop up appeared on this page which can be removed by pressing escape\n",
    "    ActionChains(driver).send_keys(Keys.ESCAPE).perform()\n",
    "\n",
    "    #The script will search for the searchbar and the button(somatic)\n",
    "    searchbar = driver.find_element_by_xpath(\"/html/body/app-root/div/gnx-home-page/div/gnx-search/div      [2]/input\")\n",
    "\n",
    "    button = driver.find_element_by_xpath(\"/html/body/app-root/div/gnx-home-page/div/gnx-search/div[2]/     gnx-variant-type-toggle/button[2]\")\n",
    "\n",
    "    button.click()\n",
    "\n",
    "    searchbar.send_keys(search)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "\n",
    "    #time.sleep(10)\n",
    "    try:\n",
    "        element = WebDriverWait(driver, 10).until(EC.presence_of_element_located((By.ID, \"mat-tab-label-2-3\")))\n",
    "\n",
    "    except TimeoutException:\n",
    "        print(\"Timed out waiting for page to load\")\n",
    "\n",
    "    driver.find_element_by_xpath(\"//*[@id='mat-tab-label-2-3']\").click()\n",
    "\n",
    "    time.sleep(3)\n",
    "\n",
    "    classification = driver.find_element_by_xpath(\"//*[@class='indicator-text ng-star-inserted']\")\n",
    "\n",
    "    row_cells[0].text = brca_input\n",
    "    row_cells[1].text = classification.text\n",
    "    franklin_class = classification.text\n",
    "\n",
    "\n",
    "    p = document.add_paragraph(\"Results: \")    \n",
    "    site_location = str(driver.current_url)\n",
    "    add_hyperlink(p, 'Link', driver.current_url)\n",
    "\n",
    "    driver.quit()\n",
    "\n",
    "except Exception:\n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 447,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "https://varsome.com/variant/hg19/BRCA2%20c.8954-5A%3EG?annotation-mode=germline\n"
     ]
    }
   ],
   "source": [
    "#Varsome login required if too many searches are done\n",
    "#10\n",
    "try:\n",
    "\n",
    "    option = webdriver.ChromeOptions()\n",
    "    option.add_argument('headless')\n",
    "    PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "    driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "    driver.get(\"https://varsome.com/\")\n",
    "\n",
    "    document.add_heading(\"Varsome\")\n",
    "    table_word = document.add_table(rows=1, cols=2)\n",
    "    table_word.style = 'Colorful List Accent 1'\n",
    "    hdr_cells = table_word.rows[0].cells\n",
    "    hdr_cells[0].text = \"Gene\"\n",
    "    hdr_cells[1].text = 'Classification'\n",
    "    row_cells = table_word.add_row().cells\n",
    "\n",
    "    time.sleep(1)\n",
    "\n",
    "    searchbar = driver.find_element_by_xpath(\"//*[@id='search']\")\n",
    "    searchbar.send_keys(brca_input + \" \" + search)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "\n",
    "    time.sleep(1)\n",
    "\n",
    "    #Popup asking to accept cookies \n",
    "    cookie = driver.find_element_by_xpath(\"//*[@id='onetrust-accept-btn-handler']\")\n",
    "    cookie.click()\n",
    "\n",
    "    button = driver.find_element_by_xpath(\"//*[@id='proceedBtn']/h4\")\n",
    "    button.click()\n",
    "\n",
    "    time.sleep(10)\n",
    "    try:\n",
    "        element = WebDriverWait(driver, 10).until(EC.presence_of_element_located((By.ID, \"acmg_annotation\")))\n",
    "\n",
    "    except TimeoutException:\n",
    "        print(\"Timed out waiting for page to load\")\n",
    "    \n",
    "\n",
    "    classification = driver.find_element_by_xpath(\"//*[@id='acmg_annotation']/div/div[1]/div/div[2]/div/    div/div[1]/div/div/span/div[2]/div/div\")\n",
    "\n",
    "    row_cells = table_word.add_row().cells\n",
    "    row_cells[0].text = brca_input + \" \" + search\n",
    "    row_cells[1].text = classification.text\n",
    "\n",
    "    p = document.add_paragraph(\"Results: \")    \n",
    "    site_location = str(driver.current_url)\n",
    "    add_hyperlink(p, 'Link', driver.current_url)\n",
    "    print(driver.current_url)\n",
    "\n",
    "    driver.quit()\n",
    "\n",
    "except Exception:\n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 448,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "WARNING:urllib3.connectionpool:Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPConnection object at 0x00000248AAA4FCA0>: Failed to establish a new connection: [WinError 10061] Kan geen verbinding maken omdat de doelcomputer de verbinding actief heeft geweigerd')': /session/757f173d08a89f6ed35478bd78332f1f/url\n",
      "WARNING:urllib3.connectionpool:Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPConnection object at 0x00000248A39C3B20>: Failed to establish a new connection: [WinError 10061] Kan geen verbinding maken omdat de doelcomputer de verbinding actief heeft geweigerd')': /session/757f173d08a89f6ed35478bd78332f1f/url\n",
      "WARNING:urllib3.connectionpool:Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPConnection object at 0x00000248BDFDE4C0>: Failed to establish a new connection: [WinError 10061] Kan geen verbinding maken omdat de doelcomputer de verbinding actief heeft geweigerd')': /session/757f173d08a89f6ed35478bd78332f1f/url\n"
     ]
    }
   ],
   "source": [
    "#Database of Functional Classifications of BRCA1 Variants based on Saturation Genome Editing\n",
    "#\n",
    "try:\n",
    "    \n",
    "    if brca_input == \"BRCA1\":\n",
    "\n",
    "        option = webdriver.ChromeOptions()\n",
    "        option.add_argument('headless')\n",
    "        PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "        driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "        document.add_heading(\"Database of Functional Classifications of BRCA1 Variants based on Saturation Genome Editing\")\n",
    "        table_word = document.add_table(rows=1, cols=3)\n",
    "        table_word.style = 'Colorful List Accent 1'\n",
    "        hdr_cells = table_word.rows[0].cells\n",
    "        hdr_cells[0].text = 'Gene'\n",
    "        hdr_cells[1].text = 'Transcript Variant'\n",
    "        hdr_cells[2].text = 'Functional CLass'\n",
    "\n",
    "\n",
    "        driver.get(\"https://sge.gs.washington.edu/BRCA1/#tab-9159-2\")\n",
    "\n",
    "        time.sleep(1)\n",
    "        #The script will look up and click on the right tab\n",
    "        tab = driver.find_element_by_xpath(\"//*[@id='main_page']/div/div/nav/div/ul/li[2]/a\")\n",
    "        tab.click()\n",
    "        '''\n",
    "        tab = driver.find_element_by_xpath(\"//ul[@class='nav navbar-nav']\")\n",
    "        tab.find_element_by_xpath(\".//a[@href='#tab-9159-2']\").click()\n",
    "        '''\n",
    "        #time.sleep(5)\n",
    "        try:\n",
    "            element = WebDriverWait(driver, 10).until(EC.presence_of_element_located((By.ID, \"DataTables_Table_0_filter\")))\n",
    "\n",
    "        except TimeoutException:\n",
    "            print(\"Timed out waiting for page to load\")\n",
    "\n",
    "        # look up the searchbar and enter the query\n",
    "        searchbar = driver.find_element_by_xpath(\"//*[@id='DataTables_Table_0_filter']/label/input\")\n",
    "        searchbar.send_keys(search)\n",
    "        searchbar.send_keys(Keys.RETURN)\n",
    "    \n",
    "        table = driver.find_element_by_xpath(\"//*[@id='DataTables_Table_0']/tbody\")\n",
    " \n",
    "        count = 0\n",
    "        time.sleep(5)\n",
    "        hit = False\n",
    "\n",
    "        for row in table.find_elements_by_xpath(\"//tr\"):\n",
    "            count = 0\n",
    "            for table_data in row.find_elements_by_xpath((\".//td\")):\n",
    "                count = count + 1\n",
    "                if count == 6 and search in table_data.text:\n",
    "                    row_cells = table_word.add_row().cells\n",
    "                    row_cells[0].text = brca_input\n",
    "                    hit = True\n",
    "                    row_cells[1].text = table_data.text \n",
    "                if count == 10 and hit == True:\n",
    "                    row_cells[2].text = table_data.text\n",
    "                    hit = False\n",
    "\n",
    "    p = document.add_paragraph(\"Results: \")    \n",
    "    site_location = str(driver.current_url)\n",
    "    add_hyperlink(p, 'Link', driver.current_url)\n",
    "    print(driver.current_url)\n",
    "\n",
    "    driver.quit()\n",
    "except Exception:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 449,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "https://www.oncokb.org/gene/BRCA2/c.8954-5A%3EG\n"
     ]
    }
   ],
   "source": [
    "#Oncokb.org\n",
    "\n",
    "try:\n",
    "\n",
    "    option = webdriver.ChromeOptions()\n",
    "    option.add_argument('headless')\n",
    "    PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "    driver = webdriver.Chrome(PATH, options = option)\n",
    "\n",
    "    driver.get(\"https://www.oncokb.org/\")\n",
    "\n",
    "    document.add_heading(\"OncoKB\")\n",
    "    table_word = document.add_table(rows=1, cols=2)\n",
    "    table_word.style = 'Colorful List Accent 1'\n",
    "    hdr_cells = table_word.rows[0].cells\n",
    "    hdr_cells[0].text = 'Search'\n",
    "    hdr_cells[1].text = 'Results'\n",
    "\n",
    "    #time.sleep(5)\n",
    "\n",
    "    searchbar = driver.find_element_by_xpath(\"/html/body/div[1]/div[2]/div[3]/div/div/div[3]/div/div/       div/div[1]/div[2]/div/input\")\n",
    "    searchbar.send_keys(brca_input + \" \" + search)\n",
    "    time.sleep(2)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "    time.sleep(5)\n",
    "    result = driver.find_element_by_xpath(\"//*[@id='root']/div[2]/div[3]/div/div/div/div[2]/div/div[2]\")\n",
    "\n",
    "    row_cells = table_word.add_row().cells\n",
    "    row_cells[0].text = brca_input + search\n",
    "    row_cells[1].text = result.text\n",
    "\n",
    "    p = document.add_paragraph(\"Results: \")    \n",
    "    site_location = str(driver.current_url)\n",
    "    add_hyperlink(p, 'Link', driver.current_url)\n",
    "    print(driver.current_url)\n",
    "\n",
    "    driver.quit()\n",
    "    \n",
    "except Exception:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 450,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "https://pubmed.ncbi.nlm.nih.gov/?term=BRCA2+c.8954-5A>G\n"
     ]
    }
   ],
   "source": [
    "# Pubmed\n",
    "p = document.add_paragraph(\"pubmed: \") \n",
    "pubmed_site = \"https://pubmed.ncbi.nlm.nih.gov/?term=\"  \n",
    "pubmed_full = pubmed_site + brca_input + \"+\" + search\n",
    "add_hyperlink(p, 'Link', pubmed_full)\n",
    "print(pubmed_full)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 451,
   "metadata": {},
   "outputs": [],
   "source": [
    "document.save(\"output.docx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 452,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "'\\npdf_enigma = \"/ENIGMA_Rules_2017-06-29.pdf\"\\nimport tabula\\nimport tabulate\\nfolder_location = \"C:/Users/jensv/Desktop/BRCA_-_prostate_cancer_and_ovarian_cancer_-_PARPi\"\\ntables_enigma = read_pdf(folder_location + pdf_enigma, pages=\\'14-16\\')\\ntables_output_location = \"/tables ENIGMA\"\\n\\nprint(tables_enigma)\\ntables_folder = folder_location + tables_output_location\\n\\nif not os.path.exists(tables_folder):\\n    os.makedirs(tables_folder)\\n\\nfor i, table_pdf in enumerate(tables_enigma, start=1):\\n    table_pdf.to_excel(os.path.join(tables_folder, f\"table_{i}.xlsx\"), index=False)\\n'"
      ]
     },
     "metadata": {},
     "execution_count": 452
    }
   ],
   "source": [
    "'''\r\n",
    "pdf_enigma = \"/ENIGMA_Rules_2017-06-29.pdf\"\r\n",
    "import tabula\r\n",
    "import tabulate\r\n",
    "folder_location = \"C:/Users/jensv/Desktop/BRCA_-_prostate_cancer_and_ovarian_cancer_-_PARPi\"\r\n",
    "tables_enigma = read_pdf(folder_location + pdf_enigma, pages='14-16')\r\n",
    "tables_output_location = \"/tables ENIGMA\"\r\n",
    "\r\n",
    "print(tables_enigma)\r\n",
    "tables_folder = folder_location + tables_output_location\r\n",
    "\r\n",
    "if not os.path.exists(tables_folder):\r\n",
    "    os.makedirs(tables_folder)\r\n",
    "\r\n",
    "for i, table_pdf in enumerate(tables_enigma, start=1):\r\n",
    "    table_pdf.to_excel(os.path.join(tables_folder, f\"table_{i}.xlsx\"), index=False)\r\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 453,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "\"\\n# API requst varsome 1 request per day\\nimport requests\\nimport json\\n\\nresponse = requests.get('https://api.varsome.com/lookup/gene/BRCA1/hg19?add-all-data=1')\\n\\nprint(response.json())\\n\\n\""
      ]
     },
     "metadata": {},
     "execution_count": 453
    }
   ],
   "source": [
    "'''\n",
    "# API requst varsome 1 request per day\n",
    "import requests\n",
    "import json\n",
    "\n",
    "response = requests.get('https://api.varsome.com/lookup/gene/BRCA1/hg19?add-all-data=1')\n",
    "\n",
    "print(response.json())\n",
    "\n",
    "'''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 454,
   "metadata": {},
   "outputs": [
    {
     "output_type": "error",
     "ename": "SyntaxError",
     "evalue": "EOL while scanning string literal (<ipython-input-454-44987550c2d7>, line 21)",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-454-44987550c2d7>\"\u001b[1;36m, line \u001b[1;32m21\u001b[0m\n\u001b[1;33m    ''''\u001b[0m\n\u001b[1;37m        \n^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m EOL while scanning string literal\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "#LOVD database\n",
    "import requests as rq\n",
    "import json\n",
    "import pprint\n",
    "import feedparser\n",
    "import bs4\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "query = \"\"\"\n",
    "https://databases.lovd.nl/shared/api/rest.php/variants/IVD\n",
    "\"\"\"\n",
    "response = rq.get(\"https://databases.lovd.nl/shared/variants/BRCA1/unique1\")\n",
    "\n",
    "responsesoup = bs4.BeautifulSoup(response.text)\n",
    "for i in responsesoup:\n",
    "    if \"5074\" in i:\n",
    "        print(i)\n",
    "''''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "output_type": "error",
     "ename": "NameError",
     "evalue": "name 'BeautifulSoup' is not defined",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-23-af9a62e61fc0>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     15\u001b[0m \u001b[0mresponse\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mrq\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"https://databases.lovd.nl/shared/api/rest.php/variants/BRCA1\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     16\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 17\u001b[1;33m \u001b[0mresponsesoup\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mBeautifulSoup\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mresponse\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'lxml'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m: name 'BeautifulSoup' is not defined"
     ]
    }
   ],
   "source": [
    "'''\n",
    "#LOVD database\n",
    "import requests as rq\n",
    "import json\n",
    "import pprint\n",
    "import feedparser\n",
    "import bs4\n",
    "from xml.etree import cElementTree as ET\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "query = \"\"\"\n",
    "https://databases.lovd.nl/shared/api/rest.php/variants/IVD\n",
    "\"\"\"\n",
    "'''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "output_type": "error",
     "ename": "TypeError",
     "evalue": "'str' object is not callable",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-24-395fd9e5d58a>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     10\u001b[0m \u001b[0mquery\u001b[0m\u001b[1;33m=\u001b[0m \u001b[1;34m\"https://api.genoox.com/v1/variant/search?search_text=VHL:c.293A>G\"\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     11\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 12\u001b[1;33m \u001b[0mr\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mrq\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mquery\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mauth\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'franklin_token'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m~\\AppData\\Roaming\\Python\\Python39\\site-packages\\requests\\api.py\u001b[0m in \u001b[0;36mget\u001b[1;34m(url, params, **kwargs)\u001b[0m\n\u001b[0;32m     74\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     75\u001b[0m     \u001b[0mkwargs\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msetdefault\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'allow_redirects'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;32mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 76\u001b[1;33m     \u001b[1;32mreturn\u001b[0m \u001b[0mrequest\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'get'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0murl\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mparams\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mparams\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     77\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     78\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Roaming\\Python\\Python39\\site-packages\\requests\\api.py\u001b[0m in \u001b[0;36mrequest\u001b[1;34m(method, url, **kwargs)\u001b[0m\n\u001b[0;32m     59\u001b[0m     \u001b[1;31m# cases, and look like a memory leak in others.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     60\u001b[0m     \u001b[1;32mwith\u001b[0m \u001b[0msessions\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mSession\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0msession\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 61\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0msession\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mrequest\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmethod\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mmethod\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0murl\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0murl\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     62\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     63\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Roaming\\Python\\Python39\\site-packages\\requests\\sessions.py\u001b[0m in \u001b[0;36mrequest\u001b[1;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[0;32m    526\u001b[0m             \u001b[0mhooks\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mhooks\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    527\u001b[0m         )\n\u001b[1;32m--> 528\u001b[1;33m         \u001b[0mprep\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mprepare_request\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    529\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    530\u001b[0m         \u001b[0mproxies\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mproxies\u001b[0m \u001b[1;32mor\u001b[0m \u001b[1;33m{\u001b[0m\u001b[1;33m}\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Roaming\\Python\\Python39\\site-packages\\requests\\sessions.py\u001b[0m in \u001b[0;36mprepare_request\u001b[1;34m(self, request)\u001b[0m\n\u001b[0;32m    454\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    455\u001b[0m         \u001b[0mp\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mPreparedRequest\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 456\u001b[1;33m         p.prepare(\n\u001b[0m\u001b[0;32m    457\u001b[0m             \u001b[0mmethod\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mrequest\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmethod\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mupper\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    458\u001b[0m             \u001b[0murl\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mrequest\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0murl\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Roaming\\Python\\Python39\\site-packages\\requests\\models.py\u001b[0m in \u001b[0;36mprepare\u001b[1;34m(self, method, url, headers, files, data, params, auth, cookies, hooks, json)\u001b[0m\n\u001b[0;32m    318\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mprepare_cookies\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcookies\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    319\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mprepare_body\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mfiles\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mjson\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 320\u001b[1;33m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mprepare_auth\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mauth\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0murl\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    321\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    322\u001b[0m         \u001b[1;31m# Note that prepare_auth must be last to enable authentication schemes\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Roaming\\Python\\Python39\\site-packages\\requests\\models.py\u001b[0m in \u001b[0;36mprepare_auth\u001b[1;34m(self, auth, url)\u001b[0m\n\u001b[0;32m    549\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    550\u001b[0m             \u001b[1;31m# Allow auth to make its changes.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 551\u001b[1;33m             \u001b[0mr\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mauth\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    552\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    553\u001b[0m             \u001b[1;31m# Update self to reflect the auth changes.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mTypeError\u001b[0m: 'str' object is not callable"
     ]
    }
   ],
   "source": [
    "'''\n",
    "\n",
    "\n",
    "#token = {{franklin token}}\n",
    "franklin_token = \"franklin token\"\n",
    "\n",
    "query= \"https://api.genoox.com/v1/variant/search?search_text=VHL:c.293A>G\"\n",
    "\n",
    "r = rq.get(query, auth='franklin_token')\n",
    "''''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "import myvariant\n",
    "mv = myvariant.MyVariantInfo()\n",
    "\n",
    "out = mv.query('clinvar.gene.symbol:BRCA1 AND clinvar.rcv.clinical_significance:pathogenic')\n",
    "'''"
   ]
  }
 ]
}