{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  },
  "orig_nbformat": 2,
  "kernelspec": {
   "name": "python395jvsc74a57bd070de0b75b2a3f35b1bb892b463cd56f2ed390cd02c9f7d4168b3d96fea273b5e",
   "display_name": "Python 3.9.4 64-bit"
  },
  "metadata": {
   "interpreter": {
    "hash": "70de0b75b2a3f35b1bb892b463cd56f2ed390cd02c9f7d4168b3d96fea273b5e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "1ok\n"
     ]
    }
   ],
   "source": [
    "### pip is the package installer for python3. The command \"py -m pip install\"\n",
    "### installs the packages where in \"-m pip\" executes pip using the latest python interpreter installed in windows.\n",
    "\n",
    "import os\n",
    "import subprocess\n",
    "import re\n",
    "\n",
    "### packages for reading excel files\n",
    "from openpyxl import Workbook\n",
    "from openpyxl import load_workbook\n",
    "\n",
    "### packages for writing word documents\n",
    "from docx import Document\n",
    "from docx.shared import Cm\n",
    "import docx\n",
    "from docx.enum.dml import MSO_THEME_COLOR_INDEX\n",
    "#py -m pip install --user python-docx\n",
    "\n",
    "### packages for webscraping\n",
    "from selenium import webdriver\n",
    "from selenium.webdriver.common.keys import Keys\n",
    "import time\n",
    "from selenium.webdriver.common.by import By\n",
    "from selenium.webdriver.support.ui import WebDriverWait\n",
    "from selenium.webdriver.support import expected_conditions as EC\n",
    "from selenium.webdriver.common.action_chains import ActionChains\n",
    "\n",
    "#Setting up selenium driver\n",
    "#Install the corresponding version of chromedriver on https://sites.google.com/a/chromium.org/chromedriver/downloads and place the file in a map. Store the path to the file in the PATH variable below.\n",
    "'''\n",
    "option = webdriver.ChromeOptions()\n",
    "option.add_argument('headless')\n",
    "PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "driver = webdriver.Chrome(PATH, options=option)\n",
    "'''\n",
    "# package to read the pdf files\n",
    "from tabula import read_pdf\n",
    "from tabulate import tabulate\n",
    "import tabula\n",
    "\n",
    "# package to read text files\n",
    "import csv\n",
    "\n",
    "# package to split the nucleotide input\n",
    "import re\n",
    "import requests as rq\n",
    "\n",
    "# package to handle json format\n",
    "import json\n",
    "import pprint\n",
    "\n",
    "print(\"1ok\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "2ok\n"
     ]
    }
   ],
   "source": [
    "# Creating a function that allows for the creation of working hyperlinks\n",
    "\n",
    "def add_hyperlink(paragraph, text, url):\n",
    "    # This gets access to the document.xml.rels file and gets a new relation id value\n",
    "    part = paragraph.part\n",
    "    r_id = part.relate_to(url, docx.opc.constants.RELATIONSHIP_TYPE.HYPERLINK, is_external=True)\n",
    "\n",
    "    # Create the w:hyperlink tag and add needed values\n",
    "    hyperlink = docx.oxml.shared.OxmlElement('w:hyperlink')\n",
    "    hyperlink.set(docx.oxml.shared.qn('r:id'), r_id, )\n",
    "\n",
    "    # Create a w:r element and a new w:rPr element\n",
    "    new_run = docx.oxml.shared.OxmlElement('w:r')\n",
    "    rPr = docx.oxml.shared.OxmlElement('w:rPr')\n",
    "\n",
    "    # Join all the xml elements together add add the required text to the w:r element\n",
    "    new_run.append(rPr)\n",
    "    new_run.text = text\n",
    "    hyperlink.append(new_run)\n",
    "\n",
    "    # Create a new Run object and add the hyperlink into it\n",
    "    r = paragraph.add_run ()\n",
    "    r._r.append (hyperlink)\n",
    "\n",
    "    # A workaround for the lack of a hyperlink style (doesn't go purple after using the link)\n",
    "    # Delete this if using a template that has the hyperlink style in it\n",
    "    r.font.color.theme_color = MSO_THEME_COLOR_INDEX.HYPERLINK\n",
    "    r.font.underline = True\n",
    "\n",
    "    return hyperlink\n",
    "print(\"2ok\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "'\\nprint(\"Insert gene BRCA1/BRCA2)\\nbrca_input = input()\\nbrca_input = brca_input.upper()\\nprint(brca_input)\\nprint(\"What sort of sequence? Nucleotide/Protein\")\\nsequence_type = input()\\nsequence_type = sequence_type.upper()\\nprint(\"Insert sequence\")\\nsearch = input()\\n\\nprint(\"3ok\")\\n'"
      ]
     },
     "metadata": {},
     "execution_count": 90
    }
   ],
   "source": [
    "###########Input\n",
    "'''\n",
    "print(\"Insert gene BRCA1/BRCA2)\n",
    "brca_input = input()\n",
    "brca_input = brca_input.upper()\n",
    "print(brca_input)\n",
    "print(\"What sort of sequence? Nucleotide/Protein\")\n",
    "sequence_type = input()\n",
    "sequence_type = sequence_type.upper()\n",
    "print(\"Insert sequence\")\n",
    "search = input()\n",
    "\n",
    "print(\"3ok\")\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "\n",
      "\n",
      "5074\n"
     ]
    }
   ],
   "source": [
    "from tkinter import *\n",
    "\n",
    "#Click function\n",
    "def click():\n",
    "    global brca_input\n",
    "    global sequence_type\n",
    "    global search\n",
    "    brca_input = gene_entry.get()\n",
    "    sequence_type = sequence_entry.get()\n",
    "    search = search_entry.get()\n",
    "    print(search)\n",
    "    #window.quit()\n",
    "\n",
    "#creating a window\n",
    "\n",
    "window = Tk()\n",
    "window.title(\"BRCA1/2 automatic consultation\")\n",
    "\n",
    "#Canvas\n",
    "\n",
    "canvas = tk.Canvas(window, width=600, height=300, bg=\"white\")\n",
    "canvas.grid(columnspan=3, rowspan=4)\n",
    "\n",
    "#Creating a label widget\n",
    "\n",
    "Intro = Label(canvas, text=\"Automatic consultation of diverse databases required for BRCA1/2 gene identification\", bg=\"lightblue\")\n",
    "Intro.grid(column=0,row=0)\n",
    "\n",
    "Post_intro = Label(window, text=\" \", bg='white')\n",
    "Post_intro.grid(column=0,row=1)\n",
    "\n",
    "gene = Label(canvas, text=\"Enter the gene: BRCA1/BRCA2:\", bg=\"lightblue\")\n",
    "gene.grid(column=0, row=2)\n",
    "gene_entry = Entry(canvas, width=20, bg=\"lightblue\")\n",
    "gene_entry.grid(row=2, column=1)\n",
    "#Button (canvas, text=\"SUBMIT\", width=6, command=click) .grid(row=2, column=2)\n",
    "\n",
    "\n",
    "sequence = Label(canvas, text=\"Protein/nucleotide?:\", bg=\"lightblue\")\n",
    "sequence.grid(column=0, row=3)\n",
    "sequence_entry = Entry(canvas, width=20, bg=\"lightblue\")\n",
    "sequence_entry.grid(row=3, column=1)\n",
    "#Button (canvas, text=\"SUBMIT\", width=6, command=click) .grid(row=3, column=2)\n",
    "\n",
    "\n",
    "\n",
    "search = Label(canvas, text=\"Enter sequence:\", bg=\"lightblue\")\n",
    "search.grid(column=0, row=4)\n",
    "search_entry = Entry(canvas, width=20, bg=\"lightblue\")\n",
    "search_entry.grid(row=4, column=1)\n",
    "Button (canvas, text=\"SUBMIT\", width=6, command=click) .grid(row=4, column=2)\n",
    "\n",
    "brca_input = gene_entry.get()\n",
    "sequence_type = sequence_entry.get()\n",
    "search = search_entry.get()\n",
    "print(search)\n",
    "print(sequence_type)\n",
    "\n",
    "\n",
    "window.mainloop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "transformok\n"
     ]
    }
   ],
   "source": [
    "### Transforming the input\n",
    "### Generating a list of numbers and converting them from integers into strings. The script will use this list as a comparison, if the search output gives a nucleotide squence of +5 or -5 it will not be given as output because the list of disallowed numbers ranges from 4 to 9999 in this case, this can be changed below.\n",
    "import re\n",
    "pattern = list(range(4, 9999))\n",
    "pattern_string = [str(number) for number in pattern]\n",
    "\n",
    "### If the input is a number nucleotide sequence, the output should be compared to a range of values\n",
    "### if the input is a specific place it shouldn't be compared to the list\n",
    "if sequence_type == \"PROTEIN\":\n",
    "    search_split = re.search(\"[0-9]+\", search)\n",
    "    search_mod = search_split.group()\n",
    "\n",
    "elif sequence_type == \"NUCLEOTIDE\":\n",
    "    if bool(re.search(\"[a-zA-Z]\", search)) == True:\n",
    "        search_split = re.split(\"[+, ., \\-, >, _, a-zA-Z]\", search)\n",
    "        search_mod = search_split[2]\n",
    "        range_check = False\n",
    "\n",
    "    else:\n",
    "        search_mod = search\n",
    "        range_check = True\n",
    "\n",
    "print(\"transformok\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import os\n",
    "####Setting up the path to the location of the excel files and pdf file\n",
    "#folder_location = os.getcwd()\n",
    "#folder_location = os.getcwd()\n",
    "folder_location = \"C:/Users/jensv/Desktop/BRCA_-_prostate_cancer_and_ovarian_cancer_-_PARPi\"\n",
    "functional_categorization = \"/2020 BRCA1 functional categorization of BRCA1 VUS - CCR.xlsx\"\n",
    "variants_excel = \"/allenigmavariants_BICsubmission_2013-07-01.xlsx\"\n",
    "bicbnl = \"/bicbnl 27  fh 2-10-13_nov2013.xlsx\"\n",
    "enigma = \"/BRCA_Multifac_published data_ENIGMAwebsite 2015-03-27.xlsx\"\n",
    "pdf_lindor = \"/Lindor 2012.pdf\"\n",
    "cmg_brca1 = \"/BRCA1 CMG 05 2021.txt\"\n",
    "cmg_brca2 = \"/BRCA2 CMG 05 2021.txt\"\n",
    "brca1_exchange = \"/brca1_exchange\"\n",
    "brca2_exchange = \"/brca2_exchange\"\n",
    "new_path = \"C:/Users/jensv/Stage_UZA\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Word document\n",
    "\n",
    "### Create word document\n",
    "document = Document()\n",
    "sections = document.sections\n",
    "for section in sections:\n",
    "    section.top_margin = Cm(1)\n",
    "    section.bottom_margin = Cm(1)\n",
    "    section.left_margin = Cm(1)\n",
    "    section.right_margin = Cm(1)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "uza\n"
     ]
    }
   ],
   "source": [
    "### UZA CMG\n",
    "### priority: 1\n",
    "### The script will open the text file depending on if the gene input is BRCA1/2\n",
    "document.add_heading(\"UZA CMG / BRCA Exchange / LOVD\")\n",
    "table_word = document.add_table(rows=1, cols=5)\n",
    "\n",
    "table_word.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table_word.rows[0].cells\n",
    "hdr_cells[0].text = 'Source'\n",
    "hdr_cells[1].text = 'Gene'\n",
    "hdr_cells[2].text = 'cDNA'\n",
    "hdr_cells[3].text = 'Protein'\n",
    "hdr_cells[4].text = 'Mut effect/Classification'\n",
    "\n",
    "if brca_input == \"BRCA1\":\n",
    "\n",
    "    full_path = folder_location + cmg_brca1\n",
    "    with open(full_path) as csv_file:\n",
    "        # The script will open the text file using csv.reader but using the tab (\"\\t\") delimiter so it will iterate over each line of text separated by tabs.\n",
    "        csv_reader = csv.reader(csv_file, delimiter=\"\\t\")\n",
    "        for row in csv_reader:\n",
    "            count = 0\n",
    "            hit = False\n",
    "            for cell in row:\n",
    "                count = count + 1\n",
    "                # The coding sequences are on the 8th tab\n",
    "                if count == 8 and search in cell:\n",
    "                    row_cells = table_word.add_row().cells\n",
    "                    row_cells[0].text = \"UZA CMG\"\n",
    "                    row_cells[1].text = brca_input\n",
    "                    row_cells[2].text = cell\n",
    "                    hit = True\n",
    "                    count = count + 1\n",
    "                    \n",
    "                    \n",
    "                # if it doesn't get a hit for the coding sequences it will check the protein sequence on the 9th tab\n",
    "                elif count == 9 and search in cell:\n",
    "                    row_cells = table_word.add_row().cells\n",
    "                    row_cells[0].text = \"UZA CMG\"\n",
    "                    row_cells[1].text = brca_input\n",
    "                    row_cells[3].text = cell\n",
    "                    hit = True\n",
    "                    count = count + 1\n",
    "                # The mut effect is on the 11th tab. If there is a hit for the coding sequences or the protein it will give the mut effect as output.\n",
    "                if count == 11 and hit == True:\n",
    "                    row_cells[4].text = cell\n",
    "                    hit = False\n",
    "                    count = count + 1\n",
    "                    \n",
    "             \n",
    "        \n",
    "# The files for BRCA1 and BRCA2 have the same structure so the loop can be repeated.\n",
    "elif brca_input == \"BRCA2\":\n",
    "\n",
    "    full_path = folder_location + cmg_brca2\n",
    "    with open(full_path) as csv_file:\n",
    "        csv_reader = csv.reader(csv_file, delimiter=\"\\t\")\n",
    "        for row in csv_reader:\n",
    "            count = 0\n",
    "            hit = False\n",
    "            for cell in row:\n",
    "                count = count + 1\n",
    "                if count == 8 and search in cell:\n",
    "                    row_cells = table_word.add_row().cells\n",
    "                    row_cells[0].text = \"UZA CMG\"\n",
    "                    row_cells[1].text = brca_input\n",
    "                    row_cells[2].text = cell\n",
    "                    hit = True\n",
    "                    count = count + 1\n",
    "                elif count == 9 and search in cell:\n",
    "                    row_cells = table_word.add_row().cells\n",
    "                    row_cells[0].text = \"UZA CMG\"\n",
    "                    row_cells[1].text = brca_input\n",
    "                    row_cells[3].text = cell\n",
    "                    hit = True\n",
    "                    count = count + 1\n",
    "                if count == 11 and hit == True:\n",
    "                    row_cells[4].text = cell\n",
    "                    hit = False\n",
    "                    count = count + 1\n",
    "                    \n",
    "\n",
    "print(\"uza\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Downloading the brca exchange databases locally\n",
    "download_database = \"NO\"\n",
    "if download_database == \"YES\":\n",
    "    query_brca1 = \"\"\"\n",
    "https://brcaexchange.org/backend/data/\n",
    "?format=json\n",
    "&search_term=BRCA1\n",
    "&include=Variant_in_ENIGMA\n",
    "&include=Variant_in_ClinVar\n",
    "&include=Variant_in_1000_Genomes\n",
    "&include=Variant_in_ExAC\n",
    "&include=Variant_in_LOVD\n",
    "&include=Variant_in_BIC\n",
    "&include=Variant_in_ESP\n",
    "&include=Variant_in_exLOVD\n",
    "&include=Variant_in_Findlay_BRCA1_Ring_Function_Scores\n",
    "&include=Variant_in_GnomAD\n",
    "&include=Variant_in_GnomADv3\n",
    "\"\"\"\n",
    "    response = rq.get(query_brca1.replace('\\n', '')).json()['data']\n",
    "    with open('brca1_exchange', 'w') as brca1_data:\n",
    "        json.dump(response, brca1_data)\n",
    "\n",
    "    query_brca2 = \"\"\"\n",
    "https://brcaexchange.org/backend/data/\n",
    "?format=json\n",
    "&search_term=BRCA1\n",
    "&include=Variant_in_ENIGMA\n",
    "&include=Variant_in_ClinVar\n",
    "&include=Variant_in_1000_Genomes\n",
    "&include=Variant_in_ExAC\n",
    "&include=Variant_in_LOVD\n",
    "&include=Variant_in_BIC\n",
    "&include=Variant_in_ESP\n",
    "&include=Variant_in_exLOVD\n",
    "&include=Variant_in_Findlay_BRCA1_Ring_Function_Scores\n",
    "&include=Variant_in_GnomAD\n",
    "&include=Variant_in_GnomADv3\n",
    "\"\"\"\n",
    "    response = rq.get(query_brca2.replace('\\n', '')).json()['data']\n",
    "    with open('brca2_exchange', 'w') as brca2_data:\n",
    "        json.dump(response, brca2_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "BRCA2\n"
     ]
    }
   ],
   "source": [
    "# Searching BRCA Exchange database\"\n",
    "# priority: 2\n",
    "\n",
    "option = webdriver.ChromeOptions()\n",
    "option.add_argument('headless')\n",
    "PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "\n",
    "try:\n",
    "\n",
    "    driver.get(\"https://brcaexchange.org/\")\n",
    "\n",
    "    # Looking for the searchbar and enter the search\n",
    "    searchbar = driver.find_element_by_xpath(\"//input[@type='text']\")\n",
    "    searchbar.send_keys(brca_input + \" \" +search)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "    #wait for the page to load\n",
    "    time.sleep(4)\n",
    "    \n",
    "    page = driver.current_url + \"&pageLength=100\"\n",
    "    driver.get(page)\n",
    "    time.sleep(4)\n",
    "    table = driver.find_element_by_xpath(\"//*[@id='data-table-container']/div/table/tbody\")\n",
    "    \n",
    "    tab_count = 1\n",
    "    #iterate over each cell\n",
    "    while tab_count < 6:\n",
    "        for row in table.find_elements_by_xpath(\"//tr\"):\n",
    "            count = 0\n",
    "            for table_data in row.find_elements_by_xpath('.//td'):\n",
    "                count = count + 1\n",
    "                if count == 2 and search in table_data.text:\n",
    "                    if range_check == True:\n",
    "                        cDNA = table_data.text\n",
    "                        cDNA_split = re.split(\"[+, ., \\-, >, _, a-zA-Z]\", cDNA)\n",
    "                        cDNA_split_plusmin = cDNA_split[3]\n",
    "\n",
    "                        if not any(i == cDNA_split_plusmin for i in pattern_string):\n",
    "                            row_cells = table_word.add_row().cells\n",
    "                            row_cells[0].text = \"BRCA Exchange\"\n",
    "                            row_cells[1].text = brca_input\n",
    "                            row_cells[2].text = table_data.text\n",
    "                            hit = True\n",
    "\n",
    "                    elif range_check == False:\n",
    "                        row_cells = table_word.add_row().cells\n",
    "                        row_cells[0].text = \"BRCA Exchange\"\n",
    "                        row_cells[1].text = brca_input\n",
    "                        row_cells[2].text = table_data.text\n",
    "                        hit = True\n",
    "\n",
    "                if count == 3 and hit == True:\n",
    "                    row_cells[3].text = table_data.text\n",
    "                    hit = True\n",
    "\n",
    "                if count == 3 and search in table_data.text:\n",
    "                    row_cells = table_word.add_row().cells\n",
    "                    row_cells[0].text = \"BRCA Exchange\"\n",
    "                    row_cells[1].text = brca_input\n",
    "                    row_cells[3].text = table_data.text\n",
    "                    hit = True\n",
    "                if count == 6 and hit == True:\n",
    "                    row_cells[4].text = table_data.text\n",
    "                    hit = False\n",
    "        \n",
    "\n",
    "        next_button = driver.find_element_by_xpath(\"//*[@id='variant-search-row']/div[2]/ul/li[8]/a\")\n",
    "        enabled = next_button.is_enabled()\n",
    "        if enabled == True:\n",
    "            next_button.click()\n",
    "            time.sleep(10)\n",
    "            tab_count = tab_count + 1\n",
    "            enabled = next_button.is_enabled()\n",
    "        else:\n",
    "            p = document.add_paragraph(\"Results: \")    \n",
    "            site_location = str(driver.current_url)\n",
    "            add_hyperlink(p, 'Link', driver.current_url)\n",
    "            print(driver.current_url)\n",
    "            driver.quit()\n",
    "            break\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#create and save a hyperlink to word\n",
    "        p = document.add_paragraph(\"Results: \")    \n",
    "        site_location = str(driver.current_url)\n",
    "        add_hyperlink(p, 'Link', driver.current_url)\n",
    "        print(driver.current_url)\n",
    "        print(\"test\")\n",
    "\n",
    "        driver.quit()\n",
    "\n",
    "except Exception:\n",
    "        pass\n",
    "print(\"BRCA2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "output_type": "error",
     "ename": "SessionNotCreatedException",
     "evalue": "Message: session not created\nfrom tab crashed\n  (Session info: headless chrome=90.0.4430.212)\n",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mSessionNotCreatedException\u001b[0m                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-98-512e42264382>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     13\u001b[0m \u001b[0moption\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0madd_argument\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'headless'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     14\u001b[0m \u001b[0mPATH\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;34m\"C:\\Program Files (x86)\\chromedriver.exe\"\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 15\u001b[1;33m \u001b[0mdriver\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mwebdriver\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mChrome\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mPATH\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0moptions\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0moption\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     16\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     17\u001b[0m \u001b[1;31m#Try is used so the script doesn't fail completely if something goes wrong with one website (it doesn't load, the site got updated, it can't find the search, ...)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Roaming\\Python\\Python39\\site-packages\\selenium\\webdriver\\chrome\\webdriver.py\u001b[0m in \u001b[0;36m__init__\u001b[1;34m(self, executable_path, port, options, service_args, desired_capabilities, service_log_path, chrome_options, keep_alive)\u001b[0m\n\u001b[0;32m     74\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     75\u001b[0m         \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 76\u001b[1;33m             RemoteWebDriver.__init__(\n\u001b[0m\u001b[0;32m     77\u001b[0m                 \u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     78\u001b[0m                 command_executor=ChromeRemoteConnection(\n",
      "\u001b[1;32m~\\AppData\\Roaming\\Python\\Python39\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\u001b[0m in \u001b[0;36m__init__\u001b[1;34m(self, command_executor, desired_capabilities, browser_profile, proxy, keep_alive, file_detector, options)\u001b[0m\n\u001b[0;32m    155\u001b[0m             warnings.warn(\"Please use FirefoxOptions to set browser profile\",\n\u001b[0;32m    156\u001b[0m                           DeprecationWarning, stacklevel=2)\n\u001b[1;32m--> 157\u001b[1;33m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mstart_session\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcapabilities\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mbrowser_profile\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    158\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_switch_to\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mSwitchTo\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    159\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_mobile\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mMobile\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Roaming\\Python\\Python39\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\u001b[0m in \u001b[0;36mstart_session\u001b[1;34m(self, capabilities, browser_profile)\u001b[0m\n\u001b[0;32m    250\u001b[0m         parameters = {\"capabilities\": w3c_caps,\n\u001b[0;32m    251\u001b[0m                       \"desiredCapabilities\": capabilities}\n\u001b[1;32m--> 252\u001b[1;33m         \u001b[0mresponse\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mexecute\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mCommand\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mNEW_SESSION\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mparameters\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    253\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[1;34m'sessionId'\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mresponse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    254\u001b[0m             \u001b[0mresponse\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mresponse\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'value'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Roaming\\Python\\Python39\\site-packages\\selenium\\webdriver\\remote\\webdriver.py\u001b[0m in \u001b[0;36mexecute\u001b[1;34m(self, driver_command, params)\u001b[0m\n\u001b[0;32m    319\u001b[0m         \u001b[0mresponse\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcommand_executor\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mexecute\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdriver_command\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mparams\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    320\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mresponse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 321\u001b[1;33m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0merror_handler\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mcheck_response\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mresponse\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    322\u001b[0m             response['value'] = self._unwrap_value(\n\u001b[0;32m    323\u001b[0m                 response.get('value', None))\n",
      "\u001b[1;32m~\\AppData\\Roaming\\Python\\Python39\\site-packages\\selenium\\webdriver\\remote\\errorhandler.py\u001b[0m in \u001b[0;36mcheck_response\u001b[1;34m(self, response)\u001b[0m\n\u001b[0;32m    240\u001b[0m                 \u001b[0malert_text\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mvalue\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'alert'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'text'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    241\u001b[0m             \u001b[1;32mraise\u001b[0m \u001b[0mexception_class\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmessage\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mscreen\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mstacktrace\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0malert_text\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 242\u001b[1;33m         \u001b[1;32mraise\u001b[0m \u001b[0mexception_class\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmessage\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mscreen\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mstacktrace\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    243\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    244\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_value_or_default\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mobj\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mdefault\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mSessionNotCreatedException\u001b[0m: Message: session not created\nfrom tab crashed\n  (Session info: headless chrome=90.0.4430.212)\n"
     ]
    }
   ],
   "source": [
    "# 3\n",
    "# Searching in the LOVD database using selenium webscraping.\n",
    "# The path to the chromedriver needs to be defined in PATH\n",
    "# driver will open and interact with the website and has an optional option\n",
    "# In this case the option is 'headless' which means the driver won't visibly open the website and do the search\n",
    "# removing the options=option part will disable this so you can follow how the search is done\n",
    "\n",
    "# The elements can be found by rightclicking on the page when visiting the website and clicking inspect\n",
    "# The script finds the elements mostly by xpath\n",
    "\n",
    "option = webdriver.ChromeOptions()\n",
    "\n",
    "option.add_argument('headless')\n",
    "PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "#Try is used so the script doesn't fail completely if something goes wrong with one website (it doesn't load, the site got updated, it can't find the search, ...)\n",
    "\n",
    "try:\n",
    "\n",
    "    if brca_input == \"BRCA1\" and sequence_type == \"NUCLEOTIDE\":\n",
    "\n",
    "        # The script will search for the searchbar on the site and enter the nucleotide search in it.\n",
    "\n",
    "        driver.get(\"https://databases.lovd.nl/shared/variants/BRCA1/unique\")\n",
    "        searchbar = driver.find_element_by_name(\"search_VariantOnTranscript/DNA\")\n",
    "        searchbar.send_keys(search)\n",
    "        searchbar.send_keys(Keys.RETURN)\n",
    "    \n",
    "        #time.sleep() is used to pause the script to give the website the time to load\n",
    "\n",
    "        time.sleep(1)\n",
    "\n",
    "        hit = False\n",
    "    # The script will search for the results table using xpath and iterate over it. hit = False is used so it only adds the relevant rows containing the search query.\n",
    "\n",
    "        count = 0\n",
    "        table = driver.find_element_by_xpath(\"//table[@id='viewlistTable_CustomVL_VOTunique_VOG_BRCA1']\")\n",
    "        for row in table.find_elements_by_xpath(\"//tr[@class='data']\"):\n",
    "            #The script will iterate over the results table in the page row by row\n",
    "            count = 0\n",
    "            for table_data in row.find_elements_by_xpath(\".//td\"):\n",
    "                count = count + 1\n",
    "                # For a row, the script will iterate over each cell\n",
    "                # It will keep a count for each time it goes over a cell\n",
    "                # when it reaches the 4th cell AND the search is in this cell it will append the information to the word document\n",
    "                # hit = True for this row so the script knows to append the information of the 9th cell to the word document aswell\n",
    "                if count == 4 and search in table_data.text:\n",
    "                    if range_check == True:\n",
    "                        cDNA = table_data.text\n",
    "                        cDNA_split = re.split(\"[+, ., \\-, >, _, a-zA-Z]\", cDNA)\n",
    "                        cDNA_split_plusmin = cDNA_split[3]\n",
    "\n",
    "                        if not any(i == cDNA_split_plusmin for i in pattern_string):\n",
    "                            row_cells = table_word.add_row().cells\n",
    "                            row_cells[0].text = \"LOVD\"\n",
    "                            row_cells[1].text = brca_input\n",
    "                            row_cells[2].text = table_data.text\n",
    "                            hit = True\n",
    "\n",
    "                    elif range_check == False:\n",
    "                        row_cells = table_word.add_row().cells\n",
    "                        row_cells[0].text = \"LOVD\"\n",
    "                        row_cells[1].text = brca_input\n",
    "                        row_cells[2].text = table_data.text\n",
    "                        hit = True\n",
    "                        \n",
    "                if count == 6 and hit == True:\n",
    "                    row_cells[3].text = table_data.text\n",
    "\n",
    "                if count == 6 and search in table_data.text:\n",
    "                    row_cells = table_word.add_row().cells\n",
    "                    row_cells[0].text = \"LOVD\"\n",
    "                    row_cells[1].text = brca_input\n",
    "                    row_cells[3].text = table_data.text\n",
    "                    hit = True\n",
    "\n",
    "                if count == 9 and hit == True:\n",
    "                    row_cells[4].text = table_data.text\n",
    "                    hit = False               \n",
    "\n",
    "\n",
    "\n",
    "        # It will append the results to the word document alongside with a hyperlink of the current url location\n",
    "        p = document.add_paragraph(\"Results: \")    \n",
    "        site_location = str(driver.current_url)\n",
    "        add_hyperlink(p, 'Link', driver.current_url)\n",
    "\n",
    "        driver.quit()\n",
    "\n",
    "    elif brca_input == \"BRCA2\" and sequence_type == \"NUCLEOTIDE\":\n",
    "        #If BRCA2 is given as input it will start the search on the BRCA2 tab of the website\n",
    "        driver.get(\"https://databases.lovd.nl/shared/variants/BRCA2/unique\")\n",
    "        searchbar = driver.find_element_by_name(\"search_VariantOnTranscript/DNA\")\n",
    "        searchbar.send_keys(search)\n",
    "        searchbar.send_keys(Keys.RETURN)\n",
    "    \n",
    "        time.sleep(1)\n",
    "\n",
    "        hit = False\n",
    "        count = 0\n",
    "        table = driver.find_element_by_xpath(\"//table[@id='viewlistTable_CustomVL_VOTunique_VOG_BRCA2']\")\n",
    "        for row in table.find_elements_by_xpath(\"//tr[@class='data']\"):\n",
    "            count = 0\n",
    "            for table_data in row.find_elements_by_xpath(\".//td\"):\n",
    "                count = count + 1\n",
    "                if count == 4 and search in table_data.text:\n",
    "                    if range_check == True:\n",
    "                        cDNA = table_data.text\n",
    "                        cDNA_split = re.split(\"[+, ., \\-, >, _, a-zA-Z]\", cDNA)\n",
    "                        cDNA_split_plusmin = cDNA_split[3]\n",
    "\n",
    "                        if not any(i == cDNA_split_plusmin for i in pattern_string):\n",
    "                            row_cells = table_word.add_row().cells\n",
    "                            row_cells[0].text = \"LOVD\"\n",
    "                            row_cells[1].text = brca_input\n",
    "                            row_cells[2].text = table_data.text\n",
    "                            hit = True\n",
    "\n",
    "                    elif range_check == False:\n",
    "                        row_cells = table_word.add_row().cells\n",
    "                        row_cells[0].text = \"LOVD\"\n",
    "                        row_cells[1].text = brca_input\n",
    "                        row_cells[2].text = table_data.text\n",
    "                        hit = True\n",
    "                        print(table_data.text)\n",
    "\n",
    "                if count == 6 and hit == True:\n",
    "                    row_cells[3].text = table_data.text\n",
    "                    print(table_data.text)\n",
    "\n",
    "\n",
    "                elif count == 6 and search in table_data.text:\n",
    "                    # Searching on protein lev\n",
    "                    row_cells = table_word.add_row().cells\n",
    "                    row_cells[0].text = \"LOVD\"\n",
    "                    row_cells[1].text = brca_input\n",
    "                    row_cells[3].text = table_data.text\n",
    "                    hit = True\n",
    "\n",
    "                if count == 9 and hit == True:\n",
    "                    row_cells[4].text = table_data.text\n",
    "                    hit = False\n",
    "                    print(table_data.text)\n",
    "\n",
    "        p = document.add_paragraph(\"Results: \")    \n",
    "        site_location = str(driver.current_url)\n",
    "        add_hyperlink(p, 'Link', driver.current_url)\n",
    "\n",
    "        driver.quit()\n",
    "\n",
    "    elif brca_input == \"BRCA1\" and sequence_type == \"PROTEIN\":\n",
    "\n",
    "        # The script will search for the searchbar on the site and enter the nucleotide search in it.\n",
    "\n",
    "        driver.get(\"https://databases.lovd.nl/shared/variants/BRCA1/unique\")\n",
    "        searchbar = driver.find_element_by_xpath(\"//*[@id='viewlistTable_CustomVL_VOTunique_VOG_BRCA2']/thead/tr/th[6]/input\")\n",
    "        searchbar.send_keys(search)\n",
    "        searchbar.send_keys(Keys.RETURN)\n",
    "    \n",
    "        #time.sleep() is used to pause the script to give the website the time to load\n",
    "\n",
    "        time.sleep(1)\n",
    "\n",
    "        hit = False\n",
    "    # The script will search for the results table using xpath and iterate over it. hit = False is used so it only adds the relevant rows containing the search query.\n",
    "\n",
    "        count = 0\n",
    "        table = driver.find_element_by_xpath(\"//table[@id='viewlistTable_CustomVL_VOTunique_VOG_BRCA1']\")\n",
    "        for row in table.find_elements_by_xpath(\"//tr[@class='data']\"):\n",
    "            #The script will iterate over the results table in the page row by row\n",
    "            count = 0\n",
    "            for table_data in row.find_elements_by_xpath(\".//td\"):\n",
    "                count = count + 1\n",
    "                # For a row, the script will iterate over each cell\n",
    "                # It will keep a count for each time it goes over a cell\n",
    "                # when it reaches the 4th cell AND the search is in this cell it will append the information to the word document\n",
    "                # hit = True for this row so the script knows to append the information of the 9th cell to the word document aswell\n",
    "                if count == 4 and search in table_data.text:\n",
    "                    if range_check == True:\n",
    "                        cDNA = table_data.text\n",
    "                        cDNA_split = re.split(\"[+, ., \\-, >, _, a-zA-Z]\", cDNA)\n",
    "                        cDNA_split_plusmin = cDNA_split[3]\n",
    "\n",
    "                        if not any(i == cDNA_split_plusmin for i in pattern_string):\n",
    "                            row_cells = table_word.add_row().cells\n",
    "                            row_cells[0].text = \"LOVD\"\n",
    "                            row_cells[1].text = brca_input\n",
    "                            row_cells[2].text = table_data.text\n",
    "                            hit = True\n",
    "\n",
    "                    elif range_check == False:\n",
    "                        row_cells = table_word.add_row().cells\n",
    "                        row_cells[0].text = \"LOVD\"\n",
    "                        row_cells[1].text = brca_input\n",
    "                        row_cells[2].text = table_data.text\n",
    "                        hit = True\n",
    "                        \n",
    "                if count == 6 and hit == True:\n",
    "                    row_cells[3].text = table_data.text\n",
    "\n",
    "                if count == 6 and search in table_data.text:\n",
    "                    row_cells = table_word.add_row().cells\n",
    "                    row_cells[0].text = \"LOVD\"\n",
    "                    row_cells[1].text = brca_input\n",
    "                    row_cells[3].text = table_data.text\n",
    "                    hit = True\n",
    "\n",
    "                if count == 9 and hit == True:\n",
    "                    row_cells[4].text = table_data.text\n",
    "                    hit = False               \n",
    "\n",
    "\n",
    "\n",
    "        # It will append the results to the word document alongside with a hyperlink of the current url location\n",
    "        p = document.add_paragraph(\"Results: \")    \n",
    "        site_location = str(driver.current_url)\n",
    "        add_hyperlink(p, 'Link', driver.current_url)\n",
    "\n",
    "        driver.quit()\n",
    "\n",
    "    elif brca_input == \"BRCA2\" and sequence_type == \"PROTEIN\":\n",
    "        #If BRCA2 is given as input it will start the search on the BRCA2 tab of the website\n",
    "        driver.get(\"https://databases.lovd.nl/shared/variants/BRCA2/unique\")\n",
    "        searchbar = driver.find_element_by_xpath(\"//*[@id='viewlistTable_CustomVL_VOTunique_VOG_BRCA2']/thead/tr/th[6]/input\")\n",
    "        searchbar.send_keys(search)\n",
    "        searchbar.send_keys(Keys.RETURN)\n",
    "    \n",
    "        time.sleep(1)\n",
    "\n",
    "        hit = False\n",
    "        count = 0\n",
    "        table = driver.find_element_by_xpath(\"//table[@id='viewlistTable_CustomVL_VOTunique_VOG_BRCA2']\")\n",
    "        for row in table.find_elements_by_xpath(\"//tr[@class='data']\"):\n",
    "            count = 0\n",
    "            for table_data in row.find_elements_by_xpath(\".//td\"):\n",
    "                count = count + 1\n",
    "                if count == 4 and search in table_data.text:\n",
    "                    if range_check == True:\n",
    "                        cDNA = table_data.text\n",
    "                        cDNA_split = re.split(\"[+, ., \\-, >, _, a-zA-Z]\", cDNA)\n",
    "                        cDNA_split_plusmin = cDNA_split[3]\n",
    "\n",
    "                        if not any(i == cDNA_split_plusmin for i in pattern_string):\n",
    "                            row_cells = table_word.add_row().cells\n",
    "                            row_cells[0].text = \"LOVD\"\n",
    "                            row_cells[1].text = brca_input\n",
    "                            row_cells[2].text = table_data.text\n",
    "                            hit = True\n",
    "\n",
    "                    elif range_check == False:\n",
    "                        row_cells = table_word.add_row().cells\n",
    "                        row_cells[0].text = \"LOVD\"\n",
    "                        row_cells[1].text = brca_input\n",
    "                        row_cells[2].text = table_data.text\n",
    "                        hit = True\n",
    "                        print(table_data.text)\n",
    "\n",
    "                if count == 6 and hit == True:\n",
    "                    row_cells[3].text = table_data.text\n",
    "                    print(table_data.text)\n",
    "\n",
    "\n",
    "                elif count == 6 and search in table_data.text:\n",
    "                    # Searching on protein lev\n",
    "                    row_cells = table_word.add_row().cells\n",
    "                    row_cells[0].text = \"LOVD\"\n",
    "                    row_cells[1].text = brca_input\n",
    "                    row_cells[3].text = table_data.text\n",
    "                    hit = True\n",
    "\n",
    "                if count == 9 and hit == True:\n",
    "                    row_cells[4].text = table_data.text\n",
    "                    hit = False\n",
    "                    print(table_data.text)\n",
    "\n",
    "        p = document.add_paragraph(\"Results: \")    \n",
    "        site_location = str(driver.current_url)\n",
    "        add_hyperlink(p, 'Link', driver.current_url)\n",
    "\n",
    "        driver.quit()\n",
    "\n",
    "except Exception:\n",
    "    pass\n",
    "print(\"LOVD\")"
   ]
  },
  {
   "source": [
    "\n",
    "# Lindor pdf\n",
    "# 4\n",
    "# The script will read the pdf file containing the tables\n",
    "# It will create a folder and paste the tables in excel files, this step is skipped if this folder already exists\n",
    "# The script will iterate over each excel file in the folder\n",
    "\n",
    "\n",
    "tables_pdf = read_pdf(folder_location + pdf_lindor, pages=\"17-28\")\n",
    "tables_output_location = \"/tables excel lindor 2012\"\n",
    "\n",
    "tables_folder = folder_location + tables_output_location\n",
    "\n",
    "if not os.path.exists(tables_folder):\n",
    "    os.makedirs(tables_folder)\n",
    "\n",
    "for i, table_pdf in enumerate(tables_pdf, start=1):\n",
    "    table_pdf.to_excel(os.path.join(tables_folder, f\"table_{i}.xlsx\"), index=False)\n",
    "\n",
    "# The excel files are divided in different groups based on the similarity of their structure (Equal amount of columns, columns that have the same meaning)\n",
    "group_one = ('table_1.xlsx')\n",
    "group_two = (\"2.xlsx\", \"3.xlsx\", \"4.xlsx\", \"5.xlsx\")\n",
    "group_three = (\"6.xlsx\", \"7.xlsx\", \"8.xlsx\", \"9.xlsx\", \"10.xlsx\")\n",
    "group_four = (\"11.xlsx\")\n",
    "group_five = (\"12.xlsx\")\n",
    "\n",
    "document.add_heading(\"Lindor 2012 pdf\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=5)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'HGSV nucleotide'\n",
    "hdr_cells[2].text = 'BIC nucleotide'\n",
    "hdr_cells[3].text = \"HGVS protein\"\n",
    "hdr_cells[4].text = 'Class'\n",
    "\n",
    "tables_dir = os.listdir(tables_folder)\n",
    "\n",
    "#In the first half it will search for BRCA1 genes\n",
    "\n",
    "for table_file in tables_dir:\n",
    "    if brca_input == \"BRCA1\":\n",
    "        if table_file.endswith(group_one):\n",
    "\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "\n",
    "            for row in sheet.iter_rows():\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[3].text = brca_input\n",
    "                        row_cells[4].text = str(cell.offset(column=5).value)\n",
    "                        #protein?\n",
    "\n",
    "\n",
    "        if table_file.endswith(group_two):\n",
    "\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "\n",
    "            for row in sheet.iter_rows(max_col = 3, min_col = 3):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[2].text = cell.value\n",
    "                        row_cells[3].text = str(cell.offset(column=-1).value)\n",
    "                        row_cells[4].text = str(cell.offset(column=5).value)\n",
    "                        #bic\n",
    "                        \n",
    "            for row in sheet.iter_rows(max_col = 4, min_col = 4):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[1].text = cell.value\n",
    "                        row_cells[2].text = str(cell.offset(column=-2).value)\n",
    "                        row_cells[4].text = str(cell.offset(column=4).value)\n",
    "                        #hgvs\n",
    "\n",
    "            for row in sheet.iter_rows(max_col = 2, min_col = 2):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[1].text = str(cell.offset(column=2).value)\n",
    "                        row_cells[2].text = str(cell.offset(column=1).value)\n",
    "                        row_cells[3].text = cell.value\n",
    "                        row_cells[4].text = str(cell.offset(column=6).value)\n",
    "                        #hgvs protein\n",
    "                        \n",
    "            \n",
    "        if table_file.endswith(group_four):\n",
    "\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "            for row in sheet.iter_rows():\n",
    "                for cell in row:\n",
    "                    if brca_input and search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = cell.value\n",
    "                        row_cells[4].text = str(cell.offset(column=2).value)\n",
    "                        #brca1/brca2\n",
    "#Second half: it will search for the BRCA2 gene in the applicable file\n",
    "\n",
    "    if brca_input == \"BRCA2\":\n",
    "        if table_file.endswith(group_three):\n",
    "\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "            for row in sheet.iter_rows(max_col = 3, min_col = 3):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[2].text = cell.value\n",
    "                        row_cells[3].text = str(cell.offset(column=-1).value)\n",
    "                        row_cells[4].text = str(cell.offset(column=8).value)\n",
    "                        #bic\n",
    "\n",
    "            for row in sheet.iter_rows(max_col = 4, min_col = 4):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[1].text = cell.value\n",
    "                        row_cells[3].text = str(cell.offset(column=-2).value)\n",
    "                        row_cells[4].text = str(cell.offset(column=5).value)\n",
    "                        #hgvs\n",
    "\n",
    "            for row in sheet.iter_rows(max_col = 2, min_col = 2):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[1].text = str(cell.offset(column=2).value)\n",
    "                        row_cells[2].text = str(cell.offset(column=1).value)\n",
    "                        row_cells[3].text = cell.value\n",
    "                        row_cells[4].text = str(cell.offset(column=9).value)\n",
    "                        #hgvs protein\n",
    "\n",
    "\n",
    "        if table_file.endswith(group_four):\n",
    "\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "            for row in sheet.iter_rows():\n",
    "                for cell in row:\n",
    "                    if brca_input and search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = cell.value\n",
    "                        row_cells[3].text = str(cell.offset(column=2).value)\n",
    "                        \n",
    "        if table_file.endswith(group_five):\n",
    "\n",
    "            wb = load_workbook(tables_folder + \"/\" + table_file)\n",
    "            sheet = wb.active\n",
    "            for row in sheet.iter_rows(max_col = 4, min_col = 4):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[1].text = cell.value\n",
    "                        row_cells[3].text = str(cell.offset(column=5).value)\n",
    "                        #hgvs\n",
    "\n",
    "                     \n",
    "            for row in sheet.iter_rows(max_col = 5, min_col = 5):\n",
    "                for cell in row:\n",
    "                    if search in str(cell.value):\n",
    "                        row_cells = table.add_row().cells\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[2].text = cell.value\n",
    "                        row_cells[3].text = str(cell.offset(column=4).value)\n",
    "                        #bic\n",
    "print(\"lindor\")"
   ],
   "cell_type": "code",
   "metadata": {
    "tags": []
   },
   "execution_count": 69,
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "lindor\n"
     ]
    }
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "#5) Reading excel sheet: bicbnl 27  fh 2-10-13_nov2013\n",
    "# Python gives warning it can't parse the header but the script still works\n",
    "# Similar to the other excel scripts except it differentiates between two tabs first.\n",
    "\n",
    "wb = load_workbook(folder_location + bicbnl)\n",
    "\n",
    "document.add_heading(\"LOB\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=5)\n",
    "table.allow_outfit = True\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'c.nom'\n",
    "hdr_cells[2].text = 'Type'\n",
    "hdr_cells[3].text = 'opm LOB'\n",
    "hdr_cells[4].text = 'opmerking'\n",
    "    \n",
    "if brca_input == \"BRCA1\":\n",
    "    # The script will open the right sheet first.\n",
    "    sheet = wb['totaal BRCA1']\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 5, min_col = 5):\n",
    "        # It will search in the 5th column for the nucleotide\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                row_cells = table.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                row_cells[1].text = cell.value\n",
    "                row_cells[2].text = str(cell.offset(column=2).value) # column G (type)\n",
    "                row_cells[3].text = str(cell.offset(column=14).value) # column S (opmerking LOB indeling)\n",
    "                row_cells[4].text = str(cell.offset(column=15).value) # column T (opmerking) \n",
    "\n",
    "elif brca_input == \"BRCA2\":\n",
    "    sheet = wb['totaal BRCA2']\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 5, min_col = 5):\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                row_cells = table.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                row_cells[1].text = cell.value\n",
    "                row_cells[2].text = str(cell.offset(column=2).value)\n",
    "                row_cells[3].text = str(cell.offset(column=14).value)\n",
    "                row_cells[4].text = str(cell.offset(column=15).value)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "<Element {http://schemas.openxmlformats.org/wordprocessingml/2006/main}hyperlink at 0x210941191c0>"
      ]
     },
     "metadata": {},
     "execution_count": 71
    }
   ],
   "source": [
    "from eutils import Client\n",
    "from Bio import Entrez\n",
    "import requests as rq\n",
    "import pprint\n",
    "#4/5/6 clinvar\n",
    "eclient = Client(api_key=\"84c4d9fa268d5e41511ce024a0cf00537808\")\n",
    "\n",
    "gene_esearch = eclient.esearch(db='clinvar',term='{}[gene] AND {}'.format(brca_input, search))\n",
    "\n",
    "id_list = gene_esearch.ids\n",
    "\n",
    "document.add_heading(\"Clinvar\")\n",
    "table_word = document.add_table(rows=1, cols=4)\n",
    "table_word.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table_word.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'cDNA'\n",
    "hdr_cells[2].text = 'Classification'\n",
    "hdr_cells[3].text = 'Review Status'\n",
    "\n",
    "\n",
    "for i in id_list:\n",
    "    query = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi?db=clinvar&id={}&retmode=json\".format(i)\n",
    "    response = rq.get(query).json()\n",
    "    id_string = str(i)\n",
    "    if search in response[\"result\"][id_string][\"title\"]:\n",
    "        row_cells = table_word.add_row().cells\n",
    "        row_cells[0].text = brca_input\n",
    "        sequence = response[\"result\"][id_string][\"title\"]\n",
    "        sequence_split = sequence.split(\":\")\n",
    "        row_cells[1].text = sequence_split[1]\n",
    "        row_cells[2].text = response[\"result\"][id_string][\"clinical_significance\"][\"description\"]\n",
    "        if response[\"result\"][id_string][\"clinical_significance\"][\"review_status\"] == \"practice guideline\":\n",
    "            row_cells[3].text = \"****\"\n",
    "        elif response[\"result\"][id_string][\"clinical_significance\"][\"review_status\"] == \"reviewed by expert panel\":\n",
    "            row_cells[3].text = \"***\"\n",
    "        elif response[\"result\"][id_string][\"clinical_significance\"][\"review_status\"] == \"criteria provided, multiple submitters, no conflicts\":\n",
    "            row_cells[3].text = \"**\"\n",
    "        elif response[\"result\"][id_string][\"clinical_significance\"][\"review_status\"] == \"criteria provided, conflicting interpretations\" or 'criteria provided, single submitter':\n",
    "            row_cells[3].text = \"*\"\n",
    "        elif response[\"result\"][id_string][\"clinical_significance\"][\"review_status\"] == \"no assertion for the individual variant\" or \"no assertion criteria provided\" or \"no assertion provided\":\n",
    "            row_cells[3].text = \"/\"\n",
    "\n",
    "current_url = \"https://www.ncbi.nlm.nih.gov/clinvar/?term={}%5Bgene%5D+{}\".format(brca_input, search)\n",
    "p = document.add_paragraph(\"Results: \")    \n",
    "site_location = current_url\n",
    "add_hyperlink(p, 'Link', current_url)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "# BIC\n",
    "# Priority 6\n",
    "folder_location = \"C:/Users/jensv/Desktop/BRCA_-_prostate_cancer_and_ovarian_cancer_-_PARPi\"\n",
    "bic_brca1 = \"/brca1_data_BIC.txt\"\n",
    "bic_brca2 = \"/brca2_data_BIC.txt\"\n",
    "\n",
    "document.add_heading(\"BIC\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=4)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'HGVS cDNA'\n",
    "hdr_cells[2].text = \"Protein\"\n",
    "hdr_cells[3].text = 'Category'\n",
    "\n",
    "\n",
    "if brca_input == \"BRCA1\":\n",
    "    full_path = folder_location + bic_brca1\n",
    "    with open(full_path) as csv_file:\n",
    "        csv_reader = csv.reader(csv_file, delimiter=\"\\t\")\n",
    "        for row in csv_reader:\n",
    "            count = 0\n",
    "            hit = False\n",
    "            for cell in row:\n",
    "                count = count + 1\n",
    "                if count == 8 and search in cell:\n",
    "                    row_cells = table.add_row().cells\n",
    "                    row_cells[0].text = brca_input\n",
    "                    row_cells[1].text = cell\n",
    "                    count = count + 1\n",
    "                    hit = True\n",
    "\n",
    "                elif count == 9 and search in cell:\n",
    "                    row_cells = table.add_row().cells\n",
    "                    row_cells[0].text = brca_input\n",
    "                    row_cells[1].text = cell\n",
    "                    count = count + 1\n",
    "                    hit = True\n",
    "\n",
    "                if count == 16 and hit == True:\n",
    "                    row_cells[3].text = cell\n",
    "                    hit = False\n",
    "\n",
    "elif brca_input == \"BRCA2\":\n",
    "    full_path = folder_location + bic_brca2\n",
    "    with open(full_path) as csv_file:\n",
    "        csv_reader = csv.reader(csv_file, delimiter=\"\\t\")\n",
    "        for row in csv_reader:\n",
    "            count = 0\n",
    "            hit = False\n",
    "            for cell in row:\n",
    "                count = count + 1\n",
    "                if count == 8 and search in cell:\n",
    "                    row_cells = table.add_row().cells\n",
    "                    row_cells[0].text = brca_input\n",
    "                    row_cells[1].text = cell\n",
    "                    count = count + 1\n",
    "                    hit = True\n",
    "\n",
    "                elif count == 9 and search in cell:\n",
    "                    row_cells = table.add_row().cells\n",
    "                    row_cells[0].text = brca_input\n",
    "                    row_cells[1].text = cell\n",
    "                    count = count + 1\n",
    "                    hit = True\n",
    "\n",
    "                if count == 16 and hit == True:\n",
    "                    row_cells[3].text = cell\n",
    "                    hit = False\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "'\\nimport re\\ntest = \\'c.301+7G>A\\'\\ntest_split = re.split(\"[+, ., -, >, d-zD-Z]\", test)\\nprint(test_split)\\n'"
      ]
     },
     "metadata": {},
     "execution_count": 73
    }
   ],
   "source": [
    "\"\"\"\n",
    "import re\n",
    "test = 'c.301+7G>A'\n",
    "test_split = re.split(\"[+, ., -, >, d-zD-Z]\", test)\n",
    "print(test_split)\n",
    "\"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Enigma rules manual made table\n",
    "# 4?\n",
    "enigma_table_brca1 = \"/ENIGMA RULES BRCA1.xlsx\"\n",
    "enigma_table_brca2 = \"/ENIGMA RULES BRCA2.xlsx\"\n",
    "\n",
    "if brca_input == \"BRCA1\":\n",
    "    if sequence_type == \"PROTEIN\":\n",
    "        wb = load_workbook(folder_location + enigma_table_brca1)\n",
    "        sheet = wb.active\n",
    "        document.add_heading(\"ENIGMA\")\n",
    "        table = document.add_table(rows=1, cols=6)\n",
    "\n",
    "        table.style = 'Colorful List Accent 1'\n",
    "        hdr_cells = table.rows[0].cells\n",
    "        hdr_cells[0].text = 'Gene'\n",
    "        hdr_cells[1].text = 'Domain/motif'\n",
    "        hdr_cells[2].text = 'AA start'\n",
    "        hdr_cells[3].text = 'AA end'\n",
    "        hdr_cells[4].text = 'AA alterations'\n",
    "        hdr_cells[5].text = 'Classification'\n",
    "\n",
    "        # searching for between AA start and AA end\n",
    "        found = False\n",
    "        for row in sheet.iter_rows(min_row=2, min_col=2, max_col=2):\n",
    "            for cell in row:\n",
    "                if int(search_mod) >= int(cell.value) and int(search_mod) <= int(cell.offset(column=1).value):\n",
    "                    found = True\n",
    "                    row_cells = table.add_row().cells\n",
    "    \n",
    "                    if search in cell.offset(column=2).value:\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[1].text = str(cell.offset(column=-1).value)\n",
    "                        row_cells[2].text = str(cell.value)\n",
    "                        row_cells[3].text = str(cell.offset(column=1).value)\n",
    "                        row_cells[4].text = search\n",
    "                        row_cells[5].text = \"Class 5\"\n",
    "                    else:\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[1].text = str(cell.offset(column=-1).value)\n",
    "                        row_cells[2].text = str(cell.value)\n",
    "                        row_cells[3].text = str(cell.offset(column=1).value)\n",
    "                        row_cells[4].text = search\n",
    "                        row_cells[5].text = \"Class 3\"\n",
    "\n",
    "    elif sequence_type == \"NUCLEOTIDE\":\n",
    "        for row in sheet.iter_rows(min_col=4, max_col=4):\n",
    "            for cell in row:\n",
    "                if search in str(cell.value) and found == False:\n",
    "                    row_cells = table.add_row().cells\n",
    "                    row_cells[0].text = brca_input\n",
    "                    row_cells[1].text = str(cell.offset(column=-3).value)\n",
    "                    row_cells[4].text = search\n",
    "                    row_cells[5].text = \"Class 5\"\n",
    "\n",
    "elif brca_input == \"BRCA2\":\n",
    "    if sequence_type == \"PROTEIN\":\n",
    "        wb = load_workbook(folder_location + enigma_table_brca2)\n",
    "        sheet = wb.active\n",
    "        document.add_heading(\"ENIGMA\")\n",
    "        table = document.add_table(rows=1, cols=6)\n",
    "\n",
    "        table.style = 'Colorful List Accent 1'\n",
    "        hdr_cells = table.rows[0].cells\n",
    "        hdr_cells[0].text = 'Gene'\n",
    "        hdr_cells[1].text = 'Domain/motif'\n",
    "        hdr_cells[2].text = 'AA start'\n",
    "        hdr_cells[3].text = 'AA end'\n",
    "        hdr_cells[4].text = 'AA alterations'\n",
    "        hdr_cells[5].text = 'Classification'\n",
    "\n",
    "        # searching for between AA start and AA end\n",
    "        found = False\n",
    "        for row in sheet.iter_rows(min_row=2, min_col=2, max_col=2):\n",
    "            for cell in row:\n",
    "                if int(search_mod) >= int(cell.value) and int(search_mod) <= int(cell.offset(column=1).value):\n",
    "                    found = True\n",
    "                    row_cells = table.add_row().cells\n",
    "\n",
    "                    if search in cell.offset(column=2).value:\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[1].text = str(cell.offset(column=-1).value)\n",
    "                        row_cells[2].text = str(cell.value)\n",
    "                        row_cells[3].text = str(cell.offset(column=1).value)\n",
    "                        row_cells[4].text = search\n",
    "                        row_cells[5].text = \"Class 5\"\n",
    "                    else:\n",
    "                        row_cells[0].text = brca_input\n",
    "                        row_cells[1].text = str(cell.offset(column=-1).value)\n",
    "                        row_cells[2].text = str(cell.value)\n",
    "                        row_cells[3].text = str(cell.offset(column=1).value)\n",
    "                        row_cells[4].text = search\n",
    "                        row_cells[5].text = \"Class 3\"\n",
    "\n",
    "    elif sequence_type == \"NUCLEOTIDE\":\n",
    "        for row in sheet.iter_rows(min_col=4, max_col=4):\n",
    "            for cell in row:\n",
    "                if search in str(cell.value) and found == False:\n",
    "                    row_cells = table.add_row().cells\n",
    "                    row_cells[0].text = brca_input\n",
    "                    row_cells[1].text = str(cell.offset(column=-3).value)\n",
    "                    row_cells[4].text = search\n",
    "                    row_cells[5].text = \"Class 5\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "#2) Reading excel sheet : allenigmavariants_BICsubmission_2013-07-01\n",
    "#7A\n",
    "\n",
    "# The first column in this file is BRCA1/BRCA2, the second column is hgvsnucleotide and the fourth column is bic nucleotide\n",
    "# The script will iterate over each row, if the brca_input matches with the gene in the 1st column AND the search is found in the 2nd OR 4th column it will give the class in column 7 as output\n",
    "\n",
    "#cell.offset(column=2).value gives an error as output if the cell is empty but not when it is converted to string first.\n",
    "\n",
    "# Zoeken op protein change\n",
    "wb = load_workbook(folder_location + variants_excel)\n",
    "\n",
    "sheet = wb.active\n",
    "\n",
    "document.add_heading(\"ENIGMA\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=5)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = \"List\"\n",
    "hdr_cells[1].text = 'Gene'\n",
    "hdr_cells[2].text = 'HGSV nucleotide'\n",
    "hdr_cells[3].text = 'HGSV protein'\n",
    "hdr_cells[4].text = 'Class'\n",
    "\n",
    "\n",
    "for row in sheet.iter_rows():\n",
    "    for cell in row:\n",
    "        # if BRCA gene is in 1st column AND nucleotide search is in 2ndcolumn (hgvs nucleotide)\n",
    "        if brca_input in str(cell.value) and search in str(cell.offset(column=1).value) or brca_input in str(cell.value) and search in str(cell.offset(column=2).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = \"Class 1 lijst met SNV's\"\n",
    "            row_cells[1].text = cell.value #gene\n",
    "            row_cells[2].text = cell.offset(column=1).value #hgvs nucleotide\n",
    "            row_cells[3].text = str(cell.offset(column=2).value) #hgvs protein\n",
    "            row_cells[4].text = str(cell.offset(column=6).value) #class\n",
    "        '''\n",
    "        #else if BRCA gene is in 1st column AND nucleotide search is in 4th column (bic nucleotide)  \n",
    "        elif brca_input in str(cell.value) and search in str(cell.offset(column=3).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value #gene\n",
    "            row_cells[2].text = cell.offset(column=3).value #bic nucleotide\n",
    "            row_cells[3].text = cell.offset(column=6).value #class\n",
    "        '''\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "#3)Reading excel sheet: Published Multifactoral data enigma\n",
    "#7B samen met 7A\n",
    "# This is likewise to searching in the allenigmavariants_BICsubmission_2013-07-01 excel file\n",
    "\n",
    "\n",
    "wb = load_workbook(folder_location + enigma)\n",
    "\n",
    "sheet = wb.active\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "for row in sheet.iter_rows():\n",
    "    for cell in row:\n",
    "        if brca_input in str(cell.value) and search in str(cell.offset(column=1).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = \"Enigma MLM lijst\"\n",
    "            row_cells[1].text = cell.value\n",
    "            row_cells[2].text = cell.offset(column=1).value\n",
    "            row_cells[3].text = str(cell.offset(column=2).value)\n",
    "            row_cells[4].text = cell.offset(column=6).value\n",
    "            Result = False\n",
    "        if brca_input in str(cell.value) and search in str(cell.offset(column=2).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = \"Enigma MLM lijst\"\n",
    "            row_cells[1].text = cell.value\n",
    "            row_cells[2].text = cell.offset(column=1).value\n",
    "            row_cells[3].text = str(cell.offset(column=2).value)\n",
    "            row_cells[4].text = cell.offset(column=6).value\n",
    "\n",
    "            '''\n",
    "        elif brca_input in str(cell.value) and search in str(cell.offset(column=3).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value\n",
    "            row_cells[2].text = cell.offset(column=3).value\n",
    "            row_cells[3].text = cell.offset(column=6).value\n",
    "            Result = False\n",
    "            '''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "#1) Reading excel sheet: 2020 functional categorization of BRCA1 VUS - CCR\n",
    "# Zoeken op protein\n",
    "# 8\n",
    "# Bij group B SNVS smaen met enigma en function scores\n",
    "\n",
    "### Add heading to word document\n",
    "document.add_heading(\"BRCA1 SNV lijst met functionele data (PMID: 32546644)\", 1)\n",
    "\n",
    "### Create table in word\n",
    "table = document.add_table(rows=1, cols=4)\n",
    "\n",
    "### Loading a style template and creating the header cells\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'DNA variant'\n",
    "hdr_cells[1].text = 'Cisplatin Assay'\n",
    "hdr_cells[2].text = 'Olaparib Assay'\n",
    "hdr_cells[3].text = 'DR-GFP assay'\n",
    "\n",
    "\n",
    "#If the input is BRCA1 the script will load the excel file. It will iterate over each row of the first column (max_col = 1) of the active sheet.\n",
    "#If the search input is found in the first column the script will add the values of the 3rd, 4th and 5th column\n",
    "#Example cell.offset(column=2).value means that the script will take the value of the 3rd column (1st column + 2 to the right = 3)\n",
    "if brca_input == \"BRCA1\":\n",
    "\n",
    "    wb = load_workbook(folder_location + functional_categorization)\n",
    "\n",
    "    sheet = wb.active\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 2, min_col = 2):\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                    row_cells = table.add_row().cells\n",
    "                    row_cells[0].text = cell.value\n",
    "                    row_cells[1].text = cell.offset(column=1).value\n",
    "                    row_cells[2].text = cell.offset(column=2).value\n",
    "                    row_cells[3].text = cell.offset(column=3).value\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "########Franklin database\n",
    "#9\n",
    "try:\n",
    "\n",
    "    option = webdriver.ChromeOptions()\n",
    "    option.add_argument('headless')\n",
    "    PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "    driver = webdriver.Chrome(PATH, options = option)\n",
    "\n",
    "    document.add_heading(\"Franklin/Varsome\")\n",
    "    table_word = document.add_table(rows=1, cols=3)\n",
    "    table_word.style = 'Colorful List Accent 1'\n",
    "    hdr_cells = table_word.rows[0].cells\n",
    "    hdr_cells[0].text = \"Engine\"\n",
    "    hdr_cells[1].text = 'Gene'\n",
    "    hdr_cells[2].text = 'Classification'\n",
    "    row_cells = table_word.add_row().cells\n",
    "\n",
    "    driver.get(\"https://franklin.genoox.com/clinical-db/home\")\n",
    "\n",
    "    time.sleep(10)\n",
    "\n",
    "    #a pop up appeared on this page which can be removed by pressing escape\n",
    "    ActionChains(driver).send_keys(Keys.ESCAPE).perform()\n",
    "\n",
    "    #The script will search for the searchbar and the button(somatic)\n",
    "    searchbar = driver.find_element_by_xpath(\"/html/body/app-root/div/gnx-home-page/div/gnx-search/div      [2]/input\")\n",
    "\n",
    "    button = driver.find_element_by_xpath(\"/html/body/app-root/div/gnx-home-page/div/gnx-search/div[2]/     gnx-variant-type-toggle/button[2]\")\n",
    "\n",
    "    button.click()\n",
    "\n",
    "    searchbar.send_keys(brca_input + \":\" + search)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "\n",
    "    time.sleep(10)\n",
    "\n",
    "    driver.find_element_by_xpath(\"//*[@id='mat-tab-label-2-3']\").click()\n",
    "\n",
    "    time.sleep(10)\n",
    "\n",
    "    #classification = driver.find_element_by_xpath(\"//*[@id='mat-tab-content-2-3']/div/gnx-application-container/div/div/gnx-acmg-classification-app/gnx-result-page/div/gnx-result-top-section/div[2]/gnx-classifiction-gauge/div/div/svg/g/g/text[2]\")\n",
    "    classification = driver.find_element_by_xpath(\"//*[@class='indicator-text ng-star-inserted']\")\n",
    "\n",
    "    time.sleep(10)\n",
    "\n",
    "    row_cells[0].text = \"Franklin\"\n",
    "    row_cells[1].text = brca_input + \":\" + search\n",
    "    row_cells[2].text = classification.text\n",
    "    franklin_class = classification.text\n",
    "    print(classification.text)\n",
    "    print(franklin_class)\n",
    "\n",
    "\n",
    "    p = document.add_paragraph(\"Results: \")    \n",
    "    site_location = str(driver.current_url)\n",
    "    add_hyperlink(p, 'Link', driver.current_url)\n",
    "\n",
    "    driver.quit()\n",
    "\n",
    "except Exception:\n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Varsome login required if too many searches are done\n",
    "#10\n",
    "try:\n",
    "\n",
    "    option = webdriver.ChromeOptions()\n",
    "    option.add_argument('headless')\n",
    "    PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "    driver = webdriver.Chrome(PATH, options=option)\n",
    "\n",
    "    driver.get(\"https://varsome.com/\")\n",
    "\n",
    "    time.sleep(1)\n",
    "\n",
    "    searchbar = driver.find_element_by_xpath(\"//*[@id='search']\")\n",
    "    searchbar.send_keys(brca_input + \" \" + search)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "\n",
    "    time.sleep(1)\n",
    "\n",
    "    #Popup asking to accept cookies \n",
    "    cookie = driver.find_element_by_xpath(\"//*[@id='onetrust-accept-btn-handler']\")\n",
    "    cookie.click()\n",
    "\n",
    "    button = driver.find_element_by_xpath(\"//*[@id='proceedBtn']/h4\")\n",
    "    button.click()\n",
    "\n",
    "    time.sleep(10)\n",
    "    try:\n",
    "        element = WebDriverWait(driver, 10).until(EC.presence_of_element_located((By.ID, \"acmg_annotation\")))\n",
    "\n",
    "    except TimeoutException:\n",
    "        print(\"Timed out waiting for page to load\")\n",
    "    \n",
    "\n",
    "    classification = driver.find_element_by_xpath(\"//*[@id='acmg_annotation']/div/div[1]/div/div[2]/div/    div/div[1]/div/div/span/div[2]/div/div\")\n",
    "\n",
    "    row_cells = table_word.add_row().cells\n",
    "    row_cells[0].text = \"Varsome\"\n",
    "    row_cells[1].text = brca_input + \"\" + search\n",
    "    row_cells[2].text = classification.text\n",
    "\n",
    "    p = document.add_paragraph(\"Results: \")    \n",
    "    site_location = str(driver.current_url)\n",
    "    add_hyperlink(p, 'Link', driver.current_url)\n",
    "    print(driver.current_url)\n",
    "\n",
    "    driver.quit()\n",
    "\n",
    "except Exception:\n",
    "    pass\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "https://sge.gs.washington.edu/BRCA1/#tab-9159-2\n"
     ]
    }
   ],
   "source": [
    "#Database of Functional Classifications of BRCA1 Variants based on Saturation Genome Editing\n",
    "#\n",
    "\n",
    "try:\n",
    "    \n",
    "    if brca_input == \"BRCA1\":\n",
    "\n",
    "        option = webdriver.ChromeOptions()\n",
    "        option.add_argument('headless')\n",
    "        PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "        driver = webdriver.Chrome(PATH, options = option)\n",
    "\n",
    "        document.add_heading(\"Database of Functional Classifications of BRCA1 Variants based on Saturation Genome Editing\")\n",
    "        table_word = document.add_table(rows=1, cols=3)\n",
    "        table_word.style = 'Colorful List Accent 1'\n",
    "        hdr_cells = table_word.rows[0].cells\n",
    "        hdr_cells[0].text = 'Gene'\n",
    "        hdr_cells[1].text = 'Transcript Variant'\n",
    "        hdr_cells[2].text = 'Functional CLass'\n",
    "\n",
    "\n",
    "        driver.get(\"https://sge.gs.washington.edu/BRCA1/#tab-9159-2\")\n",
    "\n",
    "        time.sleep(1)\n",
    "        #The script will look up and click on the right tab\n",
    "        tab = driver.find_element_by_xpath(\"//*[@id='main_page']/div/div/nav/div/ul/li[2]/a\")\n",
    "        tab.click()\n",
    "        '''\n",
    "        tab = driver.find_element_by_xpath(\"//ul[@class='nav navbar-nav']\")\n",
    "        tab.find_element_by_xpath(\".//a[@href='#tab-9159-2']\").click()\n",
    "        '''\n",
    "        #time.sleep(5)\n",
    "        try:\n",
    "            element = WebDriverWait(driver, 10).until(EC.presence_of_element_located((By.ID, \"DataTables_Table_0_filter\")))\n",
    "\n",
    "        except TimeoutException:\n",
    "            print(\"Timed out waiting for page to load\")\n",
    "\n",
    "        # look up the searchbar and enter the query\n",
    "        searchbar = driver.find_element_by_xpath(\"//*[@id='DataTables_Table_0_filter']/label/input\")\n",
    "        searchbar.send_keys(search)\n",
    "        searchbar.send_keys(Keys.RETURN)\n",
    "    \n",
    "        table = driver.find_element_by_xpath(\"//*[@id='DataTables_Table_0']/tbody\")\n",
    " \n",
    "        count = 0\n",
    "        time.sleep(5)\n",
    "        hit = False\n",
    "\n",
    "        for row in table.find_elements_by_xpath(\"//tr\"):\n",
    "            count = 0\n",
    "            for table_data in row.find_elements_by_xpath((\".//td\")):\n",
    "                count = count + 1\n",
    "                if count == 6 and search in table_data.text:\n",
    "                    row_cells = table_word.add_row().cells\n",
    "                    row_cells[0].text = brca_input\n",
    "                    hit = True\n",
    "                    row_cells[1].text = table_data.text \n",
    "                if count == 10 and hit == True:\n",
    "                    row_cells[2].text = table_data.text\n",
    "                    hit = False\n",
    "\n",
    "    p = document.add_paragraph(\"Results: \")    \n",
    "    site_location = str(driver.current_url)\n",
    "    add_hyperlink(p, 'Link', driver.current_url)\n",
    "    print(driver.current_url)\n",
    "\n",
    "    driver.quit()\n",
    "except Exception:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "https://www.oncokb.org/gene/BRCA1/5074\n"
     ]
    }
   ],
   "source": [
    "#Oncokb.org\n",
    "# Using webscraping because the API is not free.\n",
    "\n",
    "try:\n",
    "\n",
    "    option = webdriver.ChromeOptions()\n",
    "    option.add_argument('headless')\n",
    "    PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "    driver = webdriver.Chrome(PATH, options = option)\n",
    "\n",
    "    driver.get(\"https://www.oncokb.org/\")\n",
    "\n",
    "    document.add_heading(\"OncoKB\")\n",
    "    table_word = document.add_table(rows=1, cols=2)\n",
    "    table_word.style = 'Colorful List Accent 1'\n",
    "    hdr_cells = table_word.rows[0].cells\n",
    "    hdr_cells[0].text = 'Search'\n",
    "    hdr_cells[1].text = 'Results'\n",
    "\n",
    "    #time.sleep(5)\n",
    "\n",
    "    searchbar = driver.find_element_by_xpath(\"/html/body/div[1]/div[2]/div[3]/div/div/div[3]/div/div/       div/div[1]/div[2]/div/input\")\n",
    "    searchbar.send_keys(brca_input + \" \" + search)\n",
    "    time.sleep(2)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "    time.sleep(5)\n",
    "    result = driver.find_element_by_xpath(\"//*[@id='root']/div[2]/div[3]/div/div/div/div[2]/div/div[2]\")\n",
    "\n",
    "    row_cells = table_word.add_row().cells\n",
    "    row_cells[0].text = brca_input + search\n",
    "    row_cells[1].text = result.text\n",
    "\n",
    "    p = document.add_paragraph(\"Results: \")    \n",
    "    site_location = str(driver.current_url)\n",
    "    add_hyperlink(p, 'Link', driver.current_url)\n",
    "    print(driver.current_url)\n",
    "\n",
    "    driver.quit()\n",
    "    \n",
    "except Exception:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "https://pubmed.ncbi.nlm.nih.gov/?term=BRCA1+5074\n"
     ]
    }
   ],
   "source": [
    "# Pubmed\n",
    "p = document.add_paragraph(\"pubmed: \") \n",
    "pubmed_site = \"https://pubmed.ncbi.nlm.nih.gov/?term=\"  \n",
    "pubmed_full = pubmed_site + brca_input + \"+\" + search\n",
    "add_hyperlink(p, 'Link', pubmed_full)\n",
    "print(pubmed_full)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "document.save(\"output.docx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "window.mainloop()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "'\\npdf_enigma = \"/ENIGMA_Rules_2017-06-29.pdf\"\\nimport tabula\\nimport tabulate\\nfolder_location = \"C:/Users/jensv/Desktop/BRCA_-_prostate_cancer_and_ovarian_cancer_-_PARPi\"\\ntables_enigma = read_pdf(folder_location + pdf_enigma, pages=\\'14-16\\')\\ntables_output_location = \"/tables ENIGMA\"\\n\\nprint(tables_enigma)\\ntables_folder = folder_location + tables_output_location\\n\\nif not os.path.exists(tables_folder):\\n    os.makedirs(tables_folder)\\n\\nfor i, table_pdf in enumerate(tables_enigma, start=1):\\n    table_pdf.to_excel(os.path.join(tables_folder, f\"table_{i}.xlsx\"), index=False)\\n'"
      ]
     },
     "metadata": {},
     "execution_count": 85
    }
   ],
   "source": [
    "'''\r\n",
    "pdf_enigma = \"/ENIGMA_Rules_2017-06-29.pdf\"\r\n",
    "import tabula\r\n",
    "import tabulate\r\n",
    "folder_location = \"C:/Users/jensv/Desktop/BRCA_-_prostate_cancer_and_ovarian_cancer_-_PARPi\"\r\n",
    "tables_enigma = read_pdf(folder_location + pdf_enigma, pages='14-16')\r\n",
    "tables_output_location = \"/tables ENIGMA\"\r\n",
    "\r\n",
    "print(tables_enigma)\r\n",
    "tables_folder = folder_location + tables_output_location\r\n",
    "\r\n",
    "if not os.path.exists(tables_folder):\r\n",
    "    os.makedirs(tables_folder)\r\n",
    "\r\n",
    "for i, table_pdf in enumerate(tables_enigma, start=1):\r\n",
    "    table_pdf.to_excel(os.path.join(tables_folder, f\"table_{i}.xlsx\"), index=False)\r\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "\"\\n# API requst varsome 1 request per day\\nimport requests\\nimport json\\n\\nresponse = requests.get('https://api.varsome.com/lookup/gene/BRCA1/hg19?add-all-data=1')\\n\\nprint(response.json())\\n\\n\""
      ]
     },
     "metadata": {},
     "execution_count": 86
    }
   ],
   "source": [
    "'''\n",
    "# API requst varsome 1 request per day\n",
    "import requests\n",
    "import json\n",
    "\n",
    "response = requests.get('https://api.varsome.com/lookup/gene/BRCA1/hg19?add-all-data=1')\n",
    "\n",
    "print(response.json())\n",
    "\n",
    "'''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "output_type": "error",
     "ename": "SyntaxError",
     "evalue": "EOL while scanning string literal (<ipython-input-87-44987550c2d7>, line 21)",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-87-44987550c2d7>\"\u001b[1;36m, line \u001b[1;32m21\u001b[0m\n\u001b[1;33m    ''''\u001b[0m\n\u001b[1;37m        \n^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m EOL while scanning string literal\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "#LOVD database\n",
    "import requests as rq\n",
    "import json\n",
    "import pprint\n",
    "import feedparser\n",
    "import bs4\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "query = \"\"\"\n",
    "https://databases.lovd.nl/shared/api/rest.php/variants/IVD\n",
    "\"\"\"\n",
    "response = rq.get(\"https://databases.lovd.nl/shared/variants/BRCA1/unique1\")\n",
    "\n",
    "responsesoup = bs4.BeautifulSoup(response.text)\n",
    "for i in responsesoup:\n",
    "    if \"5074\" in i:\n",
    "        print(i)\n",
    "''''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "output_type": "error",
     "ename": "NameError",
     "evalue": "name 'BeautifulSoup' is not defined",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-23-af9a62e61fc0>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     15\u001b[0m \u001b[0mresponse\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mrq\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"https://databases.lovd.nl/shared/api/rest.php/variants/BRCA1\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     16\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 17\u001b[1;33m \u001b[0mresponsesoup\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mBeautifulSoup\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mresponse\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m'lxml'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m: name 'BeautifulSoup' is not defined"
     ]
    }
   ],
   "source": [
    "'''\n",
    "#LOVD database\n",
    "import requests as rq\n",
    "import json\n",
    "import pprint\n",
    "import feedparser\n",
    "import bs4\n",
    "from xml.etree import cElementTree as ET\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "query = \"\"\"\n",
    "https://databases.lovd.nl/shared/api/rest.php/variants/IVD\n",
    "\"\"\"\n",
    "'''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "output_type": "error",
     "ename": "AttributeError",
     "evalue": "'str' object has no attribute 'items'",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-109-20d51a559409>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      6\u001b[0m \u001b[0mheaders\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;34m'PMAK-60aeb05843c1c000510a6469-4b4474e4637a9568b84f7ab2aabd4f8c21'\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      7\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 8\u001b[1;33m \u001b[0mresponse\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mrequests\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mrequest\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"GET\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0murl\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mheaders\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mheaders\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mpayload\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      9\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     10\u001b[0m \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mresponse\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtext\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Roaming\\Python\\Python39\\site-packages\\requests\\api.py\u001b[0m in \u001b[0;36mrequest\u001b[1;34m(method, url, **kwargs)\u001b[0m\n\u001b[0;32m     59\u001b[0m     \u001b[1;31m# cases, and look like a memory leak in others.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     60\u001b[0m     \u001b[1;32mwith\u001b[0m \u001b[0msessions\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mSession\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0msession\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 61\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0msession\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mrequest\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmethod\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mmethod\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0murl\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0murl\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     62\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     63\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Roaming\\Python\\Python39\\site-packages\\requests\\sessions.py\u001b[0m in \u001b[0;36mrequest\u001b[1;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[0;32m    526\u001b[0m             \u001b[0mhooks\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mhooks\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    527\u001b[0m         )\n\u001b[1;32m--> 528\u001b[1;33m         \u001b[0mprep\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mprepare_request\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mreq\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    529\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    530\u001b[0m         \u001b[0mproxies\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mproxies\u001b[0m \u001b[1;32mor\u001b[0m \u001b[1;33m{\u001b[0m\u001b[1;33m}\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Roaming\\Python\\Python39\\site-packages\\requests\\sessions.py\u001b[0m in \u001b[0;36mprepare_request\u001b[1;34m(self, request)\u001b[0m\n\u001b[0;32m    454\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    455\u001b[0m         \u001b[0mp\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mPreparedRequest\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 456\u001b[1;33m         p.prepare(\n\u001b[0m\u001b[0;32m    457\u001b[0m             \u001b[0mmethod\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mrequest\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmethod\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mupper\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    458\u001b[0m             \u001b[0murl\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mrequest\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0murl\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Roaming\\Python\\Python39\\site-packages\\requests\\models.py\u001b[0m in \u001b[0;36mprepare\u001b[1;34m(self, method, url, headers, files, data, params, auth, cookies, hooks, json)\u001b[0m\n\u001b[0;32m    315\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mprepare_method\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmethod\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    316\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mprepare_url\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0murl\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mparams\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 317\u001b[1;33m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mprepare_headers\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mheaders\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    318\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mprepare_cookies\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcookies\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    319\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mprepare_body\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mfiles\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mjson\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\AppData\\Roaming\\Python\\Python39\\site-packages\\requests\\models.py\u001b[0m in \u001b[0;36mprepare_headers\u001b[1;34m(self, headers)\u001b[0m\n\u001b[0;32m    447\u001b[0m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mheaders\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mCaseInsensitiveDict\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    448\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mheaders\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 449\u001b[1;33m             \u001b[1;32mfor\u001b[0m \u001b[0mheader\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mheaders\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    450\u001b[0m                 \u001b[1;31m# Raise exception on invalid header value.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    451\u001b[0m                 \u001b[0mcheck_header_validity\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mheader\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mAttributeError\u001b[0m: 'str' object has no attribute 'items'"
     ]
    }
   ],
   "source": [
    "'''\n",
    "import requests\n",
    "\n",
    "url = \"https://api.genoox.com/v1/auth/login?email=<jens.verheyden@student.howest.be>&password=<Rainbowroad1>\"\n",
    "\n",
    "payload = \"{\\r\\n    \\\"email\\\":\\\"<jens.verheyden@student.howest.be>\\\",\\r\\n     \\\"password\\\":\\\"<Rainbowroad1>\\\"\\r\\n}\"\n",
    "headers = 'PMAK-60aeb05843c1c000510a6469-4b4474e4637a9568b84f7ab2aabd4f8c21'\n",
    "\n",
    "response = requests.request(\"GET\", url, headers=headers, data=payload)\n",
    "\n",
    "print(response.text)\n",
    "\n",
    " \n",
    "import requests\n",
    "\n",
    "url = \"https://franklin.genoox.com/api/parse_search\"\n",
    "\n",
    "payload = \"{\\\"search_text_input\\\": \\\"vhl:c.293A>C\\\"}\"\n",
    "headers = {}\n",
    "\n",
    "response = requests.request(\"POST\", url, headers=headers, data=payload)\n",
    "\n",
    "print(response.text)\n",
    "\n",
    "\n",
    "print(token)\n",
    "url = \"https://api.genoox.com/v1/variant/search?search_text=VHL:c.293A>G\"\n",
    " \n",
    "payload={}\n",
    "headers = {}\n",
    " \n",
    "response = requests.request(\"GET\", url, headers=headers, data=payload)\n",
    " \n",
    "print(response.text)\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "output_type": "error",
     "ename": "TypeError",
     "evalue": "lookup() missing 1 required positional argument: 'query'",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-122-3fa2c61bef24>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      8\u001b[0m \u001b[0mapi\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mVarSomeAPIClient\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      9\u001b[0m \u001b[1;31m# fetch information about a variant into a dictionary\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 10\u001b[1;33m \u001b[0mresult\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mapi\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mlookup\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'chr7-140453136-A-T'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mparams\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;33m{\u001b[0m\u001b[1;34m'add-source-databases'\u001b[0m\u001b[1;33m:\u001b[0m \u001b[1;34m'gnomad-exomes,refseq-transcripts'\u001b[0m\u001b[1;33m}\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mref_genome\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'hg19'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     11\u001b[0m \u001b[0mannotated_variant\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mAnnotatedVariant\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m**\u001b[0m\u001b[0mresult\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mTypeError\u001b[0m: lookup() missing 1 required positional argument: 'query'"
     ]
    }
   ],
   "source": [
    "'''\n",
    "####https://github.com/saphetor/varsome-api-client-python/archive/master.zip\n",
    "from varsome_api.client import VarSomeAPIClient\n",
    "# API key is not required for single variant lookups\n",
    "from varsome_api.client import VarSomeAPIClient\n",
    "from varsome_api.models.variant import AnnotatedVariant\n",
    "# API key is not required for single variant lookups\n",
    "api_key = 'Your token'\n",
    "api = VarSomeAPIClient\n",
    "# fetch information about a variant into a dictionary\n",
    "result = api.lookup('chr7-140453136-A-T', params={'add-source-databases': 'gnomad-exomes,refseq-transcripts'}, ref_genome='hg19')\n",
    "annotated_variant = AnnotatedVariant(**result)\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "output_type": "error",
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-2-9e77d902c908>, line 1)",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-2-9e77d902c908>\"\u001b[1;36m, line \u001b[1;32m1\u001b[0m\n\u001b[1;33m    varsome_api_run.py -g -hg19 -k api_key -q 'chr7-140453136-A-T' -p add-all-data=1\u001b[0m\n\u001b[1;37m                                   ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "varsome_api_run.py -g -hg19 -k api_key -q 'chr7-140453136-A-T' -p add-all-data=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ]
}