{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.4"
  },
  "orig_nbformat": 2,
  "kernelspec": {
   "name": "python394jvsc74a57bd070de0b75b2a3f35b1bb892b463cd56f2ed390cd02c9f7d4168b3d96fea273b5e",
   "display_name": "Python 3.9.4 64-bit"
  },
  "metadata": {
   "interpreter": {
    "hash": "70de0b75b2a3f35b1bb892b463cd56f2ed390cd02c9f7d4168b3d96fea273b5e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "###pip is the package installer for python3. The command \"py -m pip install\"\n",
    "### installs the packages where in \"-m pip\" executes pip using the latest python\n",
    "### interpreter installed in windows.\n",
    "\n",
    "import os\n",
    "import subprocess\n",
    "import re\n",
    "\n",
    "packagelist = \"openpyxl\", \"beautifulsoup4\"\n",
    "\n",
    "for package in packagelist:\n",
    "    try:\n",
    "        import package\n",
    "    except:\n",
    "        subprocess.call([\"py\",\"-m\",\"pip\", \"install\", package], stdout=subprocess.DEVNULL)\n",
    "       # os.system(\"py -m pip install \" + package)\n",
    "\n",
    "from openpyxl import Workbook\n",
    "from openpyxl import load_workbook\n",
    "from selenium import webdriver\n",
    "from selenium.webdriver.common.keys import Keys\n",
    "import time\n",
    "from selenium.webdriver.common.by import By\n",
    "from selenium.webdriver.support.ui import WebDriverWait\n",
    "from selenium.webdriver.support import expected_conditions as EC\n",
    "from selenium.webdriver.common.action_chains import ActionChains\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "###########Input\n",
    "#print(\"BRCA1/BRCA2?\")\n",
    "#brca_input = input()\n",
    "brca_input = \"BRCA1\"\n",
    "brca_input = brca_input.upper()\n",
    "#print(\"Search?\")\n",
    "#search = input()\n",
    "search = \"4308\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "####Setting up the location of the excel files\n",
    "\n",
    "folder_location = \"C:/Users/jensv/Desktop/BRCA_-_prostate_cancer_and_ovarian_cancer_-_PARPi\"\n",
    "functional_categorization = \"/2020 BRCA1 functional categorization of BRCA1 VUS - CCR.xlsx\"\n",
    "variants_excel = \"/allenigmavariants_BICsubmission_2013-07-01.xlsx\"\n",
    "bicbnl = \"/bicbnl 27  fh 2-10-13_nov2013.xlsx\"\n",
    "enigma = \"/BRCA_Multifac_published data_ENIGMAwebsite 2015-03-27.xlsx\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "##################Write data to excel\n",
    "#Writing data to word\n",
    "from docx import Document\n",
    "from docx.shared import Cm\n",
    "#py -m pip install --user python-docx\n",
    "document = Document()\n",
    "\n",
    "document.add_heading(\"2020 functional categorization of BRCA1 VUS - CCR\", 1)\n",
    "\n",
    "table = document.add_table(rows=1, cols=4)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'DNA variant'\n",
    "hdr_cells[1].text = 'Cisplatin Assay'\n",
    "hdr_cells[2].text = 'Olaparib Assay'\n",
    "hdr_cells[3].text = 'DR-GFP assay'\n",
    "\n",
    "document.save('test.docx')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "#1) Reading excel sheet: 2020 functional categorization of BRCA1 VUS - CCR\n",
    "#The script will load the excel workbook and sheet and create empty lists of the different assays and found dna variants\n",
    "#it will iterate over each cell and append the results to the lists if the input is found in the first column: DNA variant\n",
    "\n",
    "##################Write data to excel\n",
    "#Writing data to word\n",
    "from docx import Document\n",
    "from docx.shared import Cm\n",
    "#py -m pip install --user python-docx\n",
    "document = Document()\n",
    "\n",
    "document.add_heading(\"2020 functional categorization of BRCA1 VUS - CCR\", 1)\n",
    "\n",
    "table = document.add_table(rows=1, cols=4)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'DNA variant'\n",
    "hdr_cells[1].text = 'Cisplatin Assay'\n",
    "hdr_cells[2].text = 'Olaparib Assay'\n",
    "hdr_cells[3].text = 'DR-GFP assay'\n",
    "\n",
    "\n",
    "if brca_input == \"BRCA1\":\n",
    "\n",
    "    wb = load_workbook(folder_location + functional_categorization)\n",
    "\n",
    "    sheet = wb.active\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 1):\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                row_cells = table.add_row().cells\n",
    "                Result = True\n",
    "                if Result == True:\n",
    "                    row_cells[0].text = cell.value\n",
    "                    row_cells[1].text = cell.offset(column=2).value\n",
    "                    row_cells[2].text = cell.offset(column=3).value\n",
    "                    row_cells[3].text = cell.offset(column=4).value\n",
    "                    Result = False\n",
    "\n",
    "document.save('test.docx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "#2) Reading excel sheet : allenigmavariants_BICsubmission_2013-07-01\n",
    "#The script will iterate over cell but will only append to the lists if the search is found in the cell and the right BRCA input is present in the first column\n",
    "#Another elif statement is added which seperates the search found between HGVS and BIC\n",
    "\n",
    "wb = load_workbook(folder_location + variants_excel)\n",
    "\n",
    "sheet = wb.active\n",
    "\n",
    "search = \"4308\"\n",
    "brca_input = \"BRCA1\"\n",
    "\n",
    "document.add_heading(\"allenigmavariants_BICsubmission_2013-07-01\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=4)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'HGSV nucleotide'\n",
    "hdr_cells[2].text = 'BIC nucleotide'\n",
    "hdr_cells[3].text = 'Class'\n",
    "\n",
    "\n",
    "for row in sheet.iter_rows():\n",
    "    for cell in row:\n",
    "        if brca_input in str(cell.value) and search in str(cell.offset(column=1).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value\n",
    "            row_cells[1].text = cell.offset(column=1).value\n",
    "            row_cells[3].text = cell.offset(column=6).value\n",
    "            Result = False\n",
    "            \n",
    "        elif brca_input in str(cell.value) and search in str(cell.offset(column=3).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value\n",
    "            row_cells[2].text = cell.offset(column=3).value\n",
    "            row_cells[3].text = cell.offset(column=6).value\n",
    "            Result = False\n",
    "\n",
    "\n",
    "document.save('test.docx')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "#3)Reading excel sheet: Published Multifactoral data enigma\n",
    "\n",
    "\n",
    "wb = load_workbook(folder_location + enigma)\n",
    "\n",
    "sheet = wb.active\n",
    "\n",
    "\n",
    "search = \"2002\"\n",
    "brca_input = \"BRCA1\"\n",
    "\n",
    "document.add_heading(\"BRCA_Multifac_published data_ENIGMAwebsite 2015-03-27\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=4)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'HGSV nucleotide'\n",
    "hdr_cells[2].text = 'BIC nucleotide'\n",
    "hdr_cells[3].text = 'Class'\n",
    "\n",
    "\n",
    "for row in sheet.iter_rows():\n",
    "    for cell in row:\n",
    "        if brca_input in str(cell.value) and search in str(cell.offset(column=1).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value\n",
    "            row_cells[1].text = cell.offset(column=1).value\n",
    "            row_cells[3].text = cell.offset(column=6).value\n",
    "            Result = False\n",
    "        elif brca_input in str(cell.value) and search in str(cell.offset(column=3).value):\n",
    "            row_cells = table.add_row().cells\n",
    "            row_cells[0].text = cell.value\n",
    "            row_cells[2].text = cell.offset(column=3).value\n",
    "            row_cells[3].text = cell.offset(column=6).value\n",
    "            Result = False\n",
    "\n",
    "document.save('test.docx')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "#4) Reading excel sheet: bicbnl 27  fh 2-10-13_nov2013\n",
    "\n",
    "wb = load_workbook(folder_location + bicbnl)\n",
    "\n",
    "search = \"2685\"\n",
    "brca_input = \"BRCA1\"\n",
    "\n",
    "document.add_heading(\"bicbnl 27  fh 2-10-13_nov2013\")\n",
    "\n",
    "table = document.add_table(rows=1, cols=5)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'Gene'\n",
    "hdr_cells[1].text = 'c.nom'\n",
    "hdr_cells[2].text = 'Type'\n",
    "hdr_cells[3].text = 'opm LOB indeling'\n",
    "hdr_cells[4].text = 'opmerking'\n",
    "\n",
    "if brca_input == \"BRCA1\":\n",
    "    sheet = wb['totaal BRCA1']\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 5, min_col = 5):\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                row_cells = table.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                row_cells[1].text = cell.value\n",
    "                row_cells[2].text = str(cell.offset(column=2).value)\n",
    "                row_cells[3].text = str(cell.offset(column=14).value)\n",
    "                row_cells[4].text = str(cell.offset(column=15).value)\n",
    "                Result = False\n",
    "\n",
    "elif brca_input == \"BRCA2\":\n",
    "    sheet = wb['totaal BRCA2']\n",
    "\n",
    "    for row in sheet.iter_rows(max_col = 5, min_col = 5):\n",
    "        for cell in row:\n",
    "            if search in str(cell.value):\n",
    "                row_cells = table.add_row().cells\n",
    "                row_cells[0].text = brca_input\n",
    "                row_cells[1].text = cell.value\n",
    "                row_cells[2].text = str(cell.offset(column=2).value)\n",
    "                row_cells[3].text = str(cell.offset(column=14).value)\n",
    "                row_cells[4].text = str(cell.offset(column=15).value)\n",
    "                Result = False\n",
    "\n",
    "document.save('test.docx')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "['BRCA1']\n['c.5095C>T']\n['NA, pathogenic, pathogenic (dominant), unclassified']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "from selenium import webdriver\n",
    "from selenium.webdriver.common.keys import Keys\n",
    "import time\n",
    "from selenium.webdriver.common.by import By\n",
    "from selenium.webdriver.support.ui import WebDriverWait\n",
    "from selenium.webdriver.support import expected_conditions as EC\n",
    "\n",
    "#Install the corresponding version of chromedriver on https://sites.google.com/a/chromium.org/chromedriver/downloads and place the file in a map. Store the path to the file in the PATH variable below.\n",
    "brca_input=\"BRCA1\"\n",
    "\n",
    "PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "driver = webdriver.Chrome(PATH)\n",
    "\n",
    "lovd_gene = []\n",
    "lovd_nucleotide = []\n",
    "lovd_class = []\n",
    "\n",
    "if brca_input == \"BRCA1\":\n",
    "    driver.get(\"https://databases.lovd.nl/shared/variants/BRCA1/unique\")\n",
    "    searchbar = driver.find_element_by_name(\"search_VariantOnTranscript/DNA\")\n",
    "    search = \"5095\"\n",
    "    searchbar.send_keys(search)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "    \n",
    "    time.sleep(1)\n",
    "    \n",
    "    lovd_gene.append(brca_input)\n",
    "    count = 0\n",
    "    table = driver.find_element_by_xpath(\"//table[@id='viewlistTable_CustomVL_VOTunique_VOG_BRCA1']\")\n",
    "    for row in table.find_elements_by_xpath(\"//tr[@class='data']\"):\n",
    "        count = 0\n",
    "        for table_data in row.find_elements_by_xpath(\".//td\"):\n",
    "            count = count + 1\n",
    "            if count == 4:\n",
    "                lovd_nucleotide.append(table_data.text)\n",
    "            if count == 9:\n",
    "                lovd_class.append(table_data.text)\n",
    "\n",
    "    driver.quit()\n",
    "\n",
    "elif brca_input == \"BRCA2\":\n",
    "    driver.get(\"https://databases.lovd.nl/shared/variants/BRCA2/unique\")\n",
    "    searchbar = driver.find_element_by_name(\"search_VariantOnTranscript/DNA\")\n",
    "    search = \"5095\"\n",
    "    searchbar.send_keys(search)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "    \n",
    "    time.sleep(1)\n",
    "    \n",
    "    lovd_gene.append(brca_input)\n",
    "    count = 0\n",
    "    table = driver.find_element_by_xpath(\"//table[@id='viewlistTable_CustomVL_VOTunique_VOG_BRCA2']\")\n",
    "    for row in table.find_elements_by_xpath(\"//tr[@class='data']\"):\n",
    "        count = 0\n",
    "        for table_data in row.find_elements_by_xpath(\".//td\"):\n",
    "            count = count + 1\n",
    "            if count == 4:\n",
    "                lovd_nucleotide.append(table_data.text)\n",
    "            if count == 9:\n",
    "                lovd_class.append(table_data.text)\n",
    "\n",
    "    driver.quit()\n",
    "\n",
    "print(lovd_gene)\n",
    "print(lovd_nucleotide)\n",
    "print(lovd_class)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "BRCA2\n['c.5073A>T', 'c.5068_5073dup', 'c.5073A>G', 'c.5073dup', 'c.4845_4846del', 'c.5070_5073del', 'c.5073del']\n['Not Yet Reviewed', 'Not Yet Reviewed', 'Not Yet Reviewed', 'Pathogenic', 'Pathogenic', 'Pathogenic', 'Pathogenic']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "search = \"5073\"\n",
    "brca_input = \"BRCA2\"\n",
    "\n",
    "PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "driver = webdriver.Chrome(PATH)\n",
    "\n",
    "brca_exchange_gene = brca_input\n",
    "brca_exchange_nucleotide = []\n",
    "brca_exchange_class = []\n",
    "\n",
    "\n",
    "driver.get(\"https://brcaexchange.org/\")\n",
    "\n",
    "searchbar = driver.find_element_by_xpath(\"//input[@type='text']\")\n",
    "searchbar.send_keys(brca_input + \" \" +search)\n",
    "searchbar.send_keys(Keys.RETURN)\n",
    "\n",
    "time.sleep(5)\n",
    "table = driver.find_element_by_xpath(\"//*[@id='data-table-container']/div/table/tbody\")\n",
    "\n",
    "for row in table.find_elements_by_xpath(\"//tr\"):\n",
    "    count = 0\n",
    "    for table_data in row.find_elements_by_xpath('.//td'):\n",
    "        #print(table_data.text)\n",
    "        count = count + 1\n",
    "        if count == 2:\n",
    "            brca_exchange_nucleotide.append(table_data.text)\n",
    "        if count == 6:\n",
    "            brca_exchange_class.append(table_data.text)\n",
    "driver.quit()\n",
    "\n",
    "print(brca_exchange_gene)\n",
    "print(brca_exchange_nucleotide)\n",
    "print(brca_exchange_class)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "BRCA1\n['c.5073A>T', 'c.5073A>G', 'c.5073A>C']\n['LOF', 'FUNC', 'LOF']\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "search = \"5073\"\n",
    "brca_input = \"BRCA1\"\n",
    "if brca_input == \"BRCA1\":\n",
    "    PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "    driver = webdriver.Chrome(PATH)\n",
    "\n",
    "\n",
    "    washington_gene = brca_input\n",
    "    washington_nucleotide = []\n",
    "    washington_class = [] \n",
    "\n",
    "    driver.get(\"https://sge.gs.washington.edu/BRCA1/#tab-9159-2\")\n",
    "    time.sleep(1)\n",
    "    tab = driver.find_element_by_xpath(\"//ul[@class='nav navbar-nav']\")\n",
    "    tab.find_element_by_xpath(\".//a[@href='#tab-9159-2']\").click()\n",
    "\n",
    "    time.sleep(5)\n",
    "\n",
    "    searchbar = driver.find_element_by_xpath(\"//*[@id='DataTables_Table_0_filter']/label/input\")\n",
    "    searchbar.send_keys(search)\n",
    "    searchbar.send_keys(Keys.RETURN)\n",
    "    \n",
    "    table = driver.find_element_by_xpath(\"//*[@id='DataTables_Table_0']/tbody\")\n",
    " \n",
    "    count = 0\n",
    "    time.sleep(5)\n",
    "    hit = False\n",
    "\n",
    "    for row in table.find_elements_by_xpath(\"//tr\"):\n",
    "        count = 0\n",
    "        for table_data in row.find_elements_by_xpath((\".//td\")):\n",
    "            count = count + 1\n",
    "            if count == 6 and search in table_data.text:\n",
    "                hit = True\n",
    "                washington_nucleotide.append(table_data.text)\n",
    "            if count == 10 and hit == True:\n",
    "                washington_class.append(table_data.text)\n",
    "                hit = False\n",
    "                \n",
    "print(washington_gene)\n",
    "print(washington_nucleotide)\n",
    "print(washington_class)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Likely Pathogenic\n"
     ]
    }
   ],
   "source": [
    "########Franklin database\n",
    "\n",
    "search = \"BRCA2:c.50del\"\n",
    "\n",
    "PATH = \"C:\\Program Files (x86)\\chromedriver.exe\"\n",
    "driver = webdriver.Chrome(PATH)\n",
    "\n",
    "franklin_gene = search\n",
    "#franklin_class = []\n",
    "\n",
    "driver.get(\"https://franklin.genoox.com/clinical-db/home\")\n",
    "\n",
    "time.sleep(10)\n",
    "\n",
    "ActionChains(driver).send_keys(Keys.ESCAPE).perform()\n",
    "searchbar = driver.find_element_by_xpath(\"/html/body/app-root/div/gnx-home-page/div/gnx-search/div[2]/input\")\n",
    "button = driver.find_element_by_xpath(\"/html/body/app-root/div/gnx-home-page/div/gnx-search/div[2]/gnx-variant-type-toggle/button[2]\")\n",
    "\n",
    "button.click()\n",
    "\n",
    "searchbar.send_keys(search)\n",
    "searchbar.send_keys(Keys.RETURN)\n",
    "\n",
    "time.sleep(10)\n",
    "\n",
    "driver.find_element_by_xpath(\"//*[@id='mat-tab-label-2-3']\").click()\n",
    "\n",
    "time.sleep(3)\n",
    "\n",
    "classification = driver.find_element_by_xpath(\"//*[@class='indicator-text ng-star-inserted']\")\n",
    "\n",
    "franklin_class = classification.text\n",
    "print(franklin_class)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "##################Write data to excel\n",
    "#Writing data to word\n",
    "from docx import Document\n",
    "from docx.shared import Cm\n",
    "#py -m pip install --user python-docx\n",
    "document = Document()\n",
    "\n",
    "document.add_heading(\"2020 functional categorization of BRCA1 VUS - CCR\", 1)\n",
    "\n",
    "table = document.add_table(rows=1, cols=3)\n",
    "\n",
    "table.style = 'Colorful List Accent 1'\n",
    "hdr_cells = table.rows[0].cells\n",
    "hdr_cells[0].text = 'DNA variant'\n",
    "hdr_cells[1].text = 'Cisplatin Assay'\n",
    "hdr_cells[2].text = 'Olaparib Assay'\n",
    "hdr_cells[3].text = 'DR-GFP assay'\n",
    "\n",
    "\n",
    "\n",
    " nucleotide_functional_excel = []\n",
    "    cis_assay_excel = []\n",
    "    olaparib_assay_excel = []\n",
    "    dr_gfp_assay_excel = []\n",
    "\n",
    "document.save('test.docx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ]
}